{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genomic Evolution After Chemoradiotherapy in Anal Squamous Cell Carcinoma\n",
    "See the paper here: "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This IPython notebook contains code utilized to replicate any figures generated or calculations used in the results section of the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd # Version 0.17.1\n",
    "import numpy as np  # Version 1.10.1\n",
    "import glob as glob \n",
    "import math as math\n",
    "import scipy.stats as stats\n",
    "\n",
    "import matplotlib   # Version 1.5.0\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patheffects as PathEffects\n",
    "\n",
    "from matplotlib import gridspec\n",
    "from matplotlib import colors\n",
    "from matplotlib import artist\n",
    "from __future__ import division\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## We source some directories with functions or definitions to make this file cleaner\n",
    "\n",
    "# ASCC_colors.py has any color schemes used by figures\n",
    "execfile('subscripts/ASCC-colors.py')\n",
    "\n",
    "# ASCC_comut-functions contains functions that will be commonly used to \n",
    "execfile('subscripts/ASCC-comut_functions.py')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='table.of.contents'></a>\n",
    "## Table of Contents\n",
    "### Figures\n",
    "<a href='#figure1'>Figure 1: CoMut plot of Primary and Recurrent Tumors</a><br>\n",
    "<a href='#figure2'>Figure 2: Similarity bars</a><br>\n",
    "<a href='#figure3'>Figure 3: Mutational and Neoantigen burden</a><br>\n",
    "<a href='#supplementary.figure4'>Supplementary Figure 4: CoMut plot of Primary Tumors - No Recurrence & Recurrence</a><br>\n",
    "<a href='#supplementary.figure6a'>Supplementary Figure 6a: CoMut plot of Primary & Recurrent Pairs</a><br>\n",
    "### Calculations\n",
    "<a href='#methods.tumoronly.filter'>Methods: Applying tumor only filter to pilot cohort</a><br>\n",
    "<br>\n",
    "<a href='#results.primary.ascc'>Results: Genomic Landscape of Primary ASCC</a><br>\n",
    "<a href='#results.recurrent.ascc'>Results: Genomic Landscape of Recurrent ASCC</a><br>\n",
    "<a href='#results.paired.ascc'>Results: Genomic evolution in matched primary and recurrent tumors</a><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<a id='figure1'></a>\n",
    "## Figure 1: CoMut plot of Primary and Recurrent Tumors\n",
    "<a href='#table.of.contents'>Return to top</a><br>\n",
    "\n",
    "In this section, we will generate Figure 1, which contains 4 CoMut plots <br>\n",
    "...(A) will contain the primary tumors in the pilot cohort <br>\n",
    "...(B) will contain the primary tumors in the extension cohort <br>\n",
    "...(C) will contain the recurrent tumors in the pilot cohort <br>\n",
    "...(D) will contain the recurrent tumors in the extension cohort <br>\n",
    "...... Note: We display only one sample per patient per section <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# First and foremost, we import some files that contain annotations\n",
    "input_info_pilot = r'storage/reference/ASCC_samples-pilot.txt'\n",
    "info_pilot = pd.read_csv(input_info_pilot, sep = '\\t')\n",
    "\n",
    "input_info_extension = r'storage/reference/ASCC_samples-extension.txt'\n",
    "info_extension = pd.read_csv(input_info_extension, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## =================================\n",
    "# Import MAFs and format\n",
    "## =================================\n",
    "# We first import, extract columns of interest, and subset for our genes of interest\n",
    "\n",
    "# Import pilot cohort and create a 'read_depth' annotation, which will be used later\n",
    "df_pilot = pd.read_csv('storage/mutsig-pilot/final_analysis_set.maf', sep = '\\t',\n",
    "                      comment = '#', low_memory = False)\n",
    "df_pilot['read_depth'] = df_pilot['t_alt_count'] + df_pilot['t_ref_count']\n",
    "\n",
    "# Import extension cohort \n",
    "df_extension = pd.read_csv('filter_unmatched/ascc_genes/df_filter_pass.txt', sep = '\\t', \n",
    "                     comment = '#', low_memory = False)\n",
    "df_extension['read_depth'] = df_extension['t_alt_count'] + df_extension['t_ref_count']\n",
    "\n",
    "# Subset both sets for our gene list\n",
    "input_selected_genes = 'storage/reference/ASCC_genes.txt'\n",
    "df_genes = pd.read_csv(input_selected_genes, sep = '\\t')\n",
    "\n",
    "variants_pilot = df_pilot[df_pilot['Hugo_Symbol'].isin(df_genes['gene'])]\n",
    "variants_pilot.index = range(0, len(variants_pilot))\n",
    "\n",
    "variants_extension = df_extension[df_extension['Hugo_Symbol'].isin(df_genes['gene'])]\n",
    "variants_extension.index = range(0, len(variants_extension))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## ========================================\n",
    "# Subset for primary and recurrent samples\n",
    "## ========================================\n",
    "# R = Response (No Recurrence), NR = Non Response (Recurrence) \n",
    "\n",
    "# We separate out sample names by tumor type for each cohort\n",
    "ids_pilot_R = populate_samples_by_type(info_pilot, 'Response')\n",
    "ids_pilot_NR = populate_samples_by_type(info_pilot, 'No Response')\n",
    "\n",
    "ids_extension_R = populate_samples_by_type(info_extension, 'Response')\n",
    "ids_extension_NR = populate_samples_by_type(info_extension, 'No Response')\n",
    "\n",
    "pilot_primary_R_samples = ids_pilot_R.ix['Primary', :].tolist()\n",
    "pilot_primary_NR_samples = ids_pilot_NR.ix['Primary', :].tolist()\n",
    "\n",
    "ext_primary_R_samples = ids_extension_R.ix['Primary', :].tolist()\n",
    "ext_primary_NR_samples = ids_extension_NR.ix['Primary', :].tolist()\n",
    "\n",
    "pilot_recurrent_R_samples = ids_pilot_R.ix['Recurrent', :].tolist()\n",
    "pilot_recurrent_NR_samples = ids_pilot_NR.ix['Recurrent', :].tolist()\n",
    "pilot_recurrent_samples = pilot_recurrent_R_samples + pilot_recurrent_NR_samples\n",
    "\n",
    "ext_recurrent_R_samples = ids_extension_R.ix['Recurrent', :].tolist()\n",
    "ext_recurrent_NR_samples = sorted(ids_extension_NR.ix['Recurrent', :].tolist() + \\\n",
    "    ids_extension_NR.ix['Metastasis', :].tolist())\n",
    "\n",
    "# In order to keep only one recurrent sample per individual, we remove two\n",
    "ext_recurrent_NR_samples.remove('AC-V08-Tumor-SM-9J461') # Drop Metastasis sample because recurrent exists\n",
    "ext_recurrent_NR_samples.remove('AC-V18-Tumor-SM-9J46K') # Drop Metastasis sample because recurrent exists\n",
    "\n",
    "ext_recurrent_samples = ext_recurrent_R_samples + ext_recurrent_NR_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ========================================\n",
    "# Populate dataframes\n",
    "## ========================================\n",
    "\n",
    "# We populate our dataframes based on our list of mutations for each cohort\n",
    "df_pilot_primary_R = populate(pilot_primary_R_samples, df_genes, variants_pilot)\n",
    "df_pilot_primary_NR = populate(pilot_primary_NR_samples, df_genes, variants_pilot)\n",
    "\n",
    "df_ext_primary_R = populate(ext_primary_R_samples, df_genes, variants_extension)\n",
    "df_ext_primary_NR = populate(ext_primary_NR_samples, df_genes, variants_extension)\n",
    "\n",
    "df_pilot_recurrent = populate(pilot_recurrent_samples, df_genes, variants_pilot)\n",
    "df_ext_recurrent = populate(ext_recurrent_samples, df_genes, variants_extension)\n",
    "\n",
    "## ========================================================\n",
    "# Reorder samples and genes by frequency\n",
    "## ========================================================\n",
    "\n",
    "# We manipulate the order of our dataframes vertically to show the most\n",
    "# commonly mutated genes on top and horzitonally to group samples that are \n",
    "# mutated together for aesthics, the latter will only be applied to the\n",
    "# pilot cohort because we will be aligning this figure by individual\n",
    "\n",
    "## Primary\n",
    "df_freq_pilot_primary_R = df_pilot_primary_R.replace(0, np.nan)\n",
    "df_freq_pilot_primary_NR = df_pilot_primary_NR.replace(0, np.nan)\n",
    "\n",
    "df_freq_ext_primary_R = df_ext_primary_R.replace(0, np.nan)\n",
    "df_freq_ext_primary_NR = df_ext_primary_NR.replace(0, np.nan)\n",
    "\n",
    "freq_primary = (df_freq_pilot_primary_R.count(axis = 1, numeric_only = True) + \\\n",
    "    df_freq_pilot_primary_NR.count(axis = 1, numeric_only = True) + \\\n",
    "    df_freq_ext_primary_R.count(axis = 1, numeric_only = True) + \\\n",
    "    df_freq_ext_primary_NR.count(axis = 1, numeric_only = True))\n",
    "\n",
    "freq_primary.sort_values(ascending = False, inplace = True)\n",
    "\n",
    "## Recurrent\n",
    "df_freq_pilot_recurrent = df_pilot_recurrent.replace(0, np.nan)\n",
    "df_freq_ext_recurrent = df_ext_recurrent.replace(0, np.nan)\n",
    "\n",
    "freq_recurrent = (df_freq_pilot_recurrent.count(axis = 1, numeric_only = True) + \\\n",
    "                 df_freq_ext_recurrent.count(axis = 1, numeric_only = True))\n",
    "\n",
    "freq_recurrent.sort_values(ascending = False, inplace = True)\n",
    "\n",
    "length_genes = len(df_genes)\n",
    "\n",
    "# We sort our dataframes across genes depending on frequency across all cohorts\n",
    "df_pilot_primary_R_sorted = format_freq_sort(df_pilot_primary_R, freq_primary, length_genes)\n",
    "df_pilot_primary_NR_sorted = format_freq_sort(df_pilot_primary_NR, freq_primary, length_genes)\n",
    "\n",
    "df_ext_primary_R_sorted = format_freq_sort(df_ext_primary_R, freq_primary, length_genes)\n",
    "df_ext_primary_NR_sorted = format_freq_sort(df_ext_primary_NR, freq_primary, length_genes)\n",
    "\n",
    "df_pilot_recurrent_sorted = format_freq_sort(df_freq_pilot_recurrent, freq_recurrent, length_genes)\n",
    "df_ext_recurrent_sorted = format_freq_sort(df_freq_ext_recurrent, freq_recurrent, length_genes)\n",
    "\n",
    "# And also sort the samples across each subcohort for aesthic purposes\n",
    "df_pilot_primary_R_sorted = df_pilot_primary_R_sorted.T.sort_values(by = range(length_genes), ascending = [0] * length_genes).T\n",
    "df_pilot_primary_NR_sorted = df_pilot_primary_NR_sorted.T.sort_values(by = range(length_genes), ascending = [0] * length_genes).T\n",
    "\n",
    "df_ext_primary_R_sorted = df_ext_primary_R_sorted.T.sort_values(by = range(length_genes), ascending = [0] * length_genes).T\n",
    "df_ext_primary_NR_sorted = df_ext_primary_NR_sorted.T.sort_values(by = range(length_genes), ascending = [0] * length_genes).T\n",
    "\n",
    "# Note: These six dataframes will be the start for supplementary figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Concatenate\n",
    "data_primary_pilot = pd.concat([df_pilot_primary_R_sorted.ix[:,1:], df_pilot_primary_NR_sorted.ix[:,1:]], axis =1)\n",
    "data_primary_ext = pd.concat([df_ext_primary_R_sorted.ix[:,1:], df_ext_primary_NR_sorted.ix[:,1:]], axis =1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Sort recurrent dataframes by to align with individuals in pilot cohort\n",
    "# For better or for worse, this was done manually... \n",
    "\n",
    "## Pilot Cohort's Recurrent Samples\n",
    "data_pilot_recurrent = pd.DataFrame(index = range(0, length_genes), columns = range(0, len(data_primary_pilot.columns)))\n",
    "\n",
    "data_pilot_recurrent[7] = df_pilot_recurrent_sorted['AC-P10-Tumor-SM-63B55'].fillna(0)\n",
    "data_pilot_recurrent[9] = df_pilot_recurrent_sorted['AC-P09-Tumor-SM-63B52'].fillna(0)\n",
    "data_pilot_recurrent[10] = df_pilot_recurrent_sorted['AC-P12-Tumor-SM-63B59'].fillna(0)\n",
    "data_pilot_recurrent[11] = df_pilot_recurrent_sorted['AC-P15-Tumor-SM-63B5F'].fillna(0)\n",
    "data_pilot_recurrent[12] = df_pilot_recurrent_sorted['AC-P04-Tumor-SM-63B4Q'].fillna(0)\n",
    "\n",
    "data_pilot_recurrent.columns = [0, 1, 2, 3, 4, 5, 6, 'AC-P10-Tumor-SM-63B55', 8, 'AC-P09-Tumor-SM-63B52', \n",
    "                             'AC-P12-Tumor-SM-63B59', 'AC-P15-Tumor-SM-63B5F', 'AC-P04-Tumor-SM-63B4Q']\n",
    "data_pilot_recurrent_filled = data_pilot_recurrent.fillna(10)\n",
    "\n",
    "## Extension Cohort's Recurrent Samples \n",
    "data_ext_recurrent = pd.DataFrame(index = range(0, length_genes), columns = range(0, len(data_primary_ext.columns)))\n",
    "\n",
    "data_ext_recurrent[7] = df_ext_recurrent_sorted['AC-V15-Tumor-SM-9J46F'].fillna(0)\n",
    "data_ext_recurrent[8] = df_ext_recurrent_sorted['AC-V14-Tumor-SM-9J46D'].fillna(0)\n",
    "data_ext_recurrent[10] = df_ext_recurrent_sorted['AC-V08-Tumor-SM-9J462'].fillna(0)\n",
    "data_ext_recurrent[11] = df_ext_recurrent_sorted['AC-V09-Tumor-SM-9J464'].fillna(0)\n",
    "data_ext_recurrent[12] = df_ext_recurrent_sorted['AC-V16-Tumor-SM-9J46H'].fillna(0)\n",
    "data_ext_recurrent[13] = df_ext_recurrent_sorted['AC-V12-Tumor-SM-9J46A'].fillna(0)\n",
    "data_ext_recurrent[14] = df_ext_recurrent_sorted['AC-V10-Tumor-SM-9J466'].fillna(0)\n",
    "data_ext_recurrent[15] = df_ext_recurrent_sorted['AC-V11-Tumor-SM-9J468'].fillna(0)\n",
    "\n",
    "data_ext_recurrent.columns = [0, 1, 2, 3, 4, 5, 6, 7, 'AC-V15-Tumor-SM-9J46F', 'AC-V14-Tumor-SM-9J46D', 9, \n",
    "                              'AC-V08-Tumor-SM-9J462', 'AC-V09-Tumor-SM-9J464', 'AC-V16-Tumor-SM-9J46H', \n",
    "                              'AC-V12-Tumor-SM-9J46A', 'AC-V10-Tumor-SM-9J466', 'AC-V11-Tumor-SM-9J468', \n",
    "                              17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33]\n",
    "\n",
    "recurrent_remaining = list(set(df_ext_recurrent_sorted.ix[:,1:].columns.tolist()) - set(data_ext_recurrent.columns.tolist()))\n",
    "recurrent_remaining.pop(0) # to remove nan\n",
    "\n",
    "df_recurrent_remaining = df_ext_recurrent_sorted.ix[:, recurrent_remaining].fillna(0)\n",
    "df_recurrent_remaining_sorted = df_recurrent_remaining.T.sort_values(by = range(length_genes), ascending = [0] * length_genes).T\n",
    "\n",
    "for i in range(18,34):\n",
    "    j = i - 18\n",
    "    #print i, j, df_recurrent_remaining_sorted.columns.tolist()[j]\n",
    "    data_ext_recurrent.ix[:,i] = df_recurrent_remaining_sorted.ix[:,j]\n",
    "\n",
    "data_ext_recurrent_filled = data_ext_recurrent.fillna(10)\n",
    "\n",
    "data_ext_recurrent_filled.columns = [0, 1, 2, 3, 4, 5, 6, 'AC-V15-Tumor-SM-9J46F', 'AC-V14-Tumor-SM-9J46D', 9, \n",
    "                              'AC-V08-Tumor-SM-9J462', 'AC-V09-Tumor-SM-9J464', 'AC-V16-Tumor-SM-9J46H', \n",
    "                              'AC-V12-Tumor-SM-9J46A', 'AC-V10-Tumor-SM-9J466', 'AC-V11-Tumor-SM-9J468', \n",
    "                               17, 18, 'AC-V25-Tumor-SM-9J46R', 'AC-V28-Tumor-SM-9J46U', 'AC-V29-Tumor-SM-9J46V', \n",
    "                              'AC-V18-Tumor-SM-9J46J', 'AC-V21-Tumor-SM-9J46N', 'AC-V32-Tumor-SM-9J46Y', \n",
    "                              'AC-V23-Tumor-SM-9J46P', 'AC-V20-Tumor-SM-9J46M', 'AC-V31-Tumor-SM-9J46X', \n",
    "                              'AC-V17-Tumor-SM-9J46I', 'AC-V30-Tumor-SM-9J46W', 'AC-V26-Tumor-SM-9J46S', \n",
    "                              'AC-V19-Tumor-SM-9J46L', 'AC-V24-Tumor-SM-9J46Q', 'AC-V22-Tumor-SM-9J46O', \n",
    "                              'AC-V27-Tumor-SM-9J46T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# How many genes would you like to display\n",
    "display_primary = 22\n",
    "display_recurrent = 21\n",
    "\n",
    "# Fill any NA columns with 10s\n",
    "\n",
    "data_primary_pilot_filled = fillna_columns(data_primary_pilot)\n",
    "data_primary_ext_filled = fillna_columns(data_primary_ext)\n",
    "\n",
    "data_00 = data_primary_pilot_filled.iloc[:display_primary, :]\n",
    "data_01 = data_primary_ext_filled.iloc[:display_primary, :]\n",
    "\n",
    "data_10 = data_pilot_recurrent_filled.iloc[:display_recurrent, :]\n",
    "data_11 = data_ext_recurrent_filled.iloc[:display_recurrent, :]\n",
    "\n",
    "genes_primary = df_pilot_primary_R_sorted.ix[:,'gene']\n",
    "genes_recurrent = df_pilot_recurrent_sorted.ix[:,'gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ========================================================\n",
    "# Clinical annotation 1: Primary Samples - Recurrence status\n",
    "## ========================================================\n",
    "# Primary samples, pilot cohort\n",
    "resp = np.zeros(len(df_pilot_primary_R.columns[1:]))    # no recurrence\n",
    "noresp = np.ones(len(df_pilot_primary_NR.columns[1:])) # recurrence\n",
    "pilot_primary_outcome = np.concatenate((resp, noresp), axis = 0)\n",
    "\n",
    "pilot_primary_outcome = pd.DataFrame(pilot_primary_outcome).T\n",
    "pilot_primary_outcome.columns = [data_00.columns.tolist()]\n",
    "pilot_primary_outcome = pilot_primary_outcome.rename(index={0: 'Cohort Status'})\n",
    "\n",
    "# Primary samples, extension cohort\n",
    "resp = np.zeros(len(df_ext_primary_R.columns[1:]))   # no recurrence\n",
    "noresp = np.ones(len(df_ext_primary_NR.columns[1:])) # recurrence\n",
    "ext_primary_outcome = np.concatenate((resp, noresp), axis = 0)\n",
    "\n",
    "ext_primary_outcome = pd.DataFrame(ext_primary_outcome).T\n",
    "ext_primary_outcome.columns = [data_01.columns.tolist()]\n",
    "ext_primary_outcome = ext_primary_outcome.rename(index={0: 'Cohort Status'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ========================================================\n",
    "# Clinical annotation 1: Primary Samples - Recurrence status\n",
    "## ========================================================\n",
    "pilot_resp = format_resp(info_pilot, data_00)\n",
    "ext_resp = format_resp(info_extension, data_01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ==============================================================\n",
    "# Clinical annotation 1: Recurrent Samples - Recurrence location\n",
    "## ==============================================================\n",
    "# Recurrent samples, pilot cohort\n",
    "pilot_loc = format_loc(info_pilot, data_10)\n",
    "ext_loc = format_loc(info_extension, data_11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===============================================\n",
    "# Clinical annotation 2: All Samples - HPV Status\n",
    "## ===============================================\n",
    "pilot_primary_hpv = format_hpv(info_pilot, data_00)\n",
    "ext_primary_hpv = format_hpv(info_extension, data_01)\n",
    "\n",
    "pilot_recurrent_hpv = format_hpv(info_pilot, data_10)\n",
    "ext_recurrent_hpv = format_hpv(info_extension, data_11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===============================================\n",
    "# Genomic annotation 1: All Samples - 3q CNV\n",
    "## ===============================================\n",
    "# We extract information on 3q from each sample\n",
    "# For convenience and cleanliness of code, we know from inspecting the results that no samples have a 'high gain'\n",
    "# as judged by GISTIC for each sample (/storage/gistic/*/sample_cutoffs.txt). We will report all reported positive\n",
    "# events as Gains and negative events as Losses.\n",
    "\n",
    "gistic_ice = pd.read_csv('storage/gistic/ice/broad_values_by_arm.txt', sep = '\\t')\n",
    "gistic_agilent = pd.read_csv('storage/gistic/agilent/broad_values_by_arm.txt', sep = '\\t')\n",
    "gistic_pd1 = gistic_ice.ix[:, 'AC-DFCI_AC_PD1-1-Tumor-SM-9LRI9'].to_frame()\n",
    "\n",
    "gistic_pilot = pd.concat([gistic_agilent, gistic_pd1], axis = 1)\n",
    "gistic_ext = gistic_ice.copy(deep = True)\n",
    "gistic_ext.drop(['AC-DFCI_AC_PD1-1-Tumor-SM-9LRI9'], axis = 1, inplace = True)\n",
    "\n",
    "gistic_pilot_3q = gistic_pilot.ix[5, :]\n",
    "gistic_ext_3q = gistic_ext.ix[5, :]\n",
    "\n",
    "pilot_primary_3q = format_gistic_3q(gistic_pilot_3q, data_00)\n",
    "ext_primary_3q = format_gistic_3q(gistic_ext_3q, data_01)\n",
    "\n",
    "pilot_recurrent_3q = format_gistic_3q(gistic_pilot_3q, data_10)\n",
    "ext_recurrent_3q = format_gistic_3q(gistic_ext_3q, data_11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ==================================================\n",
    "# Genomic annotation 2: All Samples - Mutations / MB\n",
    "## ==================================================\n",
    "\n",
    "# We calculate the mutations / Mb for all samples illustrated\n",
    "# For the pilot cohort, MutSig provides mutations / Mb of both total and nonsyn mutations\n",
    "# For the extension cohort, we calculate an estimate of nonsyn mutational burden\n",
    "\n",
    "# Pilot Cohort\n",
    "mutsig_rates = pd.read_csv('storage/mutsig-pilot/patient_counts_and_rates.txt', sep = '\\t')\n",
    "\n",
    "pilot_primary_load = format_pilot_load(mutsig_rates, data_00)\n",
    "pilot_recurrent_load = format_pilot_load(mutsig_rates, data_10)\n",
    "\n",
    "# Extension Cohort - Estimate of Mutations / Mb\n",
    "extension_all_calls = pd.read_csv('filter_unmatched/all_genes/df_filter_pass.txt', sep = '\\t', low_memory = False)\n",
    "\n",
    "ext_primary_load = format_extension_load(info_extension, extension_all_calls, data_01)\n",
    "ext_recurrent_load = format_extension_load(info_extension, extension_all_calls, data_11)\n",
    "\n",
    "pilot_primary_load.to_csv('pilot_primary_load.txt', sep = '\\t')\n",
    "pilot_recurrent_load.to_csv('pilot_recurrent_load.txt', sep = '\\t')\n",
    "ext_primary_load.to_csv('extension_primary_load.txt', sep = '\\t')\n",
    "ext_recurrent_load.to_csv('extension_recurrent_load.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "nsamples_primary_pilot = 16\n",
    "nsamples_primary_ext = 34\n",
    "\n",
    "## ===============================================\n",
    "# Figure 1: CoMut of Primary / Recurrent Tumors\n",
    "## ===============================================\n",
    "\n",
    "fig = plt.figure(figsize = (27,16)) \n",
    "gs = gridspec.GridSpec(11, 3, \n",
    "                       height_ratios = [2, 2, 2*display_primary, 2, display_primary/2, 1, 2, 2, 2*display_primary, 2, display_primary/2],\n",
    "                       width_ratios = [nsamples_primary_pilot, nsamples_primary_ext, nsamples_primary_pilot/4])\n",
    "gs.update(wspace = 0.08, hspace = 0.04)\n",
    "\n",
    "#### ==================================== ####\n",
    "####           Primary Samples            ####\n",
    "#### ==================================== ####\n",
    "\n",
    "##============================================\n",
    "# ax020 - CoMut - Pilot, Primary Samples\n",
    "##============================================\n",
    "ax020, cax20 = plot_comut(2, 0, data_00, genes_primary, 1, 'italic', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "plt.ylabel('Primary Tumor Samples', labelpad = 35, fontsize = 24)\n",
    "\n",
    "##============================================\n",
    "# ax021 - CoMut - No Response Primary Samples\n",
    "##============================================\n",
    "ax021, cax021 = plot_comut(2, 1, data_01, genes_primary, 0, 'normal', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "##============================================\n",
    "# ax000 - Recur Status - Response Primary Samples\n",
    "##============================================\n",
    "ax000, cax000 = plot_comut(0, 0, pilot_resp, pilot_resp.index.tolist(), 1, 'normal', \n",
    "                         cmapResp, normResp, 'black', sharex = ax020)\n",
    "\n",
    "plt.title('Pilot Cohort', y = 1.15, fontsize = 24)\n",
    "\n",
    "##============================================\n",
    "# ax001 - Recur Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax001, cax001 = plot_comut(0, 1, ext_resp, ext_resp.index.tolist(), 0, 'normal', \n",
    "                         cmapResp, normResp, 'black', sharex = ax021, sharey = ax000)\n",
    "\n",
    "plt.title('Extension Cohort', y = 1.15, fontsize = 24)\n",
    "\n",
    "##============================================\n",
    "# ax010 - Hpv Status - Response Primary Samples\n",
    "##============================================\n",
    "ax010, cax010 = plot_comut(1, 0, pilot_primary_hpv, pilot_primary_hpv.index.tolist(), 1, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax020)\n",
    "\n",
    "##============================================\n",
    "# ax011 - Hpv Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax011, cax011 = plot_comut(1, 1, ext_primary_hpv, ext_primary_hpv.index.tolist(), 0, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax021, sharey = ax010)\n",
    "\n",
    "##============================================\n",
    "# ax030 - 3q Status - Response Primary Samples\n",
    "##============================================\n",
    "ax030, cax030 = plot_comut(3, 0, pilot_primary_3q, pilot_primary_3q.index.tolist(), 1, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax020)\n",
    "\n",
    "##============================================\n",
    "# ax031 - 3q Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax031, cax031 = plot_comut(3, 1, ext_primary_3q, ext_primary_3q.index.tolist(), 0, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax021, sharey = ax030)\n",
    "\n",
    "##==============================================\n",
    "# ax040 - Mutations / Mb - Response Primary Samples\n",
    "##==============================================\n",
    "ax040 = plot_burden(4, 0, pilot_primary_load, 1, colors_bar, 10, 39, \n",
    "                                      display_label = 'Mutations / Mb', sharex = ax020, label_bottom = 'off')\n",
    "\n",
    "##==============================================\n",
    "# ax041 - Mutations / Mb - No Response Primary Samples\n",
    "##==============================================\n",
    "ax041 = plot_burden(4, 1, ext_primary_load, 0, colors_bar, 10, 39,\n",
    "                                     sharex = ax021, sharey = ax040, label_bottom = 'off')\n",
    "\n",
    "#### ==================================== ####\n",
    "####         Recurrent Samples            ####\n",
    "#### ==================================== ####\n",
    "\n",
    "##============================================\n",
    "# ax120 - CoMut - Pilot, Recurrent Samples\n",
    "##============================================\n",
    "ax120, cax120 = plot_comut(8, 0, data_10, genes_recurrent, 1, 'italic', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "plt.ylabel('Recurrent Tumor Samples', labelpad = 35, fontsize = 24)\n",
    "\n",
    "##============================================\n",
    "# ax121 - CoMut - No Response Recurrent Samples\n",
    "##============================================\n",
    "ax121, cax121 = plot_comut(8, 1, data_11, genes_recurrent, 0, 'normal', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "##============================================\n",
    "# ax100 - Recurrence Location - Response Primary Samples\n",
    "##============================================\n",
    "ax100, cax100 = plot_comut(6, 0, pilot_loc, pilot_loc.index.tolist(), 1, 'normal', \n",
    "                         cmapDistance, normDistance, 'black', sharex = ax120)\n",
    "\n",
    "##============================================\n",
    "# ax101 - Recurrence Location - No Response Primary Samples\n",
    "##============================================\n",
    "ax101, cax101 = plot_comut(6, 1, ext_loc, ext_loc.index.tolist(), 0, 'normal', \n",
    "                         cmapDistance, normDistance, 'black', sharex = ax121, sharey = ax100)\n",
    "\n",
    "\n",
    "##============================================\n",
    "# ax110 - Hpv Status - Response Primary Samples\n",
    "##============================================\n",
    "ax110, cax110 = plot_comut(7, 0, pilot_recurrent_hpv, pilot_recurrent_hpv.index.tolist(), 1, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax120)\n",
    "\n",
    "##============================================\n",
    "# ax111 - Hpv Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax111, cax111 = plot_comut(7, 1, ext_recurrent_hpv, ext_recurrent_hpv.index.tolist(), 0, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax121, sharey = ax110)\n",
    "\n",
    "##============================================\n",
    "# ax130 - 3q Status - Response Primary Samples\n",
    "##============================================\n",
    "ax130, cax130 = plot_comut(9, 0, pilot_recurrent_3q, pilot_recurrent_3q.index.tolist(), 1, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax120)\n",
    "\n",
    "##============================================\n",
    "# ax131 - 3q Status - No Response Recurrent Samples\n",
    "##============================================\n",
    "ax131, cax131 = plot_comut(9, 1, ext_recurrent_3q, ext_recurrent_3q.index.tolist(), 0, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax121, sharey = ax130)\n",
    "\n",
    "##==============================================\n",
    "# ax140 - Mutations / Mb - Response Primary Samples\n",
    "##==============================================\n",
    "\n",
    "ax140 = plot_burden(10, 0, pilot_recurrent_load, 1, colors_bar, 10, 39, \n",
    "                    display_label = 'Mutations / Mb', sharex = ax120, label_bottom = 'on')\n",
    "\n",
    "samples_pilot_display = ['AC-P06','AC-P11','AC-P02','AC-P08','AC-P07','AC-P14','AC-P05','AC-P10',\n",
    " 'AC_PD1-1','AC-P09','AC-P12','AC-P15','AC-P04']\n",
    "\n",
    "ax140.set_xticklabels(samples_pilot_display, rotation = 45, ha = 'right')\n",
    "\n",
    "\n",
    "##==============================================\n",
    "# ax141 - Mutations / Mb - No Response Primary Samples\n",
    "##==============================================\n",
    "ax141 = plot_burden(10, 1, ext_recurrent_load, 0, colors_bar, 10, 39,\n",
    "                    sharex = ax021, sharey = ax140, label_bottom = 'on')\n",
    "\n",
    "samples_extension_display = ['AC-V05','AC-V04','AC-V06','AC-V03','AC-V07','AC-V01','AC-V02','AC-V15',\n",
    " 'AC-V14','AC-V34','AC-V08','AC-V09','AC-V16','AC-V12','AC-V10','AC-V11',\n",
    " 'AC-V13','AC-V33','AC-V25','AC-V28','AC-V29','AC-V18','AC-V21','AC-V32',\n",
    " 'AC-V23','AC-V20','AC-V31','AC-V17','AC-V30','AC-V26','AC-V19','AC-V24',\n",
    " 'AC-V22','AC-V27']\n",
    "ax141.set_xticklabels(samples_extension_display, rotation = 45, ha = 'right')\n",
    "\n",
    "\n",
    "#### ==================================== ####\n",
    "####               Legend                 ####\n",
    "#### ==================================== ####\n",
    "\n",
    "\n",
    "##============================================\n",
    "# ax022 - Legend - Mutation Type\n",
    "##============================================\n",
    "ax022 = plt.subplot(gs[2,2])\n",
    "\n",
    "# Remove plot frame lines\n",
    "ax022.spines['top'].set_visible(False)\n",
    "ax022.spines['bottom'].set_visible(False)\n",
    "ax022.spines['right'].set_visible(False)\n",
    "ax022.spines['left'].set_visible(False)\n",
    "\n",
    "plt.tick_params(axis = 'both', which = 'both', \n",
    "    bottom = 'off', top = 'off', left = 'off', right = 'off',\n",
    "    labelbottom = 'off', labelleft = 'off', labelright = 'off')\n",
    "\n",
    "labels = [\"Synonymous\", \"Missense\", \"Splice Site\", \"Nonsense\", \"Frameshift\", \"InDel\"]\n",
    "\n",
    "ax022.bar(0, 0, label = labels[0], color = colors_fig1_comut[1]) # Syn\n",
    "ax022.bar(1, 0, label = labels[1], color = colors_fig1_comut[2]) # Missense\n",
    "ax022.bar(2, 0, label = labels[2], color = colors_fig1_comut[3]) # Splice Site\n",
    "ax022.bar(4, 0, label = labels[3], color = colors_fig1_comut[4]) # Nonsense\n",
    "ax022.bar(4, 0, label = labels[4], color = colors_fig1_comut[5]) # Frameshift\n",
    "ax022.bar(4, 0, label = labels[5], color = colors_fig1_comut[6]) # Indel\n",
    "\n",
    "plt.ylim(0)\n",
    "\n",
    "legend = ax022.legend(bbox_to_anchor = (1.35, 0.75), ncol = 1, handlelength = 3, fontsize = 14, columnspacing = 1, \n",
    "               frameon = False)\n",
    "\n",
    "##============================================\n",
    "# ax002 - Legend - Cohort Status\n",
    "##============================================\n",
    "ax002 = plt.subplot(gs[0,2])\n",
    "\n",
    "# Remove plot frame lines\n",
    "ax002.spines['top'].set_visible(False)\n",
    "ax002.spines['bottom'].set_visible(False)\n",
    "ax002.spines['right'].set_visible(False)\n",
    "ax002.spines['left'].set_visible(False)\n",
    "\n",
    "plt.tick_params(axis = 'both', which = 'both', \n",
    "    bottom = 'off', top = 'off', left = 'off', right = 'off',\n",
    "    labelbottom = 'off', labelleft = 'off', labelright = 'off')\n",
    "\n",
    "labels = ['No Recurrence', 'Recurrence']\n",
    "\n",
    "ax002.bar(0, 0, label = labels[0], color = colors_tableauResp[1]) \n",
    "ax002.bar(1, 0, label = labels[1], color = colors_tableauResp[0]) \n",
    "\n",
    "plt.ylim(0)\n",
    "\n",
    "legend = ax002.legend(bbox_to_anchor = (2.69, 2.35), ncol = 2, handlelength = 3, fontsize = 14, columnspacing = 1, \n",
    "               frameon = False)\n",
    "\n",
    "##============================================\n",
    "# ax012 - Legend - HPV Status\n",
    "##============================================\n",
    "labels = [' Neg ', 'Positive', 'Unknown']\n",
    "ax012 = create_legend(1, 2, labels, colors_tableauHPV, 2.10, 1.75, 2)\n",
    "\n",
    "##============================================\n",
    "# ax032 - Legend - 3q CNV status\n",
    "##============================================\n",
    "labels = [\"Loss\", \"Gain\"]\n",
    "ax032 = plt.subplot(gs[3,2])\n",
    "\n",
    "# Remove plot frame lines\n",
    "ax032.spines['top'].set_visible(False)\n",
    "ax032.spines['bottom'].set_visible(False)\n",
    "ax032.spines['right'].set_visible(False)\n",
    "ax032.spines['left'].set_visible(False)\n",
    "\n",
    "plt.tick_params(axis = 'both', which = 'both', \n",
    "    bottom = 'off', top = 'off', left = 'off', right = 'off',\n",
    "    labelbottom = 'off', labelleft = 'off', labelright = 'off')\n",
    "\n",
    "ax032.bar(0, 0, label = labels[0], color = colors_tableau3q[1]) # Loss\n",
    "ax032.bar(1, 0, label = labels[1], color = colors_tableau3q[3]) # Gain\n",
    "\n",
    "plt.ylim(0)\n",
    "\n",
    "legend = ax032.legend(bbox_to_anchor = (1.60, 3.00), ncol = 2, handlelength = 3, fontsize = 14, columnspacing = 1, \n",
    "               frameon = False)\n",
    "\n",
    "##============================================\n",
    "# ax042 - Legend - Mutational Burden\n",
    "##============================================\n",
    "labels = [\"Nonsyn burden\", \"Syn burden\"]\n",
    "ax042 = create_legend(4, 2, labels, colors_bar, 1.46, 1.0, 1)\n",
    "\n",
    "##============================================\n",
    "# ax102 - Legend - Recurrence Location\n",
    "##============================================\n",
    "labels = ['Local', 'Distant']\n",
    "ax102 = create_legend(6, 2, labels, colors_fig1_distance, 1.8, 1.75, 2)\n",
    "\n",
    "plt.savefig('figures/figure1.png', bbox_inches = 'tight')\n",
    "#plt.show()\n",
    "\n",
    "# The extra lines and boxes were cleaned up with photoshop :) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figure2'></a>\n",
    "## Figure 2: Similarity bars\n",
    "<a href='#table.of.contents'>Return to top</a><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "In this section, we will generate similarity bars shown in Figure 2 for each primary and recurrent sample pair in the pilot cohort.\n",
    "\n",
    "These similarity bars were made by simply counting the overall number of unique and shared mutations between the two samples for each individual, and then illustrated with matplotlib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Specify which patients are interested in\n",
    "patients = ['AC-P04', 'AC-P09', 'AC-P10', 'AC-P12', 'AC-P15']\n",
    "\n",
    "# We define a function which will generate these figures for us using the snv and indel files\n",
    "# from each \n",
    "for patient in patients:\n",
    "    # Import snv and indel MAF files for our current patient\n",
    "    snvs = glob.glob('storage/mafs/pilot_snv/' + patient + '*')\n",
    "    indels = glob.glob('storage/mafs/pilot_indel/' + patient + '*')\n",
    "\n",
    "    primary_snv = pd.read_csv(snvs[0], sep = '\\t', comment = '#')\n",
    "    primary_indel = pd.read_csv(indels[0], sep = '\\t', comment = '#')\n",
    "    recurrent_snv = pd.read_csv(snvs[1], sep = '\\t', comment = '#')\n",
    "    recurrent_indel = pd.read_csv(indels[1], sep = '\\t', comment = '#')\n",
    "    \n",
    "    # Extract columns of interest\n",
    "    cols = ['Hugo_Symbol', 'Chromosome', 'Start_position',  'Reference_Allele', 'Tumor_Seq_Allele1', 'Tumor_Seq_Allele2','Tumor_Sample_Barcode']\n",
    "    primary = pd.concat([primary_snv.ix[:,cols], primary_indel.ix[:,cols]], ignore_index = True)\n",
    "    recurrent = pd.concat([recurrent_snv.ix[:,cols], recurrent_indel.ix[:,cols]], ignore_index = True)\n",
    "\n",
    "    # Next, we perform a comparison of specific mutations\n",
    "    # In all cases, \"1\" is the primary tumor and \"2\" is the recurrent tumor\n",
    "    id_1 = primary.ix[:,'Hugo_Symbol'] + primary.ix[:,'Chromosome'].astype(str) + primary.ix[:,'Start_position'].astype(str) + primary.ix[:,'Reference_Allele'] + \\\n",
    "        primary.ix[:,'Tumor_Seq_Allele1'] + primary.ix[:,'Tumor_Seq_Allele2']\n",
    "    id_2 = recurrent.ix[:,'Hugo_Symbol'] + recurrent.ix[:,'Chromosome'].astype(str) + recurrent.ix[:,'Start_position'].astype(str) + recurrent.ix[:,'Reference_Allele'] + \\\n",
    "        recurrent.ix[:,'Tumor_Seq_Allele1'] + recurrent.ix[:,'Tumor_Seq_Allele2']\n",
    "    \n",
    "    idx_1 = (id_1).isin(id_2)\n",
    "    idx_2 = (id_2).isin(id_1)\n",
    "\n",
    "    in_1 = len(primary[~idx_1])\n",
    "    in_both = len(primary[idx_1])\n",
    "    in_2 = len(recurrent[~idx_2])\n",
    "    total = len(primary) + len(recurrent) - in_both\n",
    "\n",
    "    # And calculate the fractions\n",
    "    frac_1 = in_1*100/total\n",
    "    frac_both = in_both*100/total\n",
    "    frac_2 = in_2*100/total\n",
    "    \n",
    "    width = 0.5\n",
    "\n",
    "    # And finally plot with matplotlib\n",
    "    fig = plt.figure(1, figsize = (2.5, 10))\n",
    "    ax = plt.subplot()\n",
    "    # We utilize the same colors as in the additional illustrations in this figure, these colors are defined\n",
    "    # at the beginning of the notebook. \n",
    "    primary = ax.bar(0, frac_1, width, color = colors_fig2[0])\n",
    "    both = ax.bar(0, frac_both, width, color = colors_fig2[1], bottom = frac_1)\n",
    "    recurrent = ax.bar(0, frac_2, width, color = colors_fig2[2], bottom = (frac_1 + frac_both))\n",
    "\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(True)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "\n",
    "    plt.tick_params(axis = 'both', which = 'both',\n",
    "                top = 'off', bottom = 'off', right = 'off', left = 'off', \n",
    "                labeltop = 'on', labelbottom = 'off', labelright = 'on', labelleft = 'off')\n",
    "\n",
    "    ind = np.arange(1)\n",
    "    ax.set_xticks(ind + width/2)\n",
    "    ax.set_xticklabels((patient,''), fontsize = 17)\n",
    "    ax.set_yticks((0.5*frac_1, frac_1+0.5*frac_both, frac_1+frac_both+0.5*frac_2))\n",
    "    if in_both != 0:\n",
    "        ax.set_yticklabels(('Primary only', 'Shared', 'Recurrent only'), fontsize = 16)\n",
    "    else:\n",
    "        ax.set_yticklabels(('Primary only', '', 'Recurrent only'), fontsize = 16) # This removes excess text if no mutations are shared\n",
    "\n",
    "    # We define a function to label the number of mutations for each section, as well as plot a title \n",
    "    # on the right hand side of each section.\n",
    "    def autolabel(rect1, rect2, rect3):\n",
    "        ax.text(rect1[0].get_x() + rect1[0].get_width()/2., 0.45*rect1[0].get_height(),\n",
    "            '%d' % int(in_1) + '(' + '%d' %int(rect1[0].get_height()) + '%)',\n",
    "            ha='center', va='bottom', fontsize = 18, color = 'white', weight = 'bold',\n",
    "            path_effects=[PathEffects.withStroke(linewidth=5,foreground=\"black\")]) \n",
    "        if in_both != 0:\n",
    "            ax.text(rect2[0].get_x() + rect2[0].get_width()/2., rect1[0].get_height() + 0.45*rect2[0].get_height(),\n",
    "                '%d' % int(in_both) + '(' + '%d' %int(rect2[0].get_height()) + '%)',\n",
    "                ha='center', va='bottom', fontsize = 18, color = 'white', weight = 'bold',\n",
    "                path_effects=[PathEffects.withStroke(linewidth=5,foreground=\"black\")]) \n",
    "        ax.text(rect3[0].get_x() + rect3[0].get_width()/2., rect1[0].get_height() + rect2[0].get_height() + 0.45*rect3[0].get_height(),\n",
    "            '%d' % int(in_2) + '(' + '%d' %int(rect3[0].get_height()) + '%)',\n",
    "            ha='center', va='bottom', fontsize = 18, color = 'white', weight = 'bold',\n",
    "            path_effects=[PathEffects.withStroke(linewidth=5,foreground=\"black\")]) \n",
    "        \n",
    "    autolabel(primary, both, recurrent)\n",
    "    plt.savefig('figures/figure2/' + patient + '.pdf', bbox_inches = 'tight')\n",
    "    plt.savefig('figures/figure2/' + patient + '.svg', bbox_inches = 'tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figure3'></a>\n",
    "## Figure 3: Mutational and Neoantigen burden\n",
    "<a href='#table.of.contents'>Return to top</a><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "In this section, we will produce Figure 3 by calculating both the mutational and neoantigen burden for each primary and recurrent pair in the pilot cohort. Results are then visualized with matplotlib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# First, import sample annotations\n",
    "df_info = pd.read_csv('storage/reference/ASCC_samples-pilot.txt', sep = '\\t', index_col = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing and formatting neoantigen calls...\n",
      "...loading combined binders...\n",
      "...subsetting for calls with < 500 nM binding affinity\n"
     ]
    }
   ],
   "source": [
    "## =================================\n",
    "# Import neoantigen data and format\n",
    "## =================================\n",
    "print 'Importing and formatting neoantigen calls...'\n",
    "\n",
    "# We import our files\n",
    "input_path = 'storage/neoantigen_binders/'\n",
    "input_suffix = '.combined.binders.txt'\n",
    "\n",
    "allFiles = glob.glob(input_path + '*' + input_suffix)\n",
    "df_neoantigen = pd.DataFrame()\n",
    "list_ = []\n",
    "\n",
    "print '...loading combined binders...'\n",
    "for file_ in allFiles:\n",
    "    df_ = pd.read_csv(file_, index_col = None, sep = '\\t', comment = '#')\n",
    "    list_.append(df_)\n",
    "df_neoantigen = pd.concat(list_, ignore_index = True)\n",
    "\n",
    "# The binding affinity is sometimes outputted as a percentage. \n",
    "index = df_neoantigen[df_neoantigen['affinity_mut'].str.contains('rank')==True].index.tolist()\n",
    "for i in index:\n",
    "    percent = df_neoantigen.ix[i, 'affinity_mut'].split(' (%rank)')[0]\n",
    "    percent = float(percent)\n",
    "    affinity = percent*500\n",
    "    df_neoantigen.ix[i, 'affinity_mut'] = affinity\n",
    "\n",
    "print '...subsetting for calls with < 500 nM binding affinity'\n",
    "df_neoantigen_500 = df_neoantigen[df_neoantigen['affinity_mut'].astype(float) < 500]\n",
    "\n",
    "# We subset our binders into primary and recurrent samples\n",
    "primary_samples = df_info[df_info['Tumor Type'] == 'Primary'].ix[:, 'Tumor Sample ID'].unique().tolist()\n",
    "recurrent_samples = df_info[df_info['Tumor Type'] == 'Recurrent'].ix[:, 'Tumor Sample ID'].unique().tolist()\n",
    "\n",
    "df_neoantigen_p = df_neoantigen_500[df_neoantigen_500['sample'].isin(primary_samples)]\n",
    "df_neoantigen_r = df_neoantigen_500[df_neoantigen_500['sample'].isin(recurrent_samples)]\n",
    "\n",
    "# Place mutational burden of affinity < 500\n",
    "neoantigen_p = pd.DataFrame([df_neoantigen_p['sample'].value_counts().tolist()], index = ['< 500 load'],\n",
    "                    columns = df_neoantigen_p['sample'].value_counts().index.tolist())\n",
    "neoantigen_r = pd.DataFrame([df_neoantigen_r['sample'].value_counts().tolist()], index = ['< 500 load'],\n",
    "                    columns = df_neoantigen_r['sample'].value_counts().index.tolist())\n",
    "neoantigen_p = neoantigen_p.T\n",
    "neoantigen_r = neoantigen_r.T\n",
    "\n",
    "# We sort our dataframes by index so they will be in alphabetical order and create numerical arrays\n",
    "# that we will be using to plot. Since there is the extra \n",
    "neoantigen_p = neoantigen_p.sort_index()\n",
    "neoantigen_p = neoantigen_p[neoantigen_p.index != 'AC-P05-Tumor-SM-63B4R'] # Just a quick hack to get rid of that sample\n",
    "neoantigen_r = neoantigen_r.sort_index()\n",
    "\n",
    "pre_neo = neoantigen_p.ix[:, '< 500 load']\n",
    "post_neo_ = neoantigen_r.ix[:, '< 500 load']\n",
    "post_neo = pd.Series([0]).append(post_neo_) # Likewise, we do not have a post sample for PD1 sample, so just set to zero.\n",
    "\n",
    "pre_neo.index = range(0, len(pre_neo))\n",
    "post_neo.index = range(0, len(post_neo))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing mutational burden for extension cohort from MutSig results...\n",
      "...calculating for primary-recurrent pairs\n",
      "...calculating for only primary samples\n"
     ]
    }
   ],
   "source": [
    "## ==========================================================\n",
    "# Import mutational burden data and format - Pilot Cohort\n",
    "## ==========================================================\n",
    "print 'Importing mutational burden for extension cohort from MutSig results...'\n",
    "\n",
    "mutsig_burden = pd.read_csv('storage/mutsig-pilot/patient_counts_and_rates.txt', sep = '\\t')\n",
    "\n",
    "recurrent_samples = df_info[df_info.ix[:,'Tumor Type'] == 'Recurrent'].ix[:,'Tumor Sample ID'].tolist()\n",
    "recurrent_individual = df_info[df_info.ix[:,'Tumor Type'] == 'Recurrent'].ix[:,'Individual ID'].tolist()\n",
    "\n",
    "primary_samples_rec = df_info.ix[df_info[df_info.ix[:,'Individual ID'].isin(recurrent_individual)].ix[:,'Tumor Type'].index.tolist(),:]\n",
    "primary_samples_rec = primary_samples_rec[primary_samples_rec.ix[:,'Tumor Type'] == 'Primary'].ix[:,'Tumor Sample ID'].tolist()\n",
    "\n",
    "print '...calculating for primary-recurrent pairs'\n",
    "\n",
    "df_burden = pd.DataFrame([], index = range(0, 5), columns = ['Primary', 'Recurrent', 'Primary load', 'Recurrent load'])\n",
    "df_burden.ix[:,'Primary'] = primary_samples_rec\n",
    "df_burden.ix[:,'Recurrent'] = recurrent_samples\n",
    "\n",
    "for i in df_burden.ix[:,'Primary'].tolist():\n",
    "    index = mutsig_burden[mutsig_burden.ix[:,'name'] == i].index.tolist()[0]\n",
    "    index_ = df_burden[df_burden.ix[:, 'Primary'] == i].index.tolist()[0]\n",
    "    df_burden.ix[index_, 'Primary load'] = mutsig_burden.ix[index, 'rate_tot']*10*10*10*10*10*10 # Convert from Mutations per base to Mb\n",
    "\n",
    "for i in df_burden.ix[:,'Recurrent'].tolist():\n",
    "    index = mutsig_burden[mutsig_burden.ix[:,'name'] == i].index.tolist()[0]\n",
    "    index_ = df_burden[df_burden.ix[:, 'Recurrent'] == i].index.tolist()[0]\n",
    "    df_burden.ix[index_, 'Recurrent load'] = mutsig_burden.ix[index, 'rate_tot']*10*10*10*10*10*10 # Convert from Mutations per base to Mb\n",
    "\n",
    "pre_mut_ = df_burden.ix[:, 'Primary load']\n",
    "post_mut_ = df_burden.ix[:, 'Recurrent load']\n",
    "\n",
    "pd1_mut = mutsig_burden.ix[0, 'rate_tot']*10*10*10*10*10*10 # Convert from Mutations per base to Mb\n",
    "pre_mut = pd.Series(pd1_mut).append(pre_mut_)\n",
    "pre_mut.index = range(0, len(pre_mut))\n",
    "\n",
    "post_mut = pd.Series([0]).append(post_mut_)\n",
    "post_mut.index = range(0, len(post_mut))\n",
    "\n",
    "print '...calculating for only primary samples'    \n",
    "\n",
    "primary_samples_all = df_info[df_info.ix[:,'Tumor Type'] == 'Primary'].ix[:,'Tumor Sample ID'].tolist()\n",
    "df_burden_pre = pd.DataFrame([], index = range(0, len(primary_samples_all)), columns = ['Primary', 'Primary load'])\n",
    "df_burden_pre.ix[:,'Primary'] = primary_samples_all\n",
    "\n",
    "for i in df_burden_pre.ix[:,'Primary'].tolist():\n",
    "    index = mutsig_burden[mutsig_burden.ix[:,'name'] == i].index.tolist()[0]\n",
    "    index_ = df_burden_pre[df_burden_pre.ix[:,'Primary'] == i].index.tolist()[0]\n",
    "    df_burden_pre.ix[index_, 'Primary load'] = mutsig_burden.ix[index, 'rate_tot']*10*10*10*10*10*10\n",
    "\n",
    "pre_mut_all = df_burden_pre.ix[:,'Primary load']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing details necessary to calculate mutational burden for extension cohort...\n",
      "...calculating for primary-recurrent pairs\n"
     ]
    }
   ],
   "source": [
    "## ===========================================================\n",
    "# Import mutational burden data and format - Extension Cohort\n",
    "## ===========================================================\n",
    "print 'Importing details necessary to calculate mutational burden for extension cohort...'\n",
    "\n",
    "input_info_extension = r'storage/reference/ASCC_samples-extension.txt'\n",
    "info_extension = pd.read_csv(input_info_extension, sep = '\\t')\n",
    "\n",
    "# For convenience, we manually specify our list of paired primary-recurrent tumors in the extension cohort\n",
    "list_extension_primary = ['AC-V08-Tumor-SM-9J45Z', 'AC-V10-Tumor-SM-9J465',\n",
    "                         'AC-V11-Tumor-SM-9J467', 'AC-V12-Tumor-SM-9N4BH', 'AC-V14-Tumor-SM-9J46D', \n",
    "                          'AC-V15-Tumor-SM-9J46F', 'AC-V16-Tumor-SM-9J46G', 'AC-V18-Tumor-SM-9J46J']\n",
    "list_extension_recurrent =['AC-V08-Tumor-SM-9J461', 'AC-V10-Tumor-SM-9J466',\n",
    "                          'AC-V11-Tumor-SM-9J468', 'AC-V12-Tumor-SM-9J46A', 'AC-V14-Tumor-SM-9J46C', \n",
    "                           'AC-V15-Tumor-SM-9J46E','AC-V16-Tumor-SM-9J46H', 'AC-V18-Tumor-SM-9J46K']\n",
    "\n",
    "df_list_extension_primary = pd.DataFrame([], columns = list_extension_primary)\n",
    "df_list_extension_recurrent = pd.DataFrame([], columns = list_extension_recurrent)\n",
    "\n",
    "pilot_mutsig_rates = pd.read_csv('storage/mutsig-pilot/patient_counts_and_rates.txt', sep = '\\t')\n",
    "extension_all_calls = pd.read_csv('filter_unmatched/all_genes/df_filter_pass.txt', sep = '\\t', low_memory = False) \n",
    "\n",
    "print '...calculating for primary-recurrent pairs'\n",
    "\n",
    "df_extension_primary = format_extension_load(info_extension, extension_all_calls, df_list_extension_primary)\n",
    "df_extension_recurrent = format_extension_load(info_extension, extension_all_calls, df_list_extension_recurrent)\n",
    "\n",
    "pre_mut_extension = df_extension_primary.ix['Nonsyn', :].tolist()\n",
    "post_mut_extension = df_extension_recurrent.ix['Nonsyn', :].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ====================================================================\n",
    "# Create list of mutational burden for all primary tumors, pilot cohort\n",
    "## ====================================================================\n",
    "\n",
    "list_pilot_primary_plt = df_info[df_info['Tumor Type'] == 'Primary'].ix[:,'Tumor Sample ID'].tolist()\n",
    "df_list_pilot_primary_plt = pd.DataFrame([], columns = list_pilot_primary_plt)\n",
    "\n",
    "df_pilot_primary_plt = format_pilot_load(pilot_mutsig_rates, df_list_pilot_primary_plt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating figure 3 ...\n",
      "...Generating mutational burden plot for pilot cohort...\n",
      "...Generating mutational burden plot for extension cohort...\n",
      "...Generating mutational burden and neoantigen plot for pilot cohort primary tumors...\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABpkAAAGRCAYAAACezdsgAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzs3Xd8FHX+x/HXJ1QpOUFDO4QgJygCAjaQDnqc2DkVUVEB\nz4KCXU9OBc5+2OCnnqCC4gGiKJazgQUUFFSaUtRDAaUkYkVAEMj398d3EiebTbKbbLILvJ+Pxz6S\nnfnOd74zOzvz2ZlvMeccIiIiIiIiIiIiIiIiIvFIS3YBREREREREREREREREZPejh0wiIiIiIiIi\nIiIiIiISNz1kEhERERERERERERERkbjpIZOIiIiIiIiIiIiIiIjETQ+ZREREREREREREREREJG56\nyCQiIiIiIiIiIiIiIiJx00MmSSgzu8DMcsysUbLLAmBmXYPydEl2WeJhZrPM7O1kl0P2PGb2hJmt\nipi22szGJ6tMiRBsw4xkl0NEUoviksRQXCJSkJmNMLOcZJcjkczs/OAcdUyyyyIiew8zaxyce85L\ndlkktZhZLzNbYGZbzWyXmTVLQJ5ZZvZw6H3z4Pg7s7R5J0toG65Odllk76WHTCkkFNTnmNlfCknz\nfDD/t1KsZ7iZnVTykhaZhwteqaTY8oSCmvDrFzNbZmY3mVm18ihoSErtw9BNsdzXLjPbaGYvmdmR\nyS5fogQPKsLbucXMFprZ5WZmZbTOw4LvU8NS5hNr2R0QeUOkxMebmV1hZv1LsFwvM5tuZhvMbHtw\nPL1hZueZWUmuTeX6nQnO10PKc50i5U1xSZlRXFJKikt2y7gk/Pq6BPk1DsrVsjTlSrBoMVW5MrNa\nZnarmS02s03BTbgVZjbazP5UwmzL7fueop+riMQhIl7MMbOdwW+8KWZ2UBxZpVSsUVpm9k4R18Hw\nS5V4CmFm+wHPAjuAy4H+wPpC0ja3grHhD2Y2w8x6RCTPIUHHm5ntF1zH2se5XCUzu8zM3gvKud3M\n1pjZf8ysUyLKVpbMrEKw3ccnuyySOiomuwAS1a/AucDr4YlmVgvoHcyvVIr8hwOPAS+XQR4TgSnO\nuRLfbEqyZ4EXg/9rAJ2BfwKHA6clq1Ap5CHgA/y5owUwGOhpZkc755YmtWSJ4YClwJ2AAXWA84Ex\nQD3gpjJYZxv892kmsLYU+cRa9gtJbAWDK4H/AU/FuoCZPQRcGpT3YeAboBZwHDAeaADclcAyloUL\ngD8C/5fkcoiUB8UlyaO4pGiKSxKvrOKSsC0lyC8zKNf/gnxTwa347UsKMzsMeA0fQ00FHsHfiGsB\n9AUuBqomq3wxyiT1PlcRiZ/Dxyf/A6rg45QL8dfkVs657CIXdm6Nme2DP4ftKW4DHg297wxcxO/7\nKVeR+2Yv1x6oDtzsnJsZ4zKTgVeBCsCf8LHh62Z2nHNudpCmMbArQWXcH38d+wWYF8sCwcOz1/Df\nk9fxx8RPQbn6ALPN7HDn3OIElbEsVMRv94P4bRHRQ6YU9V/gFDOr5pzbGpreF3/xngGckJSSFcM5\n54Dd9UYOwBLn3OTQ+3FmVgXoY2Y1nHObE7ESM9vHOfdrIvIqZx+E94+ZvQ+8AFyGf2hQQFDTtrJz\nbnv5FLHUspxzU3LfmNk44DNgqJkNd84lKhjJWwWJq7VVbNmD8id6G2JmZlfgj5WHnXOXR8y+P6iB\nfkj5lyw2u/F3V6Q0FJckj+KSoiku2Y3iklIqk5ZbpeGcyyFJ5xczqwm8hL+JdoRzblnE/GHA7cko\nWyzMrCqwnRT8XEWkxGY6594P/p9gZl8A9+Mr590dbQEzqwiYc25HqlUIihL3xsU591ZEfpXwD5nC\n+2m3E2yHc87tLIfV1Q3+/hzHMgsjYsNX8A9/rgRmAzjnEvkwsyTXsf8AbYEznHPPR8wbaWYXkuSW\n0oUx3+tMJXT9lijUXV7qccAUfK2zPhHzzgVewT/hzscKGVPFQuOvBN0R5DYLvTDUjHR8ML+RmT1o\nZsvNbLOZ/WxmM82sQyi/4vKIOvaBmZ1kZh+Y7+bjRzN7wcwOjkgzIli2uZmNNbPvzXcN82xQWzqc\ntlPQ/HqVmW0zs2wze8rM/hjbbo5LNv4EmncRtSjj2gTTC2x/8NnMMLPOZjbXzLYS+tFnZtcHabYG\n8ztE5hukq2hm/zDfBcY2803Qx0bZN7nrO8LM5gT5fm1mVyVgX0R6M/jbJFh3bvc+w8zsIjNbAWwD\neoXKd6n5Lj1+NbPvgs+xyLEyzKxdkO/gKPMamm8KPTJ4n2ZmNwb7Kfd4W2hml5RkA4PAch6+Bk1G\naL2xHNPVzOxuM1sZ2t4PzKxPMH84vuUOwBz7vVl3QsbqiFb2wo7dSOa7X3nYzNYFx9sKM7vG7Pfu\neYJzQSPgWIuhub/5G6P/AD4HriikzB855yaGlzGzO4Lv+vbg7+1mVrmQdRR73MeaZ8R39/3gu3tH\nsP+6An8KbfdXRe1Pkd2Y4hLFJYpLIpjikoQx31XuTxbRPV9Q/l/M7E9mdj7wNv57/p9Quc4LpW9n\nvrvEH4Jj7CMzOyUiz9wunbqZ2Z3BMbs1OD4zI9JmBMfz6uD4/tZ810ddQmlGWJQxmcxsYOiY+tbM\nJkaeC4Lv7A4zq2Nmz5g/v/1gZo+aj5eKczHQELgm8gETgHNuu3Pu2oh1djSzt4L9+ouZvWmFd+9T\nqbh9FGueof3e08zuM7N1wGZgKMV8riKyW3sTH6/kXpNzu7rtbz5+WIVvDX+IRRmTyX6Pww4Jzo3f\nBde3R8zHINWD8/S3wfnnicjzp/k4aIaZrTf/u+9L878DI3/35Z6TG5rvCvpH4D0zuzgowxGRG2dm\nfYN5kV2xlYiZzTOzV6NMv8vMfg29rxKsd4yZnWpmS4Lz9EIz6xikOdnMFgXXoWXRymhmrc3sv+av\nwZvNx2h/jkjTK1jXmcHnsQbYChwYzL/czD6x3+P0T83s7zFu78XBsr+aj5snmFm90PwP8L0UAMwL\nylFg/xTHOfch/prTJJR3vjGZiihjkfvIzHoBy/HXsXvs998h1xeRZ0d8HDo2ygOm3DI/5pz7JLRM\nY/Px6cbQZ312EesYaGafm49fFptZ1yhpis3TQuM8me/a73N8HH0B/jhwwOWh7S52n8qeTS2ZUtO3\n+FrB/fFPuDGzA4Fj8F2jnBplmcJqHIbHItiIvyH0H+Ad4PFg+pfB3yPxN06fA9bgm30OAt4ysyOc\nc8tjyKPA2Adm1i9I/wn+BnM6MASYa2ZHOue+ilj2P0AWvguQP+F/gPwGnBPK9gygNr75cTZwEHAJ\ncJSZtS5F7dRq5puugu+WphNwHjDVObctlK6wMR6iTXf4i/AL+B/u4/Gfce6P+dwuSf4VbMd/gR+B\nyP7qnwd64i+0nwZ5DgWONLP2oZo/Dt/M9iV8F2ZP4Wub32NmS+NoZhyL3D6Wv4uY3g//OY8FfgBW\nA5jZg/jaO//Bd5NWJ9iGOWbWxjn3Q7SVOOcWmq8JdVawXOS6ACYFf4cDN+P303xgH3y3IR3x3YiU\nxJ/wrX9+CrYj1mP63/h9/xCwLEjXBjga/3k+B9QH/gaM5Pdm8ytKWM5iy04M45MEQfc7+P32SFCe\n44FR+GNraJD0XOAB/Pc1tzucopr7d8KfV8bEUfN6Oj4Iewp/U+0Y4EagJXBKRNpYj/tY84z87j6O\n/+6+g+/Ob1/gmmC7E9KaQCRFKS5RXKK4JERxScwqhY7fsG3Oudwu8wbgj58n8ccTZjYQOBm41Dm3\n0sx2AXcAw/D7O7cG+PtB+s7AG8E23Ya/AXImMN3M+jnnpkasfxT+xubt+PPKdfj9Fx4DYRrQGt8N\nzFf47/fR+JrH7wZpop1f/h6U9d0g3z/iK9Z0NrO2zrlwPAa+m5xlwPX4c95A/PfxH1H2W9gp+JZA\nkdsWlfmHYzOAdfhu/gz/oOodM+vhnPsgnBy4h2L2UZx5go8ZN+FjqGr4z6zQz1VEdnuFXZOvx58v\nHsRXmPkB3yozUu558j/AKvx5sRP+GrUF/9vtF/w1tiM+PlpL/q5kB+Nb/76O/712TLD+A/BxbeS6\n3gCWADfgK+U/jW+N1R/4OKJ8/YFvnHOJGlMpltg5rDM+Bn8Yfz24AXjFfCWWUfj9ux34OzDNzA7I\nvfaa2aHAXPw+yb0mDgReNbNTnXP/jVjXzfi444GgLD+b2aX47nunButKAw4m/7U0KjMbAdyCr2gw\nFh8nDgE6BdfKzcH8M4Ny3YK/Fkcdj6mYddXDx9Dh4zCW8VFj2UdL8NfHUfzeTR/AoiKyPi1Y/5Mx\nlr8uv1cuGo2PEc7CV87Y1zkXLQ6tBYzDf/7X4OOhRrk9IJQgzwFB2nH46/jcYNoEfByQW0n4i1i2\nSfZgzjm9UuSF72N9F/7C1w/fH23dYN4t+JNiRfwX+beIZVcB46PkOQH4KmJaDjAuStoqUabti79Z\nMjbGPHK3oVHwviL+QvAZsE8oXSt8QPF0aNrwIN9JEXnej7+ZUzM0rWqUdR8TLN8vNK1rUJ4uxez7\nxsGyu4K/4dfz+G5Vityv0bY/9NnsAk6JSLsf/kfwTHwT8dzpuU1j3w5N6xfk0SMij2ODtIOirO+4\n0LTK+BtkU0t4bHYN1nNRUO66QHf8RXUXcFLEftwM1I/Io0Mwb2DE9EPwF79biynD8OCYaRgxfQGw\nIPR+IfByCbdzFf4m5X7B6xB8wJQDPFeCY/oH4P+KWWfe974kZY6n7IUdu0ScP/CDau4CLoxINzWY\nfkjEsjNiLOOQaN+FItKfGJT/tojp/wry+Uu8x30J8yxQ3mBff1Gaz0wvvVL9heISxSVOcUkRZVBc\nUnzZox2/OcDEiLR9g+lX4cfo+Rl4pZDP/Owo61oBvBNl+nvAmohty8E/AAof41cQim/wD99y8K2E\nijsGdkV8h34NPrO00PTeQX53RXxndwG3R+Q5HciOYf9+DyyK4/P4OFhm/9C0evgHlfPi3UclzPND\noEKsn6teeum1e7xC143jg/NgfeAkfIWOHUCbIF3u9/0boFpEHrnX6/NC03LjsCcj0s4L1jchyvT1\nEdOixWf/wF8fG4Sm5Z6T74+Sfgo+9qwQmpaBjwVvj0wfw36Ken3FjzH5apTpdwJbQ++rBPvlVyAz\nNP3UYPoW8sd9pwbrDe/b/wbLHxialo6vUPRlaFqvIM8vI/cl/oHKRyU4Xurh46w3Iq4zfYJ1jQhN\nGxSU/agY8m0eLH9TcBxm4GPx94I8/hZKuwHffX/ksmeWYB/lLnt1jNv/SlCeAr9zCkn/YJC+U2ha\nJeAjfLxUI6IcG4F9Q2mPDqYPKEWePxG61kcch2PiPQb02nNf6i4vdb2AP6HlNlc8B3jGlWG/py5U\ny9bMqppZbfwP1w/xA9KVxOH4i8i/Xaivf+fcp/jaJMdHFgNfyzJsNr5mS17XJS5Ue9fMagS1JL/A\nn/xKWlaAJ/A3SI7F1xC8C/gLvlZCaWxwzr0YMe04/In8/5xzLqIMkX3OnomvubHEzPbLfQGLg7SR\nzZ9XuVDNYOdrE88jaNZcCv/GX7Q24Ju/H4C/mEYOtP6ic25DxLS++Ca1/43Yho347tOKa2Y+BV87\n5qzcCWbWDF+jNPz5/AQcahFdxMShS1CmjfiapZfgv48XBvPjOaZ/Ao62iC5gylBxZY/VCfha6xMi\npo/C1zor6dgr6cHfTXGUwwH3Rkz/V1COEyOmx3Lcx5tntO+uyN5IccnvFJcoLgHFJbFYhG+ddGzE\n685wIudbGk3Gt2p5Hn9TclAsKzCzw/A3QSaHP0f7fVDthmZ2UGgRh39IHT7GZ+NjgNzj8Vf8zcNu\nhbTEKsyx+AeoDzg/XlPu9r2K704nMsaA6OeX/c2sejHrSifGeCqosdwO/3Avrya3cy4L30LgSDPb\nP7RIsfuoBHkCPOoSP4aYiKQGw98834hv3fgC/nx4rnNucUTaiS72sY4cvvVEWG4ryceiTK9rfsw3\nv3AQn5n3h+Cc/h7++t0uyvoiz8ng46D9yX89PRsfCz4V22aUiZnOudWh97n75S3n3NcR08Pn70r4\n69UL7veWzjjnNuH3daaZtYxY1xMufwt68DFFppnFG+f2wsfz94WvM853Hfcl0a+V8RiJPw6z8Z91\nG3wF00djzaCE+yhW6cBOF3svByfgH+bNCZVjB74FUg2gW0T6p93vraZxzs3HP9SLvCcST57Pha/1\nIoVRd3kpyjn3q5k9D5xrfhDjg/D9XpaZ4EQ6At/tzAERs0s63kgmPjD4PMq85cDxZpbhnNsYmr4m\nIt2Pwd/aobLWx9/wPgH4Qyitw9dyLqkvXf7mzi+b2XfAKDM7wTn3SgnzXRVlWuPgb74mpc65nVZw\nfJdm+IvCRgpy+O5dwiL3Ifj92Kr4ohbpTnyT5hx8bdjlhdxgjHa8HITvGiMryjxXyDK/J3DuCzNb\niL+Zc08wuV9QlqdDSW/CB7XLg65s3gSedc7NLir/kEX4Js/gbzJ84fJ3l5NJ7Mf0Nfimw2vM7BN8\n7fCnnXMLYyxLvIore6wygZVRbgQsD/42oWRyb4bUjKMcG51zP4YnOue+C76XkeWI5biPN89o312R\nvY7iknwUlyguUVwSmx+cc+/EmPYy/I2nw/A3JaN9LtE0C/6ODV6Rco/H/4WmRXb9mO877ZzbYWbX\nAPcBWWa2AP/AbrJzrqiuYDKDv4V9FpGVdHKcc2uLKMsWCreJ+OKposoFPv4J30Aqch+VIM9iv1Mi\nstu7Al8ZYhc+PlgR8bA6V7zngsjz0U/FTK+Fr3yC+THi7sC3Xg6P11RYfBatbDPxLYb741u3gI9N\nP3bOfRZD+cvKNxHvfypmeu6YmQ3wDwCLO38vDU2Ptl9ux3eN95GZrcbHN9Odc68VU+7M4G+09a/A\nd31YGuOAZ/Hx2M/42DDebqtLso9itQmoaGaV3e9dS0dlZoav1BYtZlxOaMyzkMjvBfhjoHYp8tT1\nW2Kih0yp7T/4/i3vwHeBEtmvdVi0izdE79u2MGPwtSIfxPeF/SP+xDyM0tc0jUdhNdwM8k6Kb+J/\nMN6NvxBtxu+DqZDwFnpvBevugq+dA/Hv718LmR6rNHw3KJcHZYn0Y8T7IvdhKSx3sfU5HG170/Dl\nPL2QcsSyjybjb6wd5Jz7H/7Gzlzn3LrcBM65982sKf6HfG7N78Fm9ohzrsAA3VHEc0OkSM656WbW\nBN9dwLH4fmuvMbNhzrm7E7GOCAkrexlZgf/sW+PH5ki0sjjuS/vdFdmTKC7JT3GJ4hLFJYnTHt+9\njcM/fJwS43K536+/U3C8jFyRN4GKPR6dcw+Z2Yv4saF64rvyu9HMBjrnJhWyfLwK+97mK0shVgBt\nzaxSUPs40RRTiUg8HL6r2FjGVIv3XFDY+ai4+CwTHzOtxJ/Dv8Z3C/xH/Hg4kfHZrmgVVZxzOWb2\nFHClmaUDDfGtiC+PayuKF28sV6L9UkIFPjPn3LKgBffx+NbwvYALzWy6c+6vpVhXaf0vxtgwWVbg\neyVoTeFxS2no+i1Jo4dMqe0tfA2MbvhBbIvyI9FrYmTGsb6++P5urwpPNLNb48gj0mr8yexgfA3A\nsBbA5ojawrFohe+T/jzn3H9C5azK77UzEin3e1IjNK2w/R1PC481+H3THH+jBgAzqxjkE25WvhLf\nD20qP0Aozkp88PGxc+6XEuYxFd+tWT8zewm/7+6PTBTk/zTwtJml4YPIi83sTudcZM2eeK0mjmM6\naFY8AZhgZlXwXbeMMLN7gpZCRd1gSJbVwBFmlhbu7gW/fZC/Bnw85Z+Dr2l+jpndEZF3YeX4s5nV\nCrc8Cro52J+StTJKVJ6p+LmJlDXFJdEpLtk9KS5JkbjEzGoBj+NbbX0EXG9mLznn5oWSFVaulcHf\nrYm+qRS0MnoYeNjM/oDv4vGfQGEPmVbz+2cRWbv9EBLbOvpFfM38M4soT7hcBOWKFC22i0Wi8lQ8\nJSJl5WSgKnBCuNWomR1XgryeAG7An3MPxHep+nRRC5TAj0SPG0vai0hh1uPLX+prQtCF3vTghZnd\nB1xhZkc65z4qZLHVwd+DQ/8TmpYKPYnEs4/ivY69CFyNH6OryIdMzjlnZl/HWI6YJDBPXb+lAI3J\nlMKCpsWD8X2KRvY3G2kl0CHoWgYAM2uHH+gu0hai34jIIaKWhJl1xtcsjDWPSB/juyG5xMz2CeXb\nEl/ToSTdvOTemI6s0XEtZXNMn4w/gUbeXPlD0A884MdgAM6LI9+Z+AvXkKAWdK6BFNy3TwMZZnZV\nxHTMLC34cZ7qnsZ/Pv+MNjOWPu+DmsHv4WsKn4Xvs39aRD61I5bJwTfbh9J1WZQrpmM6+FzSwwsG\nzbS/wDe9zu1rfwv+hkSBsgVjCjQPr6ecvIyvTTwgYvq1+O/Cf0PTYj0X5Aaht+Nvwo2OOO4BMLMj\nzez8UDkMH4SFXR+UI3LMjVgkKs+Yt1tkT6G4pFCKS0IUlyguKYFH8N24nIuvbf4lMDFiPbldx0WW\nayG+K7xrggdBkWWOHBeoWGa2j4XG9ABwzv2MvxlW1Gc2Ez/uwRVmlnc+MLPjgUNJbCvusfhxT+41\ns0MjZ5ofx+5eAOdcNv446R/eH2ZWD7/P58c71kIC8yzscxURKa0C8VkQ3+T+po1Z0FXqfHxX0ecA\nr5WwW/qirARam1ne+TBojVXS8ZCjClq/zgBOCVo3564rHbgIP45muAVw1H0VGd8EPgn+FnVOfwPY\niW8ZlhdvmtlpwJ8omx5P4hLnPorrOuacew8fL1xsZqdGS2Nmg8ysdfD2Zfw4hx1D8yvhu6f8BZgV\n63aFlDrPoKu/nej6LSFqyZR68t10dX5Q5lgGnR8LnAHMNLOn8U14/wZ8Sv6xAQAW4GvyX4mvkbzK\nOfdhsJ7zzGwL/sZFC/ygu0sp2Od3YXnk45zbZWZX47vYed/MngzKczm+f9SbYtg2yL9fVuB/EN9n\nZo3xA/p1w9fmi/ZjJp5moYeZ2TnB/9WDPPvj+2L9TyjdFPzg2y+Y2Wj8j/MB+B/5MQ2m7Jz73szu\nxu+DGWb2Av6ieh7+x3U47SQz+ytwT3CDbTb+hP4noA9wM76P/biY2RPB+jJd/sEhE845N9fMxgBD\ngwvma/gLWBN81zFTKORGT4Qp+JsRf8QPdhkZ3K0wszn4mrDZ+BoalwGfOj8Idmm3I9Zjuiawzsym\nA0vwLXja4b9Trzo/aCT4myMO3wXL/vibE28FP8yHALfgj+93S1v2ODyGP3/8O7hhuQLoHbwejOh7\negG+ZdIw/HH7bVE1251z95tvVj8Y6GpmzwBr8bW2egTrGBakfdXMXgeGmdkB+MC+A/7mxUvOuTfi\n3bAE5rkA6G1mo4L/Nzvn/lvMMiK7I8Ul0SkuUVySS3FJ4eqFjt9IU4IuiM7Gnyuucc6tADCz8/Ct\nn+8DLg3Sr8B31zLYzLbhb+rMd86tNrMB+JtWy81sPL5VXl3gaHwLooNC643l+9cMeMfMpuHHJ9gM\ndAX+jG/ZFJVz7gczG4HvUvStYPmG+P22Gt/qLSGcc5vM7BT8Q8QFZjYFH9PswB9jfYEM/DhcBH9n\nAvPNbBx+P1yEH6MksuJNrOeoRORZ6OcaYxlEJDWUtvvbsvA6vvLMq2Y2Fl+x5Ex8jFQSE/DxrcPf\njC+JovbTo/hr95vBtWx/4BJ8xZQClQlK6Ub8de19M3sI2IqPB+oDkQ8+Civzu2a2BvgA3/InE/8b\n/xv8NTwq51yWmd2BjydmBnFJJn7bV1KwRXiyjq2Y9pFzbp2ZZeF/s6zFx15LXNHjdZ2DjzufM7PX\n8NfSTfgxaE/Dd6XXLkh7O75r59eCuDUbPwZoO+By59zmEmxbovJcCJxgZkOAb/Hjei8oQXlkT+Gc\n0ytFXvjmkruAY4pJNwHYHmX6ZfhmjVvxP2S7B2m/jEjXEv/DcHOwvvHB9Br4cQ/WBfM+wPfXHk8e\nudvQKCL9iUF+W/DNgJ8HDo5IMzxYtkHE9K7B9C6haU3xT99/CPKbjr8p8BXweFHLFrJPGwfpwq/f\ngv35f0DtKMt0xZ9Ut+FvvgyOtv1BHm8Use7r8T+GtwBz8TW038b/oA+nM2BosM7c/bgYf1OpYXHr\nK+RzfDb4DNOL2T+5+/HsGPfjjUWkORc/tsam4LUMP+7GwUXlHVq+VrDPdwL9osy/AR/UfIv/LvwP\nf5Ni/xjyLvKziueYBirhByT/CPg+2M/Lg+O8WkReg4Nj6Lfw8Rr6ThR5/MZT9kKOg3zfm2DavsBD\n+PPBNny3L1dHya8h8Co+mNoFvB3j/vsLfiD0DfgbWBvxtYXOjUhXBR8ErQrKsQq4Fagcy/YXsr2l\nyjOYl46/sfh9sN1fxbLdeum1O71QXKK4RHFJLN8TxSWFlz3yGA6/quEfzP1AlNgBf13eBfw5NO2v\nweezPZh3XmjeIfgxsrKCz2MNvjb0GaE0Uc9poePkvOB9beABfI3sn4LjYjH+pmJaxDliZ5SyDwjS\n/xp87k9S8DxS2Hkz6jmriP1cO9hXS/APSbfiY7bRQJOItMfgx4/LPdZnAkcXsv4i91Ei8ozlc9VL\nL71S/1XcdzyUrtBrd7RzDIXHYTFPx8eNH+KveeuDc2OLKOuKek6OyD8df43dCFQsi/0EnI2vuPRr\ncA06FX/t3hpKUyXIZ3TEsvFOb4WPXX8K9s97wLERaXoFy/aJUtZL8PFhVlDeVfhKNw2L2xfB8hcF\n2/gr/iHHeKBuRJpBwfqPiiG/5kHaq2JIux54KMqyZ8a7j0LH2SJ8/LELuD6GMlTC/1Z6Dx+35cYu\nk4D2Ub4fk/GV17biK9f1i3X7I7c3EXkG89vifytsCdI9HO/3Qq8962XOOcqTmT2O/xGU7ZxrHZo+\nBP+DZifwinPu7+VaMJG9UFDj4gl930REyoaZNcS36KiL77ZjnHPu/8xsOL5lz7dB0mHOucjxVET2\nKopLRET2XEGXlo/hK4bkAAOdc/OTWyqR2JhZdfzDkAnOuSHJLo+ISKpJRnd5E/A1MPO60DCzbsBJ\nQCvn3M6S9J0tIvExsxb4gTAT1nWHiIgUsBPfCnCx+TFyFpjZzGDefc65+5JYNpGUobhERGSPNxrf\nPecZZlb8mmNhAAAgAElEQVQR36JRZHfRH9gHeCLJ5RARSUnl/pDJOTcn6K8+7FLgLufcziBNXAOP\nikj8nHPL0SB9IiJlyjmXhe9GAufcZjNbge8iClKz/3qRpFBcIiKy5zKzdKCzc+4CgODez6YiFxJJ\nAWbWHd8l6wjgXacxZ0REokpLdgECzYAuZjbPzN4xsyOSXSARERGRRDKzTKANfnB2gMvNbLGZPRZ0\nISMiIiKyJ2oCfGdmE8xsoZmNM7N9kl0okRjcgh9L8TP8GEEiIhJFqjxkqgjUcs61xw82/EySyyMi\nIiKSMEFXedOAK5xzm4GHgQOdc23wLZ3UbZ6IiIjsqSoC7fCDz7fDDzSv8fck5Tnnujvnqjrnujjn\nvkp2eUREUlUyxmSK5hvgeQDn3EdmlmNm+znnvo9MaGau3EsnIiKyF3POqVu3UgjGHZgGPOWcexHA\nObcxlORR4OVCllXcIyIiUo4U95SJtcA3zrmPg/fTgBvCCRTziIiIlK9ExjzJaslk5B+H4AWgB4CZ\nNQMqRXvAlMs5t1e+hg8fnvQy7Okv7WPt4z3lpf2sfZyolyTEeGC5c2507gQzqxea3wdYWtjCyT4G\n9vbvgbZxz3jtDdu4t2yntnHPeKXqNkrZcM5lA98E93sAegLLo6Tba16p+h3Q9mp7ta3aXm1v8stR\nHq9EK/eWTGY2GegG7GdmXwPD8TdfJpjZp8B24LzyLpeIiIhIoplZR+Ac4FMzWwQ4YBhwtpm1AXKA\n1cDFSSukiIiISNkbCkwys0rAV8CAJJdHREREEqTcHzI5584uZFb/ci2IiIiISBlzzs0FKkSZ9Xp5\nl0VEREQkWZxzS4Ajk10OERERSbxkdZcnJdCtW7dkF2GPp31c9rSPy4f2c9nTPhbZO74H2sY9w96w\njbB3bKe2cc+wN2yjSFH2tu+AtnfPtTdtK2h793R72/YmkpVFH3xlyczc7lZmERGR3ZWZ4TQAdtIo\n7hERESk/inuSRzGPiIhI+Ul0zKOWTCIiIiIiIiIiIiIiIhK3ch+TSUREvMzMTNasWZPsYojQuHFj\nVq9enexiiIjIHkxxj6QKxT0iIiIiiaXu8kREkiRomprsYogUeSyq25jkUtwjInsKxT2SKhT3pCbF\nPCIiIuVH3eWJiIiIiIiIiIiIiIhI0ukhk4iIiIiIiIiIiIiIiMRND5lEREREREREREREREQkbnrI\nJCIiIiIiIiIiIiIiInHTQyYRESk3aWlpPP/888kuhoiIiEiZU9wjIiIiInsDPWQSEUkx9Rs2wszK\n7VW/YaNy27asrCxOOumkcltfaT355JPUrFkz2cUQERHZYynuSR2Ke0RERESkJComuwAiIpJf1rpv\naHzDy+W2vjV3l/3Njx07dlCpUiXq1KlT5usqav3xcs5hZmVQosQp6baJiIikAsU9Zbf+eCnuERER\nEZGSUEsmERGJS/fu3bn00ku58sorqV27NrVr1+b666/Pl6ZJkyaMHDmSQYMGUatWLc4991wgf7cx\na9asIS0tjalTp9KtWzeqVatGu3bt+PTTT1m2bBkdO3akRo0adO7cmTVr1uTl/dVXX3HqqadSv359\natSoweGHH84rr7xS7Pp79uzJkCFD8qX75ZdfqF69Oi+88EKB7Zw9ezYDBw5ky5YtpKWlUaFCBf75\nz3/m5X/fffcV2C9Dhw7NV4Zbb72VAQMGkJ6eTqNGjXjmmWf4+eef6devHzVr1qRZs2bMnDkzXz7v\nvvsu7du3Z5999qFevXpcffXV7NixI996Bg8ezHXXXUedOnXo1KlT0R+YiIiIlJjiHsU9IiIiIlI0\nPWQSEZG4TZ48Gecc8+bNY9y4cYwbN44HHnggX5r777+fQw45hAULFnDHHXcUmteIESO48cYbWbx4\nMfvuuy/9+vVj6NCh3HnnnXz00Uds27Yt302MzZs307t3b9566y0++eQTTj/9dP7617/yxRdfFLn+\nv/3tb0yZMiXfjYvJkydTs2bNqF3ZdOzYkQceeIBq1aqRnZ3Nhg0buPbaa+PaT6NHj6Z9+/YsWrSI\nvn37cv7553POOedwwgknsGTJErp06UL//v357bffAFi/fj29e/fm8MMPZ/HixYwfP54pU6YwbNiw\nfPlOmjQJgDlz5jBx4sS4yiQiIiLxUdwTG8U9IiIiInsnPWTajWQ2rF+u/ZWHX5kN6yd780UkhdSv\nX5/Ro0fTrFkzTj/9dK677roCNVy7du3Ktddey4EHHkjTpk0Lzeuaa66hV69eNGvWjGuuuYbly5cz\ndOhQunTpwiGHHMLll1/OO++8k5e+devWXHTRRbRo0YIDDzyQG2+8kbZt2zJt2rQi19+nTx/MjOnT\np+elmTBhAueffz4VKlQoUK6KFSvyhz/8ATMjIyODOnXqUK1atbj2U69evbj44otp2rQpI0aMYPv2\n7Rx00EGce+65HHjggdx88818++23LF26FICHHnqIP/7xjzz00EM0b96c3r17c9ddd/Hggw+ybdu2\nvHybNGnCqFGjaNasGc2bN4+rTCIiIhIfxT2xUdwjIiIisnfSmEy7kTXrssgZnp6UdaeNzErKekUk\nNbVv3z7f+w4dOnDLLbewefNmatSoAcARRxwRU16tWrXK+79u3bqYGS1btsw3bcuWLWzbto2qVauy\ndetWRowYwSuvvMKGDRvYsWMH27dv57DDDsuXb+T6K1euTP/+/Rk/fjxnnnkmy5Yt46OPPirTGrGt\nW7fO+7969epUq1atwLYBfPvttwB89tlnBfZtp06d+O2331i5cmXesocffniZlVlERETyU9wTG8U9\nIiIiInsnPWQSEZEyUb169ZjShQdvzh1sOtq0nJwcwNcAnjFjBvfeey9/+tOfqFatWr6uV4pa/4UX\nXshhhx3G2rVrGT9+PB06dKBZs2bxbRh+jAXnXL5p4e5oom1b7rYUtW2FiRyIO9Z9KyIiIuVDcY/i\nHhEREZG9lbrLExGRuM2fPz/f+w8++IAGDRrk1eYtS3PnzuW8887j1FNPpWXLljRo0IAvv/wypmVb\ntGjB0Ucfzbhx45g0aRKDBg0qMn3lypXZtWtXgekZGRls2LAh7/22bdv47LPP4tuQKA455BDmzZuX\nb9p7771HlSpViux6R0RERMqO4h7FPVJ6ZrbazJaY2SIz+zDZ5REREZHE0UMmERGJ2/r167nqqqv4\n4osvmDZtGvfccw9XX311QvKOrCkbqVmzZkyfPp1Fixbx6aef0r9/f7Zv3x5z/hdeeCH/+te/2Lp1\nK2eeeWaRaTMzM9m2bRtvvvkm33//Pb/++isAPXr0YNKkScyePZtly5YxaNCgqDdl4jV48GDWr1/P\npZdeymeffcYrr7zCjTfeyJAhQ6hatWqp8xcREZH4Ke5R3CMJkQN0c861dc4dlezCiIiISOLoIZOI\niMTtnHPOYdeuXRx99NFcfPHF/O1vf+PKK6/Mmx/u4iQscnq0dIUtm+u+++6jTp06dOnShRNOOIEO\nHTrQuXPnmPPo27cvlStXpm/fvsV2v9KhQwcuueQS+vXrR506dRg1ahQAN954Iz169ODUU0/lL3/5\nC507d6Zt27Yl2rbwtAYNGvDaa6+xePFi2rZty4UXXsg555zD7bffHtO2iYiISOIp7lHcIwlh6B6U\niIjIHsmKqzmVaszM7W5lThQzI2d4elLWnTZyU7G17EQkPmYW9XtVv2EjstZ9U27lqPfHA9iw9uuY\n03fv3p1WrVoxZsyYMixV2Vm/fj2NGzfmvffeKzDY9N6qsGMxNE93eJJkb457RGTPorgnORT3FKS4\nJznM7CvgJ2AXMM4592jEfLdy5coyW3+9evU0vpeIiEgg0TFPxURlJCIiiRHPjQ+J3c6dO/nuu+8Y\nNmwY7dq1040WERGRFKC4p2wo7pEU1NE5t8HMMoCZZrbCOTcnnODQNofm/V+xSkUqVamUkBXv2LaD\nzp0689qLryUkPxERkd3NrFmzmDVrVpnlr4dMIiISl92125K5c+fSvXt3mjdvztSpU5NdHBEREdkN\nKO4RSQzn3Ibg70Yzmw4cBeR7yHTQ/x1UJuvevGwzP8z/oUzyFhER2R1069aNbt265b0fOXJkQvPX\nQyYREYnL22+/newilEjXrl3JyclJdjFERERkN6K4R6T0zKwakOac22xm1YE/A4m9uyUiIiJJo4dM\nIiIiIiIiIiJSVuoC083M4e9DTXLOzUhymURERCRB9JBJRERERERERETKhHNuFdAm2eUQERGRspGW\n7AKIiIiIiIiIiEhqMbMMM8sIvW9lZreZWb9klktERERSix4yiYiIiIiIiIhIpGeAkwDMbH/gXeA0\n4BEzuyaZBRMREZHUoYdMIiIiIiIiIiISqTUwL/j/dGClc+5Q4Dzg4qSVSkRERFKKHjKJiIiIiIiI\niEikfYDNwf/HAi8F/y8EDkhKiURERCTl6CGTiIiIiIiIiIhE+h/Qx8wOAP4MzAim1wV+SlqpRERE\nJKXoIZOISIrJbFgfMyu3V2bD+sne5JTWvXt3hg4dmrB0IiIi8jvFPalFcY9EGAncDawG5jnn5gfT\newGLklUoERERSS0Vk10AERHJb826LHKGp5fb+tJGZpVq+R07dlCpUqUElSYx605mmVLZzp07qVhR\nl34REUkdintKv27FPdEp7ik959zzZtYIaAAsCc16E3guOaUSERGRVFPuLZnM7HEzyzazT6LMu8bM\ncsysdnmXS0REYtO9e3cGDx7MddddR506dejUqRMAmzZt4qKLLqJu3bqkp6fTvXt3FixYkG/ZefPm\n0bNnT2rUqMG+++7LscceS1ZWVl6+kTViBwwYwMknn1zsutPS0nj44Yf561//So0aNfjHP/4BwPLl\nyznxxBNJT0+nbt26nH322WRnZ+fL/6STTmLMmDE0bNiQ2rVrM3DgQLZt25Y3f/bs2Tz00EOkpaVR\noUIFvv766wL7pLB0s2fPJi0tjR9++CEv7Zo1a0hLS2PhwoUAeWlef/11jjjiCKpVq0aXLl1Yt24d\ns2fPpk2bNtSsWZOTTjqJH3/8MS8f5xy33norjRo1omrVqrRu3ZqXXnqpwHqefvppevbsSfXq1Rk3\nblysH7OIiIiguEdxjzjnsp1zi4AMM0sLps13zn2W5KKJiIhIikhGd3kT8E2r8zGzhsBxwJpyL5GI\niMRl0qRJAMyZM4eJEycC0Lt3b7Kysnj11VdZvHgxXbp0oWfPnnk3N5YsWUKPHj1o1qwZ77//Ph9+\n+CH9+vVj586dpV43wD//+U9OOOEEli5dymWXXUZWVhZdu3aldevWfPzxx7z11lts2bKFU045JV9+\n7733HsuWLeOtt97imWeeYfr06YwePRqA0aNH06FDBwYMGEB2djYbNmzggAMKjnFcVDozK5A+2rQR\nI0YwZswYPvzwQ3788Uf69u3LbbfdxmOPPcbs2bNZtmwZI0aMyEv/wAMPcO+99zJq1CiWLl3Kaaed\nRp8+ffjkk/x1OIYNG8bll1/O8uXLOfXUU2PcyyIiIpJLcU9+inv2HmZWycz+ZWa/AOuAzGD63WY2\nOKmFExERkZRR7m3HnXNzzKxxlFn3A9cBL0WZJyIiKaRJkyaMGjUq7/3bb7/NJ598wsaNG6lSpQoA\nI0eO5KWXXuKpp57i2muv5V//+hdt27bl3//+d95yzZo1K/W6c5111lkMHDgw7/3w4cNp06YNd9xx\nR960J554gv3224+PP/6YI444AoA//OEPPPLII5gZzZs354wzzuCtt97ihhtuID09ncqVK1OtWjUy\nMjIKLVOs6XI55wpMu+222zjmmGMAuOSSSxg6dCgLFy7ksMMOA+D888/nued+75Xk3nvv5brrrqNv\n376A39/vvvsu99xzT76bUEOHDuW0004rtkwiIiISneKe/BT37FWGAycB5wKTQ9M/BG4AHk5GoURE\nRCS1pEQHxWZ2MvCNc+7TaLWcREQktRx++OH53i9cuJAtW7aw//7755u+fft2vvrqKwAWL15Mnz59\nEr7uwqYvWLCA2bNnU7NmzXzTzYwvv/wy72ZLixYt8tWwbdCgAR9++GGpyxkPM6NVq1Z57+vWrQtA\ny5Yt80379ttvAfjll19Yv3593s2ZXJ06deK1117LN62w/SUiIiKxUdyTWIp7div9gIHOudlmlhOa\nvhSI/6mpiIiI7JGS/pDJzPYBhuG7ysubnKTiiIhIDKpXr57vfU5ODvXq1WPOnDkFaqump8c2mHda\nWlqBZXfs2FHsuosq04knnsi9995bIN/cmxlAgYGyzYycnBwSIS3N90obXn+0bYosR+7NnwoVKsRd\nrsjKGoXtLxEREYmN4p7YKO7ZIzUg+pAGFUmB+0kiIiKSGlIhKGiK79d3ifkIsSGwwMyOcs59G22B\nCy64gMzMTAD23Xdf2rRpQ7du3QCYNWsWwJ77frXvw7tbZsVyfZ8r6duv93q/B73fk7Rr147s7GzM\njCZNmkRN07ZtW95++21uvfXWqPMzMjLYsGFDvmlLliwpNL9YyvTss8/SqFGjfDct4lW5cmV27dpV\nonQZGRk459iwYQP77bcfAIsWLYo6NkE8atasSYMGDZg7dy7du3fPmz5nzhxatGhRojxzj8/c/1ev\nXl2qMoqIiOypFPfs/nGPxGwZ0AVYHTH9TGBBuZdGREREUpJF6x+5zFdqlgm87JxrFWXeKqCdc+7H\nQpZ1yShzKjAzcobHVjMu0dJGboral7aIlJyZRf1elfd3Pd7vd/fu3WnVqhVjxozJN71r16789NNP\n3H333Rx88MFs2LCBN954g+OOO46OHTuyZMkSOnToQP/+/bnsssuoWrUq7733Hr169aJhw4aMGzeO\nq666iqeffprmzZszduxYHn/8cbp06cJLL71U5LrT0tKYNm1avm5pNmzYQNu2benYsSM33HADGRkZ\nfPnllzz77LPcd999VK9enQEDBvD999/n5Q++j//nnnsubyDpiy++mIULFzJt2jRq1KhB7dq1o94s\niUy33377sXPnTpo2bcqRRx7JnXfeyapVq7j++uv59NNP+eijj2jXrh2zZ8+me/fufPfdd9SuXRuA\n5557jjPPPDPfzZuxY8dy880353UdM3r0aIYPH84jjzzC4YcfzlNPPcWdd97JokWLaNmyJWvWrKFJ\nkyZ8/PHHtGvXrsjPtLBjMTRPLYyTZG+Oe0Rkz6K4R3GP4p7dj5mdBPwH+BfwD2AkcDBwNnCCc+7N\nBK7LtZzQsviEJbB52WbqzK/D/Hfnl0n+IiIiu5tExzxpicooVmY2GXgfaGZmX5vZgIgkDnWXJyKS\nsgqrjfrqq6/So0cPLrroIg4++GDOOussvvjiCxo0aADAYYcdxptvvsnnn39Ohw4daN++PVOnTs3r\nLmXgwIEMHDiQQYMG0alTJ9LT0wuMZVDYuqNNr1+/PnPnzqVChQocf/zxtGzZkiFDhlC1atW8Qbpj\nce2111K5cmVatGhBnTp1+Oabb2JK9/XXX1OxYkWmTp3KV199RZs2bRg5ciR33nlnTOUvztChQ7nu\nuuu44YYbaNWqFS+++CLPP/98vvEMNM6hiIhI6SjuUdyzN3POvYxvtfRnIAcYDhwEnJTIB0wiIiKy\ne0tKS6bS2Jtr9Kolk8iepbBalJkN67NmXVa5laPxH+uxeu2G4hPKHks1elPX3hz3iMieRXGPpArF\nPalJLZlERETKT6JjnlQYk0lEREJ040NERET2Fop7RPYOZpYGfAysdc6dnOzyiIiISOLoIZOIiIiI\niIiIiGBmv+CHMSiWcy6erlauAJYDyemeRURERMqMHjKJiIiIiIiIiAjA5YnO0MwaAr2B24GrE52/\niIiIJJceMomIiIiIiIiICM65J8sg2/uB64A/lEHeIiIikmR6yCQiIiIiIiIiIlGZWQ+gRfB2uXPu\n7TiWPQHIds4tNrNuQMIGGRcREZHUoIdMIiIiIiIiIiKSj5k1AZ4DWgPrg8kNzOxT4K/Oua9iyKYj\ncLKZ9Qb2AWqa2UTn3HmRCbNfyM77v/rB1alxcI3SboKIiIgAs2bNYtasWWWWvzkX03iOKcPM3O5W\n5kQxM3KGJ2eMzLSRm9hb97tIWTEzfa8kJRR1LAbzVOM0SfbmuEdE9iyKeyRVKO6JnZm9DVQA+jvn\nvg6mNQKeBJxzrkec+XUFrnHOnRxlnms5oWUCSl3Q5mWbqTO/DvPfnV8m+YuIiOxuEh3zqCWTiEiS\nNG7cGDP9hpXka9y4cbKLICIiezjFPZIqFPfEpQPQPvcBE4Bz7mszuwr4IHnFEhERkVSih0wiIkmy\nevXqZBdBREREpFwkKu5JdO8O6rFBpEhf47u4i1QV+CbezJxzs4HZpS2UiIiIpJa0ZBdARERERERE\nRERSzjXAGDNrb2YVzCzNzNoDDwTzRERERNSSSURERERERERECpgCVAHmAjnBtDRgFzAp3AWmcy45\nA0iLiIhI0ukhk4iIiIiIiIiIRLo82QUQERGR1KeHTCIiIiIiIiIiko9z7slkl0FERERSnx4yiYiI\niIiIiIhIVGZWG6hDxLjezrnlySmRiIiIpBI9ZBIRERERERERkXzMrC0wAWiVOwlwob8VklQ0ERER\nSSF6yCQiIiIiIiIiIpHGA+uAK4Bs/IMlERERkXz0kElERESkjJhZQ2AiUBfIAR51zo0xs1rAVKAx\nsBo40zn3c9IKKiIiIlLQQcAZzrmVyS6IiIiIpK604pOIiIiISAntBK52zh0KdAAuM7ODgb8Dbzrn\nmgNvAzcmsYwiIiIi0cwBDkl2IURERCS1qSWTiIiISBlxzmUBWcH/m81sBdAQOAXoGiR7EpiFf/Ak\nIiIikioGAY+Z2YHAUmBHeKZz7t2klEpERERSih4yiYiIiJQDM8sE2gDzgLrOuWzwD6LMrE4SiyYi\nIiISzUFAW6BXlHkOqFC+xREREZFUpO7yRERERMqYmdUApgFXOOc2U3DgbA2kLSIiIqlmLPAm0Aqo\nA2SEXqogIxKDRYsWUbFiRTp37hx1/qxZszjppJPIyMigWrVqtGjRgqFDh7JmzZpi8549ezZpaWl5\nrzp16tC7d28++eSTvDTdunXLm1+lShUaNGjA8ccfz6RJkwrk9+ijj9KjRw9q1apFWloaX3/9dck3\nXMpMWRxTTz75JGlpaVSoUIG0tDQaNGhA3759Wb16dV6a3377jSFDhpCRkUGNGjU45ZRTWLduXd78\nNWvWcOGFF9K0aVOqVatG06ZNGTZsGNu2bUvYtkvq0kMmERERkTJkZhXxD5iecs69GEzONrO6wfx6\nwLeFLX/BBRcwYsQIRowYwQMPPMCsWbPy5s2aNUvv9V7v9X7vfL96J7NW70zc+2Rvj94n5f2sWbMY\nMWIEF1xwARdccAFSQENguHNumXPuO+fc9+FXsgsnsjt47LHHuOyyy1i6dCmff/55vnljx47l2GOP\nJSMjg2nTpvHZZ5/x+OOP45zj9ttvjyl/M2PFihVkZWXx6quv8uOPP3L88cfzyy+/5M0fOHAg2dnZ\nrFq1ipdffpljjjmGiy++mD59+uDc73Xdtm7dSq9evRg5ciRmlridIAlVVsdU9erVycrKYsOGDUyZ\nMoXFixdzyimn5B0jV1xxBdOnT2fq1KnMmTOHTZs2ceKJJ+bN/+yzz8jJyWHs2LEsX76cBx98kIkT\nJ3LllVeWzY6QlGLhk8nuwMzc7lbmRDEzcoanJ2XdaSM3sbfudxGRvZmZ4ZzTL4xSMLOJwHfOuatD\n0+4GfnDO3W1mNwC1nHMFxmTam+MeEZFoEv2bSL9zJExxT35m9gIw2Tn3TDmsy7Wc0LJM8t68bDN1\n5tdh/rvzyyR/kcJs27aN+vXrM2fOHO6//35q1arFqFGjAFi3bh1Nmzbl0ksv5f777y+w7KZNm0hP\nL/p6N3v2bHr06MHGjRupXbs2AB988AEdO3ZkxowZHHvssXTv3p1WrVoxZsyYfMvOmDGDv/zlL0yY\nMIHzzz8/37wFCxZw1FFHsWrVKho1alSaXSAJVlbH1JNPPsmQIUPYtGlT3rQpU6Zw7rnn8vnnn1On\nTh0yMjJ48sknOeusswBYu3YtjRs35vXXX+e4446Lmu+///1vbrnlFjZu3FjaTZcES3TMo5ZMIiIi\nImXEzDoC5wA9zGyRmS00s78AdwPHmdnnQE/grmSWU0RERCSK14F7zew2M+trZn3Cr2QXTiTVPfvs\ns2RmZnLooYfSv39/Jk6cyK5duwB45pln2LFjBzfccEPUZYt7wFSYKlWqALB9+/Yi0/35z3+mVatW\nPPfccyVajyRHeR5T4WNpwYIF7Ny5M9/DpIYNG3LIIYfw/vvvF5rHzz//TK1ateJar+yeKia7ACIi\nIiJ7KufcXAofFPvY8iyLiIiISJweDv4OizLPUXiMIyLA+PHjOe+88wDo2rUr1atX58UXX6RPnz6s\nXLmS9PR06tWrl7D1ff/994wcOZL09HSOOuqoYtO3aNGCTz/9NGHrl7JXXsfU2rVrGTVqFAcccADN\nmzfnk08+oUKFCuy333750tWtW5esrKyoeaxZs4Z7772Xm266qdTlkdSnlkwiIiIiIiIiIpKPcy6t\niFfMD5jMrIqZzQ9adX9qZsPLstwiqWDlypXMmTOHfv365U07++yzefzxxwES1lWrc47MzExq1qxJ\nnTp1+Pzzz5k2bRoZGRkxLauxl3YfiTqmatasSc2aNUlPT2fw4MF50zdv3kx6ejo1atSgcePG7Nq1\ni+eff56KFeNvo5Kdnc3xxx9Pr169uOKKK+JeXnY/askkIiIiIiIiIiJlwjm33cy6O+e2mlkFYK6Z\nveac+zDZZRMpK4899hg5OTkccMABBeatW7eOZs2asWnTJrKyskrV8sTMmDVrFrVq1SIjI4MaNWrE\nvOzy5cs58MADS7xuKV+JOqaWLFmS93+4C73q1auzZMkSzIy6deuyzz775M2rV68eu3bt4vvvv8/X\nmqkEAEkAACAASURBVCk7O5suXbrkyz8rK4uePXvSunVrJk6cWKJtld2PWjKJiIiIiIiIiEgBZlbL\nzM42s7+b2S3hVzz5OOe2Bv9WwVd4TkwzDpEUtGvXLiZOnMhdd93FkiVL8r1at27NhAkTOP3006lU\nqRJ33RV9aNaff/455vVlZmbSpEmTuB4wvfHGGyxdupQzzjgj5mUkeRJ5TB144IF5r/333z9vvpnR\npEkTMjMz8z1gAjj88MOpWLEiM2fOzJu2du1aVqxYQceOHfOmbdiwge7du3PooYcyefJk0tL06GFv\noZZMIiIiIiIiIiKSj5m1B14BtgMZwDqgfvB+NfDPOPJKAxYATYGHnHMfJbq8Iqniv//9L99//z0X\nXnghtWrVyjevb9++PPLII9x0003cf//9DBkyhJ9++okBAwbQpEkT1q9fz+TJk9m+fTtjx44tdl2x\ndJG2detWsrOz2blzJxs2bOCVV15h1KhRnHbaaZxzzjl56bKzs8nKyuLzzz/HOceyZcv48ccfadSo\nUYHtkPJVnsdUNOnp6QwaNIjrr7+ejIwMateuzTXXXEObNm3o2bMn4B8wde3alYYNG3LfffexcePG\nvOUzMjL0wGkPp09XREREREREREQijQImAX8EtgE9gEbAx8Dd8WTknMtxzrUFGgJHm1mLBJdVJGWM\nHz+eHj16RH0wc8YZZ7BmzRrefPNNLr30UmbOnMl3333H6aefzsEHH8yAAQPYtm0bf//732NaVyxj\nKk2YMIEGDRrQtGlTTj75ZObPn8+4ceN47rnn8i3/yCOP0LZtW/r374+ZceKJJ9KuXTtefvnl2Dde\nykR5HlOFGT16NKeddhpnnXUWnTt3Jj09nZdeeinvGJoxYwZffvkls2fPpnHjxjRo0ID69evToEED\n1q5dW6p1S+qzRA00V17MzO1uZU4UMyNneHrxCctA2shNCRuUUEREdh9mhnNOo8Emyd4c94iIRJPo\n30T6nSNhinvyM7OfgSOdc1+Y2U9AB+fcCjM7EpjsnDuohPneDGxxzt0XmuYyTsnIS1P94OrUODj2\nrr+KsnnZZtY/vJ7ftv6WkPyiqVClAru27yqz/Os3rM/6b9aXWf4iIrJnmzVrFrNmzcp7P3LkyITG\nPOouT0REREREREREIoWfymQDjYEVwGagQayZmNn+wA7n3M9mtg9wHFBg0JC6p9YtXWmL8NvW32g5\noWWZ5b90wNIyz19Sy6ZNm6hQoQLVq1dPdlFERIrVrVs3unXrlvd+5MiRCc1f3eWJiIiIiIiIiEik\nhcCRwf+zgNvM7HxgDPBJHPnUB94xs8XAfOAN59yriSyoSFlzzjFv3jz+8Y9/0LJlS2rVqsVFF11U\npuvs3bs3NWvWLPBKT0/nrrsKPKcV2S3ddtttnH/++SxevDjZRZFSKPeWTGb/z969h1lVlv8ff9+I\nCoMpagIqIamZh1S0Mi2tqfyGWWbaQc00LC2/pZjWL00rpCw1j6WZlefy1NdDZppiGp5TMTU8l4pH\nBM8IDCBw//5YGxvGmWFg9t5rZs/7dV37Yj1rrb32Z81qxtW+1/M8cSbwaWBaZm5WWfdzYCeKySMf\nA/bJzBn1ziZJkiRJkiQAjgDeVln+AXAecArwKLBPVw+SmZOBLaueTqqxzGTSpEmcd955XHjhhcyd\nO5c5c+Ywf/58AG644Yaafv6ZZ55JS0tLu9tWW221mn62VA/z5s3jmGOOoaWlhUsuuYR11lmH7373\nu+y+++40NTWVHU9LoYzh8s6muCk5r9W6CcBhmbkwIo4Bvl95SZIk1V1ErACMAbaiePp2KsWTt+dm\nZu0G9JckSeohMnNSq+UXgE+WGEeqmwcffJAzzjiD888/n9mzZ9PS0sKCBW+dc+uVV15h6tSprLnm\nmjXJUavjSj3FtddeS79+/Vi4cCGzZ8/moYceYuzYsRx44IHst99+jBs3jlVXXbXsmOqCug+Xl5m3\nAK+0Wfe3zFxYaf4DGF7vXJIkSQARsRHwb+BXwHuABZV/fwX8JyI2LjGeJElSKSLiwxHxyYgYXHYW\nqdoyk+uuu45tt92W973vfZxyyilMnz6dmTNntltgAujfvz8333xznZNKjeO3v/0tr7/++mLrZs2a\nxezZs/nNb37DiBEjOProo5kzZ05JCdVVPXFOpq8Cfy07hCRJ6rN+C7wGrJeZW2fmZzJza2B94FXg\n9FLTSZIk1VBEHBARR7RZ9xeKeZmuAh6sPJQj9Xpz587l7LPPZr311mPXXXfl1ltvpaWl5c0h8drT\n1NTEwIEDGThwICuttFId00qN5d577+1wWLw5c+Ywc+ZMjjrqKIYPH85ZZ53VYcFX5etRRabKTcwb\nmXlB2VkkSVKf9T7gR5n5VOuVlfY4/jsBtvqAkcPXJCKq9ho53GFPJEk93leAJxc1ImIXYDSwF8V9\n0hTgh6Ukk6pkwYIF/O53v2PYsGGMHTuWJ554gpkzZ3a4/6BBg1hxxRX54Ac/yAknnMDkyZOZPn06\nO+64Yx1TS43loYce4qSTTuJd73oXTU1N9Ov31lLF7Nmzeemllxg7dizrr78+V111FZlZQlp1pow5\nmdoVEWOAHYGPLWnfMWPGMHLkSAAGDx7MqFGjaG5uBmDixIkAjdueUjxJ0Tyyf13bi5R+/rZt27Zt\nu6btRctTpkyhD5sCDOhg2wDgqQ62qQE9+ezzLBy3ctWO12/881U7liRJNbIecE+r9o7AnzPzfICI\nOBw4t4xgUjXcdNNN7Lvvvjz33HPMmjWrw/2amppYsGABo0aN4mtf+xq77rorq6++eh2TSo1tpZVW\n4utf/zpf//rXmTRpEieddBKXX345b7zxxlt6E86aNYtZs2ax2267scUWW3DxxRez1lprlZRcbUUZ\nlb+IGAlcmZmbVto7ACcAH87Ml5bw3uyr1cqIqOqXHEuj3/gZVoklqQ+KCDIzys5RTxGxM8V9yZ6Z\neUer9VsDfwC+m5l/qlOWPnvf01NU+/7Leyqpe/ydVC31xfue9kTELGCTzJxSad8P/C4zf1FpjwAe\nycyBVfzMfM/Z76nW4RYz84GZTDl+CrU6PsD9+9xf8+P7t6r7nnjiCb71rW9x4403Mnv27A73W2ml\nlRg0aBCHHnooX/7yl1ljjTXqmFLq25566ikOOuggrr32WlpaWtrdp3///gwcOJBzzjmHXXfdtc4J\nG0O173nqPlxeRFwA3AZsEBFPRcQ+wCnASsB1EfHPiDit3rkkSVLfFRF3RcSdEXEncASwMnBbREyN\niPsiYipwa2X94WVmlSRJqrEnqQwPHBFDgI2AW1ptH0YxT6XUK2Qmxx9/PJtssgkTJkxot8AUETQ1\nNbH55ptz3nnn8eyzz3LwwQdbYJLqbMSIEVx++eVcf/31bLzxxgwaNOgt+8yfP5/XX3+dvfbaiy9/\n+cudDnWp+qj7cHmZ+aV2Vp9d7xySJEmtPABkm7YkSVJfdC7wq4jYFGgGHsrMu1tt/yAwuYxg0tKa\nPn06X/jCF5g0aVKHvSKamprYdttt+clPfsJWW21V54SS2rPNNtswefJkzj//fL797W/T0tLylt/h\n2bNnc+mll3LzzTfz97//nXXXXbektOoxczJJkiSVJTPHlJ1BkiSphzgOGATsBDwPfKPN9g8BF9c7\nlLS0rrnmGnbffXdmz57NG2+88ZbtgwYNYq211uKMM87gwx/+cAkJJXWmX79+7LXXXuy666789Kc/\n5eSTT37LfE1z5szhmWeeYcstt+Taa6/lAx/4QImJ+666D5cnSZIkSZKknikzF2bmjzJzi8z8ZGY+\n1Gb7FzLzzLLySUuyYMECxo4dy6677sprr732lgLTgAEDWGWVVTjppJN46KGHLDBJPdygQYP42c9+\nxiOPPMLo0aNpampabPvChQt57bXX+NjHPsbll19eUsq+zZ5MkiSpz4uIHy3N/pn541plkSRJkrRs\n5syZwy677MJNN93U7vB4TU1N7Lbbbpx88smsvPLKJSSUtKze8Y53cOWVV3LkkUdy3HHHtTt83p57\n7skxxxzD2LFjS0rZN1lkkiRJgiOBFmAWEEvYNwGLTJIkSVIP8sorr7D99tvz4IMPMmfOnMW29e/f\nn6amJs477zx23nnnkhJK6q6IYPz48ay33nrsv//+byk0tbS0cNhhh9HU1MS+++5bUsq+Z6mKTBEx\nCNgIWBOYSjH546xaBJMkSaqjx4B1gLuBi4DLMvP1ciNJkiRJ6opnnnmG7bbbjueee4558+Yttq2p\nqYlNN92USy+9lLXXXrukhJKqae+992bEiBHstNNOzJw5c7FtLS0tjB07llVXXZXPfe5zJSXsW7o0\nJ1NE9IuI8cCzwB3AFcCdwHMR8eOIWK6GGSVJkmoqM98FfBB4APgJMC0iLouIL0TEwHLTSZIkSerI\nf/7zH7bYYguefvrptxSYBg4cyHe/+11uu+02C0xSg2lubuaOO+5gyJAh9O+/eF+alpYW9tprL667\n7rqS0vUtXSoyAccDhwG/BN4LDAe2rLQPBX5ek3SSJEl1kpmTMvO7mTkC2AF4HjgVmB4R50eEMwJL\nkqSGFhELImJIZfmsiHhb2ZmkzkybNo3tttuOl156iQULFiy2beDAgZx++umMHz+efv26+hWopN5k\n44035r777mP99ddnwIABi21raWnhs5/9LPfdd19J6fqOrv6F/Qrww8z8UWbem5nPVf79IfAjYEzN\nEkqSJNVZZt6Umd8E3gGcDuwGfLvcVJIkSTXXAqxUWf4KMKCTfaVSvf7663zkIx/hxRdfJDMX29bU\n1MTll1/O3nvvXVI6SfUybNgwJk2axDbbbMPAgYsPRDJ79mw+9alPMWPGjJLS9Q1dnZMpgHs72HYv\nS54gW5IkqdeIiA8BuwOfB94GXAL8utRQkiRJtXcb8KeIuJviu55fRkRLeztm5lfrmkxqZd68eYwe\nPZopU6Ywf/78xbatssoqXH/99bz3ve8tKZ2kehs0aBBXXXUVm266KY8//vhihecXX3yRL3/5y1xx\nxRVEWMaoha72ZLqQjnsrjQEurkYYSZKkskTElhHx84h4ErieohfTwcCQzNw9M28sN6EkSVLN7QVc\nCwwGElgdWKODV5dExPCIuCEiHoiIyRExtvqx1ZcsXLiQ3XbbjXvvvZe5c+cutu1tb3sbt99+uwUm\nqQ8aOHAgV199NU1NTYutnzt3Ltdffz2nnnpqSckaX4c9mSKi9RMp9wNHRMQ/gSuA6cAQ4LPA24Gj\naxlSkiSpliLiEeCdwA3AOOCyzLQ/vSRJ6lMycxrw/wAi4glgj8x8qZuHnQ8ckpn3RsRKwN0RMSEz\nH+7mcdVHHXfccUyYMIGWlsU72TU1NXHttdey0UYblZRMUtk22GADzjjjDL72ta8xe/bsN9fPnj2b\nQw89lI985CNsttlmJSZsTJ0Nl3dGO+vWAka1s/4U4LSqJJIkSaq/dwFzgPcCWwI/76wbfWYOqVMu\nSZKkUmTmO6t0nOeB5yvLMyPiIWBtwCKTltqkSZMYP378WwpMAwcO5KKLLmKbbbYpKZmknmL33Xfn\n+uuv5/zzz1/sb0VLSwt77bUX99xzD/36dXWAN3VFZ0WmgZ1skyRJaiTjyw4gSZLU00TEp4BDgY0p\nhs97EDg2M69exuONpHh4+Y4qRVQfMmPGDD7zmc+024Pp5JNPZqeddiopmaSe5tRTT+XWW2/lkUce\nYeHChW+uf+yxxzjrrLPYd999S0zXeDosMmXm3I62SZIkNZLMtMgkSZLUSkTsSzFqzfnAuZXV2wGX\nR8T/ZuZZS3m8lYBLgIMyc2bb7dP+NO3N5UEbDmKlDVda1uhqQJnJPvvsw8svv7zY+oEDB7LPPvuw\n3377lZRMUk+04oorctVVV7HZZpsxc+Z//5Mza9YsDjnkEHbZZRdWX331EhPW18SJE5k4cWLNjt/Z\nnExLNQxMZk7vfhxJkiRJkiT1AIdSzKXUeqb0MyPibuAwoMtFpojoT1Fg+n1mXtHePkM/O7Q7WdXg\nzj33XK655hrmzv3vM/ERwTrrrMOJJ55YYjJJPdU73/lO/vCHP7DHHnss1gNy7ty5HHTQQfzhD38o\nMV19NTc309zc/GZ7/PjqPmfb2eCDU5fyJUmSJEmSpMYwArimnfV/BdZZymOdBTyYmb/odio1vLFj\nx7L//vvz6quvAjBt2jQOPPBAZs+evdh+TU1NXHnllaywwgplxJTUC+y888588YtfXOzvxLx587js\nssuYPHkyAG+88QbnnHOOBetu6GxOpgBmAlcAVwPz6pJIkiRJkiRJZXsK+B/gP23WfwJ4sqsHiYgP\nAXsCkyPiHoq5nQ7PzPYKWBJ///vfefjhh7nwwgs5/vjjufrqqxfrwQRFgem0005j/fXXLymlpN7i\n2GOP5Y9//ONi6+bMmcO3v/1tPv/5zzNu3Dhee+01NtpoIw455JCSUvZunRWZ/gfYA9gF+DRwOXAR\n8LfMXNjJ+yRJkiRJktS7HQ+cEhFbArdV1n0I2As4sKsHycxbgeWqH0+NasaMGcyfP58ZM2Zw8MEH\ns2DBAt544403t6+wwgrssMMO7L333iWmlNRbDB06lO985zuccMIJbw6bl5nceuut3HHHHcyaNQuA\n/v07K5WoMx0Ol5eZ12fmvsAwYG9gRYrxc6dGxGkR8eE6ZZQkSaqpiNiw7AySJEk9SWb+BtgN2Iii\n4HQ8sCHwxcz8bZnZ1Nhef/31N5dnzZrFnDlzFtu+4oorcvrpp9c7lqRe7NBDD33L0Jpz5859s8AE\nsPzyy9c7VsNYYnkuM98ArgSujIiBFL2a9gCuj4iLMnOvGmeUJEmqtQci4gngqsprYmY6VLAkSerT\nMvNyipFtpLqZN6/j2/AVV1yR7bffnsGDB9cxkaTe7JVXXuHEE09crEdkewYMGFCnRI2nw55MHVgX\n2AzYlGIM3aernkiSJKn+1gSOAtYC/gi8FBF/ioj9ImLtcqNJkiRJgqLnwYQJE1hvvfW47rrryo4j\nqQfLTI444giGDx/O8ccfz+zZszvdf/jw4XVK1niWWGSKiHUj4vCImAzcB2wHnACslZmH1zqgJElS\nrWXm9Mw8JzO/ALwd2JlikutDgKci4p6IOCoitomIKDWsJEmS1MCWdLs9a9Ysnn76aUaPHs1xxx1X\np1SSeqN//vOfZOZbht1sz8iRI2sfqEF1WGSKiEMi4k7g38BngDOBEZnZnJmnZ+aL9QopSZJUL5k5\nPzNvyMzvZuZGwAbAWcD7gBuA6RFxXqkhJUmSpD6sqamJ0aNHs+eee5YdRVIPFRFcffXVHH300Qwc\nOLDTfQcOHMiaa65Zp2SNp7M5mY4HXgcuBB4HVgP27+BpgszMcdWPJ0mSVK7MfAw4BTglIpqA/wF2\nLDeVJEmS1PessMIKrLzyypxzzjl86lOfKjuOpB4uIjjooIPYbrvt+PjHP85rr71GZr5lv+WXX55h\nw4aVkLAxdFZkmk4x79LHK6/OJGCRSZIkNbTMnA1cUXlJkiRJqrKOhstramriAx/4AJdccgmrrbZa\nnVNJ6s223HJLJk2axHbbbccLL7zA/PnzF9seERaZuqHD4fIyc1hmrtnF11r1DC1JkiRJkqTaiohT\nI8Jv81VX7RWZBg4cyOGHH87f/vY3C0ySlsl6663HPffcw7rrrsuKK6642Lb58+czdOjQkpL1fh0W\nmSRJkiRJktS3RMTwVs0vAStV1k+OiHeUk0p9yfLLL7/Y8qqrrsq1117LEUccQb9+fpUpadkNHTqU\nu+66i80335wBAwa8uX7u3LnOydQN/mWWJEmSJEnSIg9HxJMRcQEwAFhUWBoJLN/hu6QqGTx4MFD0\nXtp88815+OGH2W677UpOJalRrLzyytx4441su+22bxaaVl99dZqamkpO1ntZZJIkSZIkSdIig4Ev\nAHdTfG90dUQ8CqwIjI4IxxNSTa266qr079+fj370o9xyyy0MGTKk7EiSGsyAAQO48sor2XjjjenX\nrx8bb7xx2ZF6NYtMkiRJrUTERhGxdav2wIj4WUT8KSIOLDObJElSHSyfmXdm5glAC7AFsA+wAPgq\n8EREPFJmQDW2Lbfckn322Yc///nPb5k3RZKqZcCAAVx//fWss846bLPNNmXH6dX6lx1AkiSphzkN\nuA34R6V9HMUXKzcDx0bEgMw8rqxwkiRJNfZqRNwL3AqsAAzMzFsjYj6wG/As8P4yA6qx/eY3vyk7\ngqQ+YvDgwUyePJmIKDtKr9ZhT6aImBgR34mIjeoZSJIkqWTvAW4HiIjlgb2Ab2fmDsDhFE/wSpIk\nNaq1gaOAuRQPJ98dETdTFJy2BDIzbykxnyRJVTNo0CDnY+qmzobLu4FiDN77I+LxiDglInaICPup\nSpKkRjYImFFZ3rrSvqzS/iewThmhJEmS6iEzX8zMKzPz+8Bsil5LpwAJHA+8FhE3lplRkiT1HB0W\nmTLzx5m5NTAMGA8MAS4EXoqIP0fENyJi+NJ+YEScGRHTIuJfrdatGhETIuKRiLg2IlZZhnORJEmq\nhicoiksAuwD3ZOZLlfbbgddLSSVJklSO1zLzj8AbwMeAd1IMLyxJktRpTyYAMvOFzDw3M3ej+GLl\n08DDwFjgyYi4rzIZ9ge7+JlnA6PbrDsM+FtmvpuiB9X3u3wGkiRJ1XUicFRE3EVxv/PLVtuagX+1\n9yZJkqQGtBnwTGX5SeCNzHw+My/u6gHae9hYkiQ1jiUWmVrLzAWZOTEzv5eZmwDrA78FRgHXdfEY\ntwCvtFm9M3BuZflc4LNLk0uSJKlaMvNMYHvgImB0Zv6+1eaXgZNLCSZJklRnmfl0Zi6sLL8nM59e\nhsO097CxJElqEP278+bMfAL4FfCriBjYjUMNycxplWM+HxFDupNLkiSpOzLzJuCmdtYfWf80kiRJ\nvVdm3hIRzmkpSVKD6laRqbXMbKnWsSgmk5QkSSpNRGwADAcGtN2WmVfXP5EkSZIkSVLPUrUiUzdN\ni4ihmTktIoYB0zvbecyYMYwcORKAwYMHM2rUKJqbmwGYOHEiQOO2p8wv2iP717W9SOnnb9u2bdu2\na9petDxlyhT6qojYmGKovE2AaGeXBJarayhJkiRJkqQeKDLr32koIkYCV2bmppX2scDLmXlsRBwK\nrJqZh3Xw3iwjc08QESwct3Ipn91v/Az66s9dkvqyiCAz2yu0NKyIuBkYAnwPeBCY13afzHyyTln6\n7H1PT1Ht+y/vqaTu8XdStdQX73vqpTJc3pWZuVkH23ONndd4sz1ow0GstOFKVfnsmQ/MZMrxU3jP\n2e+pyvHac/8+99f8+L35b9Va71iLqc9Mrdnx1xy+Js89/VzNji+pd/Jvz39NnDhxsYeLx48fX9V7\nnrr3ZIqIC4BmYPWIeAoYBxwD/F9EfBV4EvhivXNJkiRVbAHsnpl/KTuIJElSgwja7yH+pqGfHVqn\nKKq3qc9MrXkRTpLa8m/PfzU3N785kg0URaZqWuYiU0SsC2wA3J2ZL3T1fZn5pQ42bb+sWSRJkqro\nMdqZh2lZRMSZwKeBaYue3I2IccB+/Hd44MMz85pqfJ4kSVJP097Dxpl5drmpJElStXSpyBQRp1AM\nrXdApb0LcHHl/a9FxCcy867axZQkSaqb7wA/j4h/Zubj3TzW2cApwHlt1p+YmSd289iSJEk9XicP\nG0uSpAbQr4v77QTc3qr9M+BSYF3gxkpbkiSpERwNrA08HBGPRsSdbV9dPVBm3gK80s4m53uQJEmS\nJEm9XleHyxsKPAUQEesB7wZ2y8wpEXEacGGN8kmSJNXb/ZVXLR0QEXsBk4DvZOZrNf48SZIkSZKk\nqutqkekVYI3K8vbA9Mz8V6WdwPLVDiZJklSGzNynxh9xGvDjzMyIOAo4EfhajT9TkiRJkiSp6rpa\nZJoAHBkRqwLfAy5ptW0TYEqVc0mSJJUuIlYHVgNezsyXqnHMzHyhVfN3wJWd7T9mzBhGjhwJwODB\ngxk1ahTNzc0ATJw4EcB2jduLTJwyv9g+sn+32m8er4ecn23bvam9iL+PtqvRXrQ8ZcoUJEmStGwi\nM5e8U8RqwKnA+4F7gf0y89XKtn8At2XmIbUM2ipLdiVzI4oIFo5buZTP7jd+Bn315y5JfVlEkJl9\nbv6giNgNOBLYoNXqR4EfZeb/LeWxRgJXZuamlfawzHy+snww8P6OJsTuy/c9PUW177+8p5K6x99J\n1VJfve/pCSIi33P2e2py7JkPzGTK8VOo1fEB7t/n/pofvzf/rYoIfz6S6s6/PR2r9j1Pl3oyZebL\nQLtffmTm1tUKI0mSVLaI2AM4H/grcDQwjWJ+yt2AiyJiucy8qIvHugBoBlaPiKeAccBHI2IUsJCi\nN/g3qn0OkiRJkiRJ9dDV4fIkSZL6iiOA32bm/m3WnxcRpwM/ALpUZOqgh9LZ3cwnSZIkSZLUI3Sp\nyBQRywH/C+wKDAcGtN0nM0dUN5okSVIp1gcO7mDbpcCY+kWRJEmSJEnqubrak+kE4EBgAsXk1PNq\nlkiSJKlc04D3Ade1s+19le2SJEmSJEl9XleLTLsDh2fmsbUMI0mS1AOcDRxZ6cl9CUVRaQjwBYqh\n8o4uMZskSZIkSVKP0dUiU3/g7loGkSRJ6iF+DCwPHAaMb7W+BTi+sl2SJEmSJKnP62qR6SyKp3f/\nVsMskiRJpcvMhcAREXE88B5gTWAqcH9mvlJqOEmSJEmSpB6kq0Wmx4DvR8RfKeYneLXtDpl5VjWD\nSZIklalSULq57BySJEmSJEk9VVeLTL+u/DsCGN3O9qTo7SRJktTrRMSOwC2ZOaOy3KnMvLoOsSRJ\nkiRJknq0rhaZBtY0hSRJUrn+AmwN3FlZTiA62DeB5eqUS5IkSZIkqcfqUpEpM+fWOogkSVKJ3kkx\n79KiZUmSJEmSJC1BV3syERFvA74KbAusBrxMMU/BWZk5szbxJEmSai8zn2zdBKZm5htt94uIPrU+\nwQAAIABJREFU/sBadQsmSZLUACJiB+BkoB9wZmYeW3IkSZJUJf26slNEjAT+BRwPrA3MqPx7AvCv\niFinRvkkSZLq7Qlgiw62bV7ZLkmSpC6IiH7AqRRzfG8C7BERG5abqlwzH/ZZ7UY2ceLEsiPUTV86\nV/B8G51/m5ddl4pMwInAXGCDzPxgZu6SmR8ENgDmVLZLkiQ1go7mYgIYQHFPJEmSpK7ZCvh3Zj5Z\n6Sl+EbBzyZlKNevhWWVHUA31pS/m+9K5gufb6PzbvOy6Olzex4GvZuZiT+5m5hMRMQ44o+rJJEmS\n6iQiNgNGtVq1YztP2A4Avgg8WrdgkiRJvd/awNOt2s9QFJ4kSVID6GqRKSjmJ2jPQjp/4leSJKmn\n2wUYV1lO4Ecd7PcE8I26JJIkSepDZtw9oybHbXm6pSbHlSRJhcjsqHbUaqeIK4GRwOjMfK7V+jWB\na4EpmfmZWoVskyW7krkRRQQLx61cymf3Gz+Dvvpzl6S+LCLIzIZ/mCQilgdWoHhwZgbwMeCuNrvN\nqwzxUs9cffa+p6eo9v2X91RS9/g7qVrqK/c99RYRWwNHZuYOlfZhQGbmsa328RdRkqQ6quY9T1d7\nMh0M/B14PCL+AUwDhgDbANMr2yVJknqlSvFoUQGpq3NWSpIkacnuAtaPiHWAqcDuwB6td7C4J0lS\n79WlIlNm/ici3gXsD7wfWBN4Dvge8LvMtO+xJElqKBExHNiAYi6mxWTm1fVP1POMHL4mTz77fNWO\nt87aw5jyzNSqHU+SJJUvMxdExAHABIqHec7MzIdKjiVJkqqkqz2ZyMw5wMk1zCJJklS6iHgb8Efg\nE4tWVf5tPYzLcnUN1UM9+ezzVR62qnoFK0mS1HNk5jXAu8vOIUmSqs/hYCRJkhZ3NDAC2I6iwLQL\n0AycCTwBbF1aMkmSpAYTETtExMMR8WhEHFp2nlqLiCkRcV9E3BMRd5adp9oi4syImBYR/2q1btWI\nmBARj0TEtRGxSpkZq6WDcx0XEc9ExD8rrx3KzFhNETE8Im6IiAciYnJEjK2sb9Tr2/Z8D6ysb7hr\nHBErRsQdlb9LkyNiXGV9o17bjs634a7tIhHRr3JOf660q3ptOywyRcRTEbF5ZfnpSrvDV3dCSJIk\n9SA7Aj8F7qi0n8vMmzLz68AVwP8rLZkkSVIDiYh+wKnAaGATYI+I2LDcVDW3EGjOzC0yc6uyw9TA\n2RTXs7XDgL9l5ruBG4Dv1z1VbbR3rgAnZuaWldc19Q5VQ/OBQzJzE2Ab4FuV39dGvb5tz/eAVn+f\nGuoaZ+Zc4KOZuQUwCvhkRGxFg17bTs4XGuzatnIQ8GCrdlWvbWfD5Z0PvNhqOTvZV5IkqVEMBZ6u\nzB8wC1it1bargUvLiSVJktRwtgL+nZlPAkTERcDOwMOlpqqtoIFHFsrMWyJinTardwY+Ulk+F5hI\n8QVnr9bBucJ/h9tuKJn5PPB8ZXlmRDwEDKdxr29757t2ZXPDXePMnF1ZXJGiZpA06LWFDs8XGvDa\nVuabXvQw7SGV1VW9th0WmTLz+62WG+J/PJIkSV3wNEWhCeDfwKeBayvtDwBzygglSZLUgNamuPda\n5BmKwlMjS+C6iFgA/DYzf1d2oDoYkpnToPjiPiKGlB2oxg6IiL2AScB3MvO1sgNVW0SMpOgB8g9g\naKNf31bnewewLQ14jSs9S+8G1gN+lZl3RUTDXtsOzndHGvDaAidRjMjSeki8ql7bLj05ERFXR8QG\nHWxbPyKu7k4ISZKkHuQ64OOV5ZMohoG4LSL+DvwEOK+0ZJIkSertPpSZW1I8Vf6tiNi27EAlaOTR\nkk4D1s3MURS9YE4sOU/VRcRKwCXAQZk5k7dez4a6vu2cb0Ne48xcWBk+bjiwVURsQgNf23bOd2Ma\n8NpGxKeAaZl5L5330urWte1q99wdgMEdbBsMfKI7ISRJknqQQ4EfAmTm74HPAU8ArwAH0CDDA0iS\nJPUAzwIjWrWHV9Y1rMycWvn3BeByGr/nFsC0iBgKEBHDgOkl56mZzHwhMxd9Wfs74P1l5qm2iOhP\nUXD5fWZeUVndsNe3vfNt9GucmTMohk7bgQa+tou0Pt8GvbYfAj4TEY8DFwIfi4jfA89X89ouzRiw\nHVWztu1uCEmSpJ4iM2dn5out2pdn5p6ZuWtm/jozF5aZT5IkqYHcBawfEetExArA7sCfS85UMxHR\nVOkVQUQMonho+/5yU9VEsPgT838GxlSWvwJc0fYNvdhi51r5snaRXWm863sW8GBm/qLVuka+vm85\n30a8xhHx9ohYpbI8EPgf4CEa9Np2cL4PN+K1zczDM3NEZq5L8d/YGzJzL+BKqnhtO5yTKSKOAI5Y\nlAe4KSLaFpqWpyhU/YIqiIiDga8BC4HJwD6ZOa8ax5YkSeqKyvj422Tmne1sey9wZ2YuV/9kkiRJ\njSUzF0TEAcAEiu+XzszMh0qOVUtDgcsr36/1B87PzAklZ6qqiLgAaAZWj4ingHHAMcD/RcRXgSeB\nL5aXsHo6ONePRsQoiu82pwDfKC1glUXEh4A9gckRcQ/F98WHA8cCf2zA69vR+X6pAa/xmsC5lXmK\n+gEXZ+bVEfEPGvDa0vH5nteA17Yjx1DFa9thkQm4gWJi6wB+DvwWeKrNPvOAh4G/dScEQESsBRwI\nbJiZ8yLiYorqmvMeSJKkeupsnOLlgfn1CiJJktToMvMa4N1l56iHzHwCGFV2jlrKzC91sGn7ugap\ngw7O9ey6B6mTzLwV6Ohhu0a8vh2d7zX1zlJrmTkZ2LKd9S/TmNe2o/Pdu4Q4dZOZNwI3Vparem07\nLDJl5u3A7QAR8TpwWWW82FpaDhgUEQuBJuC5Gn+eJEkSETECGNlq1RYRMaDNbgMoupE/Ua9ckiRJ\nkiRJPVlnPZnelJm/qXWQzHwuIk6g6C01G5iQmd3uISVJktQF+1AMb5GV16872K8F2LdeoSRJkiT1\nXRHxDeCozFyj7CzLKiLeTTG/z3sy88Gy81RTRFwILJeZjTKMnLRMulRkAoiInYH9gA0onuRdTGaO\n6E6QiBgM7AysA7wGXBIRX8rMC9ruO2bMGEaOHAnA4MGDGTVqFM3NzQBMnDgRoHHbU4oReppH9q9r\ne5HSz9+2bdu2bde0vWh5ypQp9DGnAZdQDJX3L4qxt//VZp95wFOZObfO2SRJkiSVoDLaUtL+kNoJ\nnJuZX+3CcbpTjMglHPt5YAjwhcy8tM22+4BNgQMy87SufmBETAV+sjTvqbzvduDmzPxeq9WPAsOA\nF5fmWLUSEY8D+7c3H1pEXEQxN82vMvPANtt+QTHVyyUWlaTFRWanf6eKnSI+D1xQee0N/IGiQLUj\nxR+IP2bm4d0KUnzG6Mzcr9LeC/hAZh7QZr/sSuZGFBEsHLdyKZ/db/wM+urPXZL6soggMzubo6jh\nRMQ6wNTMnNcDsvTo+55q35v0xPuNvnCOUm/i76RqqS/e90jqXEQMadXciWLO+mH8t+jUkpmvd+E4\ny1RkqvRk+klmDulkn6nAXOCBzPxUq/XvBa4HFgI/KLHI1GNExOYUc9K8PTPfMtdu5TptA6wErLXo\n/xNGxAoU07q8Dty16Drak0kq9Ovift8Hfgp8rdI+uTK53TuBV4FpVcjyFLB1RAyIiAA+TtGVUpIk\nqZ4GAetHxMYdvcoOKEmSJKn2MnP6ohfFd6Bk5gut1r8OEBFbRMTfI2J2RLwYEb+LiEGVbUcDuwGf\ni4iFEbEgIraqbDshIh6tvO/xiDgqIro88lQrfwA+FhFrtlr3NeBCYE7rHSNiakR8s8262yPi54uW\ngaHAKZW8syvrh0TERRHxTETMiojJEfGlVse4EPgA8J1W5zkkIt5daW/cat+PR8SdETEnIp6LiGMj\nYrk2eU6MiJ9HxEuVzD9rk3m3SobZlX2uj4hVl/Bz+gzw1/YKTK3cCTwP7NJq3WeB6cDt7b0hIo6M\niGkRMSMifhMRyy8hh9RQulpk2gC4iaLyvQB4G0BmvgIcDXy7u0Ey806KYWruAe6jeCLgt909riRJ\n0lK6H5i8hJckSZIkERFvA66hKEy8F/g88DHg9MouRwFXAH+hKN6sCdxd2fYq8GVgQ4qh2MYA/28Z\nYjwHTKi8n4gYAOwBnEn7Q/11ZkfgBYpOB8MopjYBGEhRZPkksAnFkOPnRMQHK9u/AfyTYn7bYcCa\nleIctBryLyJGUvwsbgM2A/anmCP3yDY59qH4+WwFfAf4XmU6FyJiBEVh7TSKn92HgYu6cG6fpbgW\nS3Im/+1sAfBV4Cza/1nuAKwHNFMMtfcZ4Cdd+AypYXS1Mv46sGJmZkQ8B7ybomshFEWnqkw+l5nj\ngfHVOJYkSdIy+mg761YFRldeY+sbR5IkSVIPNoai+PCVyvBqD1V6Cl0dEYdl5rMRMYdiWLUXWr8x\nM1sXI56KiOMoihtHL0OOs4GfV977eeDpzJxUDBjVdZn5SmUuqpmtikRk5pPAL1rt+uuI+ASwO3Bb\nZs6IiDeA2a3fV9E6xIHAfzJzUaeFRyPih8AJEXFkZi6orP9nZi7qvfRYROxPMfLVFcDalWNe2uqz\nHujsvCLiHRTFsauW+EMoClg/i4jhFJ00PkIxhcx729m3Bfhaq2v/A+AXEfGDJfSYkhpGV4tMd1NU\nlq+hqDT/MCJaKCbA/jFwV23iSZIk1Vdm3tjBpj9FxFEUT6f9pY6RJEmSJPVcGwL3tJnT9RaKIshG\nwLMdvTEi9gAOANalmAeoP8X8SsviLxSFnw9T9Lw5YxmP067KcHY/AD5HUeRZofL661IeakOKXkyt\n3ULRU+qdwH8q6/7VZp/ngEVzU90F3ExRoJoAXEdRcHq5k8/9DHBjV+bQyswXIuIqip8jwDWZOb2D\ngl3ba3975VxGtjoXqaF1dbi8nwNTK8s/BB4EzuW/43ruX/1okiRJPc7fgZ3LDiFJkiSpV8iONkRE\nM3Ae8CfgU8Aoiof5V1imDyp6zfyBYpSorSvL7VnIW4d968ocQj8A/hf4KcXQcJtTFJiWKW87gsV/\nXm+02Z5UvsvOzPmZ+VGKofseqOT6d0S8u5Pjd3WovEXOpuil9hWK4fMkdaBLRabMvDkzf19Zfjkz\nRwOrAMMyc/PMfKSWISVJknqIT1GZ8FeSJEmSgIeALSJixVbrtqMo5jxcac8Dlmvzvg9SDBt3XGb+\nMzMfo+jJ0x1nUsxPdGUnvXpeoJgXCoCIGAS8q80+7eX9EHB5Zl6cmZOBJ4ANuvA+WLx49BDFube2\nHTAbmNJB5nZl5u2V6VfeC7wCfKG9/SJiFYqfy9IUmRYV0AYAV3ey36iIaF2k24ZiCL0pS/FZUq/W\npeHyIuI04NjK2JsAVLoWvl4Zz/L7mfnNGmWUJEmqm4j4YzurV6AY1uFdwOH1TSRJkiSpBzuXopfP\nOZXhtYcApwIXZOaiofKmAHtGxPoUxZBXgUeBd0bEFyimKvk0sGt3gmTmwxHxdoqCTUduAL4SEX8F\nXgN+xFt7XE0BPhIRlwJzKgWrR4EdI2LryvsOBtZi8WLKFGDriBgBzMrMlyrrW/ecOgX4ZkT8AjiN\n4v9n/Rg4sdV8TJ2KiG2BbSmGyZsObAUMo+N5mXYE/tXqeixRZi6MiA0XLXey60DgjIg4hmKIvJ8A\nv3I+JvUlXR0u7xvA0A62DalslyRJagRrtPNakWLM750y89gSs0mSJEnqQSoP4o+m+O70TuCPwPUU\nQ7gt8muKnj/3UBRF3puZl1AUXH5VWf9B4MhlidAmzyuZObej7RQFndso5nC6CriGYmqU1o6g6KX0\nOPBMZd04inmSJlAUqqYB/9fmfcdQ9GR6CJgeEYvmUHozQ6UTw6coevzcS/GzOZNimL+OMrf1KsWQ\nfVcBj1AM4Xd4Zl7ewf5LO1TeoqwzM3PmEna7FngSuAm4CLiSYroZqc+IzCX9zkJELAQ+kJl3tbPt\nS8BJmdlREaqqIiK7krkRRQQLx61cymf3Gz+Dvvpzl6S+LCLIzHZnN1Xt9fT7nmrfm/TE+42+cI5S\nb+LvpGrJ+x5JajyVoexeBD6UmfeXnUdqRB0OlxcR3wK+VWkmcHFEzGmz2wDgHcAFtYknSZIkSZKk\n3iwiDgL2rTR/l5m/LDOPpD7l7cDPLDBJtdPZnEyPU3TthGJszEkU3SBbm0cxgd351Y8mSZJUHxHx\no6XYPTPzJzULI0mS1EAiYhPga8D7gPnAXyPiL5n5eLnJJPUFmTkVcMhzqYY6LDJl5l+BvwJExEyK\nCcue6Wh/SZKkXuxIoAWYxeKT0rYnKSZzlSRJ0pJtBNyxaI6YiLgJ2BU4vtRUkiSpKjrryfSmzPx+\nrYNIkiSV6DFgHeBuislaL6tM4CtJkqTuuR84KiJWBeYCOwJvmfNbkiT1Tl0qMlWeMulUZn64+3Ek\nSZLqLzPfFRHvA3an6KX064i4BrgQ+EtmtpQaUJIkqZfKzIcj4ljgOmAmcA+woNxUkiSpWrpUZAKe\noxgaprVVga2A14FbqxlKkiSp3jJzEsUclN+NiA9TFJxOBc6KiD8Dv8nMJT54I0mSpMVl5tnA2QAR\n8VPg6dbbI6Ltd06SJKmGMnNJUwV0Wb8ufuDumblHm9cOwDuBJ4FrqhVIkiSpbJl5U2Z+E3gHcDqw\nG/DtclNJkiT1ThGxRuXfEcAuwAVt98nMhnqNGzeu9Ayel+fUSC/PqXe8PKfe8aq2rvZkaldmvhYR\nxwMnAedVJ5IkSVK5IuJDFD2ZPg+8DbgE+HWpoSRJknqvSyNiNeAN4JuZOaPsQJIkqTq6VWSqaAJW\nr8JxJEmSShMRW1IUlnYDhlL01D4Y+HNmzi4zmyRJUm+WzuMtSVLD6lKRKSI+1s7qFYCNgO8BN1cz\nlCRJUj1FxCMUwwDfAIwDLvMJW0mSJC2r5ubmsiPURCOel+fUO3hOvYPn1DdFV8bgi4iFQAJtJ4NK\n4CrgG5k5tfrx2s2StRg3sDeICBaOW7mUz+43fkZNxmuUJPVsEUFWcTLInqpyrzMHmEVxf9OpzBxS\n81D0/Pueat+b9MT7jb5wjlJv4u+kaqmv3Pf0RD39nkeSpEZS7Xuerg6Xt1E76+YAz2XmG9UKI0mS\nVJLxZQeQJEmSJEnqbbpUZMrMR2odRJIkqSyZaZFJkiRJkiRpKXVYZIqIpRoGJjOndz+OJEmSJEmS\nJEmSeoPOejIt7RxLy3UniCRJkiRJkiRJknqPzopMAcwErgCuBubVJZEkSZIkSZIkSZJ6vM6KTP8D\n7AHsAnwauBy4CPhbZi6sQzZJkiRJkiRJkiT1UP062pCZ12fmvsAwYG9gReASYGpEnBYRH65TRkmS\nJEmSJPVSEXFwRNwfEf+KiPMjYoWyM0mSpOrosMi0SGa+kZlXZuaewFDgAIrC0/UR8ftaB5QkSeqt\nIuLMiJgWEf9qtW7ViJgQEY9ExLURsUqZGSVJkmopItYCDgS2zMzNKEbV2b3cVJIkqVqWWGRqY11g\nM2BTIIGnq55IkiSph4qI30XEmUvxlrOB0W3WHUYx/PC7gRuA71crnyRJUg+1HDAoIvoDTcBzJeeR\nJElVssQiU0SsGxGHR8Rk4D5gO+AEYK3MPLzWASVJknqQj1ZeXZKZtwCvtFm9M3BuZflc4LPViSZJ\nktTzZOZzFN8jPQU8C7yamX8rN5UkSaqWDotMEXFIRNwJ/Bv4DHAmMCIzmzPz9Mx8sV4hJUmSeoLM\nXD8z1+3mYYZk5rTK8Z4HhnQ/mSRJUs8UEYMpHrJZB1gLWCkivlRuKkmSVC39O9l2PPA6cCHwOLAa\nsH9EtLdvZua46seTJElqeFl2AEmSpBraHng8M18GiIjLgA8CF7Te6cgjj3xzubm5mebm5vollCSp\ngU2cOJGJEyfW7PidFZmmU3zp8fHKqzMJWGSSJEkNIyJWBNYGBrTdlpkPduPQ0yJiaGZOi4hhFPdc\nHRozZgwjR44EYPDgwYwaNerNL10W3SSW1QaYOGU+zSP7v7kMLHN70Wf0lPNrexPe3fNb1G59rj3h\n/Gzb7k3tRfx9tF2N9qLlKVOmoJp6Ctg6IgYAcym+Y7qr7U6ti0ySJKl6mts8vDF+/PiqHj8ye9fD\nsxGRvS1ztUQEC8etXMpn9xs/g776c5ekviwiyMx2uzE3qohYC/gt8Mn2NlP04F5uKY43ErgyMzet\ntI8FXs7MYyPiUGDVzDysg/f26Pueat+b9MT7jb5wjlJv4u+kaqkv3vfUS0SMA3YH3gDuAfbNzDda\nbe/R9zySJDWSat/zdNaTSZIkqS86A9gSOAR4EJi3rAeKiAuAZmD1iHiKouf3McD/RcRXgSeBL3Y3\nsCRJUk+WmeOB6j42LUmSeoQeVWSKiFUovth5D7AQ+Gpm3lFuKkmS1Md8CNgvM//Y3QNlZkeTWm/f\n3WNLkiRJkiSVrUcVmYBfAFdn5hcioj/QVHYgSZLU50wHWsoOIUmSJEmS1NP1KzvAIhGxMrBdZp4N\nkJnzM3NGybEkSVLf8yPg0Mq9iSRJkiRJkjrQk3oyvRN4MSLOBjYHJgEHZaZPEkuSpHraFRgBPBkR\ndwGvttmemblb/WNJkiRJkiT1LD2mJxNFwWtL4FeZuSUwGzis3EiSJKkPejvwGHAvsDywRpvXkPKi\nSZIkSZIk9Rwd9mSKiK8uzYEy86xuZnkGeDozJ1XalwCHtrfjmDFjGDlyJACDBw9m1KhRNDc3AzBx\n4kSAxm1PmV+0R/ava3uR0s/ftm3btm3XtL1oecqUKfRVmfnRsjNIkiRJkiT1BpGZ7W+IWLgUx8nM\nXK7bYSJuBPbLzEcjYhzQlJmHttknO8rc6CKChePKmR6i3/gZ9NWfuyT1ZRFBZkbZOeopInYCrsrM\npbkXqlWWHn3fU+17k554v9EXzlHqTfydVC31xfuenqKn3/NIktRIqn3P09mcTAOr9SFLYSxwfkQs\nDzwO7FNCBkmS1Lf9CZgWEb8HzsnMh8oOJEmSJEmS1BN1WGTKzLn1DFL5zPuA99f7cyVJklpZj+JB\nl72B70bEncBZwMWZOaPUZJIkSb1MRGwAXAwkEMC6wA8z85elBpMkSVXRWU+mt4iINYB3AQPabsvM\nG6oVSpIkqSyZOQUYB4yLiI9RFJxOAk6OiMuAszLz7yVGlCRJqrnKd0Bk5guV9qbAbsADmXlhV4+T\nmY8CW1SO0Y9iTu7Lqx5YkiSVoktFpogYBPwB2IniqZP2dHtOJkmSpJ6k8hDNDRGxFnARsCfwpYh4\nEjgFOCUz55eZUZIkqUb+CPweOCsi3g7cBDwHHBgRa2XmCctwzO2BxzLz6SrmlCRJJepqT6afAhsB\nnwD+BuwOvAp8GfgQsFdN0kmSJJUoIj5C0ZPpc8AbwK8o5mwaDYynGOb3S6UFlCRJqp3NgH9Ulj8P\n/Ccz3x8ROwPHActSZNoNaLcX1KuvvrpMIbtqhRVWoKmpqaafIUlSX9TVItNOwJHAjZX245k5CZgQ\nEb8EDgBuq348SZKk+oqIdYCvVF4jgYnA14HLWs1ZeX1E3E7R01uSJKkRDQRmVpa3B/5cWf4n8I6l\nPVhELA98Bjisve3Dhq7x5nL/5frRv391B8xZrv/yPP3MVAtNkqQ+Z+LEiUycOLFmx+9qkWlNYEpm\nLoiI2cDqrbb9Gbik6skk1cyaw0fw/LPljE4wbO13MPWZp0r5bEnqoscphoI5h2L+pSc62O8B4M56\nhZIkSaqzfwO7RsSlFCPbHFdZP5RidJul9Ung7kVzPLXVMmHXZQrZVavudCXz5s2zyCRJ6nOam5tp\nbm5+sz1+/PiqHr+rRaangSGV5f8A/5+9O4+vojz7P/65srAFWRs2kcUVcQtYrYpCRAUBwV/dpVVA\nW61F62NdcGkL6fJTfu7VpxWL8IgPuNSlitWqqLFi3RUFQSsoi5EEtMqWsCXX74+ZxJOQwAk5OXNO\n8n2/XvPKmbnn3HNdc7bJ3HPfczLwXDg/ANic0KhEpFEVF62i96S5kWx7xdTRkWxXRKQeTgGec/eK\nna0U3sT6+OSEJCIiIpJ0BQRD290KvOjub4bLhwPv70Z951LHUHkiIiKSvuJtZHqR4CTKY8Afgelm\ndhiwhaDL9N2NE56IiIhIcrn7szWXmVlHoDewJGbIPBEREZEmy90fN7NeQA/gg5iieQTnh+JmZm0I\nzh9dlLgIRUREJBXE28g0CdgDwN1nmFkZwU0fWwPXAHc1TngiIiIiyWVmBUBLd782nB8KPAm0AVab\n2XB3/yjKGEVERESSwd1LgBIz62pma929IqZHU33qKQVyd7miiIiIpJ2MeFZy9w3u/mXM/IPufrq7\nj3T32919e+OFKCIiIpJUPwI+jpm/FZgPDAI+AW6MIigRERGRZDKzbDP7f2a2ASgC+oTLp5rZzyMN\nTkRERFJGXI1MZlZqZt+vo2yAmZUmNiwRERGRyPQAPgMws72Aw4DJ7v4GcBtwVISxiYiIiCTLZGA0\n8GOC2yVUegsYH0VAIiIiknriHS6vFXU3SLUEMhMTjoiIiEjkNgDtw8dDgW/c/a1wfjPBsHkiIiIi\nTd25wAXu/oqZVcQsXwTsH1FMIiIikmLqbGQysx5Az5hF/c2s5mqtgAnAisSHJiIiIhKJV4Brw5Mp\nVxHcj6nS/sCqSKISERERSa4e1H6+J4v4L1oWERGRJm5nBwU/Jega7eF0Xy3rGLAVuDjxoYmIiIhE\n4grgAeAhYAFwQ0zZ+cA/owhKREREJMk+AgYDy2ssPwt4N+nRiIiISEraWSPTvcDTBA1JbxH0WFpU\nY52twOfuvrFxwhMRERFJLncvIhgmrzbDCYbMExEREWnqCoD/De9RmQmcaWb9gLHAqEgjExERkZRR\nZyOTu68GVgOY2YHAcnffUtf6IiIiIk2du6+POgYRERGRZHD3uWZ2FnA9UEEw2s17wGh3nxdpcCIi\nIpIy4hpD190/MbMMMzsVOBboBPwHeBV42t0rdlqBiIiISBoxszOA0wjuT9mqZrm7H5k/cL1mAAAg\nAElEQVT0oERERESSzN2fA56LOg4RERFJXXE1MplZZ+BZ4PtAMVACdAWuBN42sxHu/p9Gi1JEREQk\nScxsCvAb4ANgMcHwwCIishu69+xFcdGqqMMQkYiZWXtgOnAwQa+oC9z9zWijEhERkUSIq5EJuBXY\nCxji7q9WLjSz4whuin0rwT2bRERERNLdhcBN7n591IGIiKS74qJV9J40N2H1rZg6OmF1iciOzGwD\n4PGs6+7t6lH1ncAz7n6mmWUBbXYnPhEREUk98TYynQJcEdvABODur5rZdcBtCY9MREREJBp7AC9G\nHYSIiIhIBC5NdIVm1g44zt3HA7j7dkD3uRQREWki4m1kak1wD6ba/CcsFxEREWkKHgJORg1NIiIi\n0sy4+/2NUG1f4CszmwkcBrwDXO7uZY2wLREREUmyeBuZ3gauNrMX3X1z5UIzawVcBbzVGMGJiIiI\nROBFYKqZfQ94Afi25gru/kzSoxIRERGJgJkNBfqHs4vd/aV6VpEFDAQmuvs7ZnYHcC0wOXalKTMX\nVT3Oz+tC/oAuux+0iIiIVCksLKSwsLDR6o+3kekq4CVgpZk9A5QAXYCRBL2Yjm+c8ERERESS7uHw\nbx9gXC3lDmQmLRoREUlp3Xv2orhoVcLq67bnXqz+YmXC6hPZXWbWF3gMOBT4Mlzcw8wWAqe7+2dx\nVvUFsMrd3wnnHwUm1VxpyoSDGxixiIiI1CY/P5/8/Pyq+YKCgoTWH1cjU3ilyf4EV5ocAQwGVgOz\ngZvdfXVCoxIRERGJTt+oAxARkfRRXLSK3pPmJqy+FVNHJ6wukQa6D9gA7O3uKwHMrBdwPzAdGBpP\nJe5eYmarzGx/d/83cAKwuJFiFhERkSSrs5HJzAYD77n7RgB3Lwb+K1mBiYiIiETB3VdEHYOIiIhI\nCjgaOKqygQnA3Vea2RXA6/Ws6xfAbDPLBj4DJiQuTBEREYnSznoyvUxwQKH7LYmIiEizYmZZwOnA\nsUAn4D/Aq8Dj7r49ythEREREkmQlwS0SamoF1GuMSHf/gGBkHBEREWliMnZSZkmLQkRERCRFmFkX\n4B3gQWAUsHf49yHgbTPLjTA8ERERkWS5EvijmR1lZplmlmFmRwF3hGUiIiIi8d2TSURERKQZuQ3o\nTDA8TFWPbjM7guDm17cB50UUm4iIiEiyPAi0BF4DKsJlGUA5wdB3VSu6e7ukRyciIiIpYVeNTCPN\nrF88Fbn7rATEIyIiIhK1kcClsQ1MAO7+tpldB9wVTVgiIiIiSXVp1AGIiIhI6ttVI9Nv4qzHATUy\niYiISFPQEthQR9kGoEUSYxERERGJhLvfH3UMIiIikvp21ch0PME9CURERESaizeASWb2krtvqlxo\nZjnApLBcREREpFkws05AF2rc19vdF0cTkYiIiKSSXTUylcWeXBERERFpBq4EXgZWmdnzQAnBiZXh\ngAH50YUmIiIikhxmNgCYCRxSuYhgJJvKv5kRhSYiIiIpZFeNTCIiIiLNirsvMLP9gKuAI4BDgdXA\nPcBt7v5VlPGJiIiIJMkMoAi4nOCiG482HBEREUlFamQSERERqSFsSLo26jhEREREIrQfcKa7L21o\nRWa2HFgHVADb3P3IhtYpIiIiqSGjrgJ3z3D3t5IZDICZZZjZe2b2VLK3LSIiIiIiIiLVde/ZCzNL\n2NS9Z6+oU5L4zAcOTFBdFUC+uw9QA5OIiEjTkoo9mS4HFgPtog5EREREmgcze6keq7u7n9BowYiI\niKSY4qJV9J40N2H1rZg6OmF1SaO6EJhuZnsDi4BtsYXu/s961GXs5EJnERERSV8p1chkZj2BkcAf\ngF9GHI6IiIg0H1/HsU534Bh0PwIRERFpHvYDBgDDaylzILMedTnwgpmVA/e6+18SEJ+IiIikgJRq\nZAJuB64G2kcdiIiIiDQf7n5mXWVm1guYBJwCfEVwvCIiIiLS1E0D5gE3AiU07EKbQe6+2sxyCRqb\nlrj7/EQEKSIiItFKmUYmMxsFlLj7AjPLJ+hKLSIiIhIJM9sXuA74MbAmfDzN3csiDUxERCTNtcwE\ns8T+y997z24s/2J1QusUegIj3X1ZQyty99Xh37Vm9gRwJME9n6pMmbmo6nF+XhfyB3Rp6GZFREQE\nKCwspLCwsNHqT5lGJmAQMMbMRgKtgT3MbJa7n19zxfHjx9OnTx8AOnToQF5eHvn5+QBVO6vJzi/f\nHsz3yUrqfKXI89d8QuYrbV65EIBWvQ5J6nylVNkfmte85qt/PxQWFrJ8+XKaKzM7CLgBOBNYRXC/\nyBnuvjXSwERERJqILeVQMTmxt2HOKChOaH0CwAvA4UCDGpnMrA2Q4e4bzSwHGAYU1FxvyoSDG7IZ\nERERqUN+fn7V+R+AgoIdfoYbxNxT77YCZjYEuNLdx9RS5qkYczKYWcIPxOOVUbCe5rrfmyIzS+iN\ne+tjxdTRei+JpBEzw92bRe9iMzucoHHpVOBTgqFh/tfdyyOMKaWPexJ9bJKKxxvNIUeRxpTo484V\nU0en3GeyMXJMte+JVH8dIXGvZXM57omHmf2M4NjofmAhsC223N0fj7OevsATBMPtZQGz3f2mGut4\nReFZiQi7Th1Hz2X5yi/p0KFDo25HREQk1SX6mCeVejKJiIiIRMLMniW4qnYhcI67/zXikERERESi\n9qfw7/W1lDmQGU8l7v45kJeooERERCS1pGQjk7u/ArwSdRwiIiLSbAwP//YE/tvM/ntnK7u7bhIg\nIiIiTZq7Z0Qdg4iIiKS+lGxkEhEREUmyxA5ILCIiIiIiIiLSDKiRSURERJo9d1cjk4iIiEgNZtYR\nGAH0AlrElrn7byMJSkRERFKKuj6LiIiIiIiIiEg1ZnYUsBS4BfgdcAFwA3AVcEaEoYmICPD++++T\nlZXFcccdV2t5YWEho0ePJjc3lzZt2tC/f39+8YtfsGLFijrrvP/++8nIyCAzM5OMjAx69OjB2Wef\nzfLly6vW2bp1K5dddhm5ubm0bduWU089laKioqryFStW8JOf/IR99tmHNm3asM8++3D99dezefPm\nhOUuqUWNTCIiIiIiIiIiUtPNwGxgT2AzMJSgR9M7wNQI45I01hgnxSu98sorZGRkVE1dunRh5MiR\nfPjhh1Xr5OfnV5W3bNmSHj16MGLECGbPnr1DfX/5y18YOnQoHTt2JCMjg5UrV+5+4iKNYPr06Uyc\nOJFFixbxySefVCubNm0aJ554Irm5uTz66KN8/PHH3Hfffbg7f/jDH3Zab05ODsXFxaxevZoHH3yQ\nBQsWcOqpp+LuAFx++eU88cQTPPzww8yfP5/169dzyimnVJV//PHHVFRUMG3aNBYvXszdd9/NrFmz\n+K//+q/G2RESOTUyiYg0QX16dsfMIpn69OwedfoiIiIiItJwhwJ3e3DWsBxo6e4lwCRgSpSBSfpq\nrJPilcyMJUuWUFxczDPPPMM333zDiBEj2LBhQ1X5BRdcQElJCZ9//jlz587lmGOO4eKLL+a0006r\nOkkOUFpayvDhwykoKMDMErcTRBJg8+bNzJkzh4suuojTTz+d6dOnV5UVFRVx+eWXc9lllzFjxgyG\nDBlCr169OProo7nrrru45ZZbdlq3mZGbm0vXrl0ZMmQIU6ZMYdGiRSxbtoz169czY8YMbrnlFoYO\nHUpeXh4PPPAAH374IfPmzQNg+PDhzJgxgxNPPJE+ffowYsQIbrjhBh577LFG3ScSHd2TSUSkCVpR\nVEzF5HaRbDujoDiS7YqIiIiISEJtjXlcAvQGlgAbgR6RRCRprfKk+Pz589m0aRPTp0/n5ptvBqqf\nFL/99turnlN5Ynz9+vVxbyc3N5dOnTrRpUsXbrvtNgYNGsSbb77JiSeeCECbNm3Izc0FoEePHhx+\n+OH84Ac/4OSTT2bWrFmMGzcOCHprALz77rsJyV8kkf7617/Sp08fDjroIM477zzOOussbrrpJjIz\nM3nkkUfYtm0bkyZNqvW57drV73xRy5YtAdiyZQvvvvsu27dv56STTqoq79mzJwceeCD/+te/qi2P\ntW7dOjp27Fiv7Ur6UE8mERERERERERGp6T3giPBxIfB7MxsH/BH4sK4nidSl5knxWbNmUV5eDpDw\nk+KVYk+O78ywYcM45JBD1NNC0saMGTM4//zzARgyZAg5OTk8+eSTACxdupR27drRrVu3Bm/niy++\n4Oabb2avvfbigAMOoLi4mMzMTDp37lxtva5du1JcXPtFxytWrODWW29l4sSJDY5HUpMamURERERE\nREREpKYbgC/Dx78C1gJ3AR2Bi+pTkZllmNl7ZvZUYkOUdJKsk+KVvv76awoKCmjXrh1HHnnkLtfv\n378/n332WcK2L9JYli5dyvz58zn33HOrlo0dO5b77rsPoNqwjzuzxx57sMcee9CuXTt+/vOfVy3f\nuHEj7dq1o23btvTu3Zvy8nIef/xxsrLqPyhaSUkJI0aMYPjw4VW9A6Xp0XB5IiIiIhEws+XAOqAC\n2Obuu/7PV0RERCRJ3P2dmMdrgRENqO5yYDEQzZjeErnKk+IPPvhg1bLKk+I174XUEO5Onz59cHdK\nS0vZb7/9ePTRR6uGx9vVc3XvJUkH06dPp6Kigr322muHsqKiIvbff3/Wr19PcXHxThtuP/jgg6rH\nsb0Fc3Jy+OCDDzAzunbtSuvWravKunXrRnl5OV9//XW13kwlJSUMHjy4Wv3FxcWccMIJHHroocya\nNWu3cpX0oEYmERERkWhUAPnu/k3UgYiIiIjsipkNBnKA193923o8rycwEvgD8MtGCk9SXKJOiu+K\nmVFYWEjHjh3Jzc2lbdu2cT938eLF7L333ru9bZFkKC8vZ9asWdx0002MGjWqWtl5553HzJkzGT9+\nPNdeey033XQTd9xxxw51rFu3jvbt29f5fjcz+vbtW2vZ4YcfTlZWFi+88ALnnHMOEAypt2TJEgYN\nGlS13urVqxk6dCiHHHIIc+bMISNDA6o1ZWpkEhEREYmGoaGLRUREJMWY2aVAe3f/Q8yypwkaigCK\nzewEd18SZ5W3A1cD7RMbqaSLRJ4Uj0efPn3o1KlTvWJ87rnnWLRoEddcc029nieSbE8//TRff/01\nP/nJT+jYsWO1srPPPpt77rmHX/3qV9x+++1cdtllfPvtt0yYMIG+ffvy5ZdfMmfOHLZs2cK0adN2\na/vt2rXjwgsv5JprriE3N5dOnTpx5ZVXkpeXxwknnAAEDUxDhgyhZ8+e3Hbbbaxdu7bq+bm5uWpw\naoLUyCQiIiISDQdeMLNy4F53/0vUAYmIiIgA44A7K2fM7IfAcOA8YAlwN/BrYOyuKjKzUUCJuy8w\ns3yCi2ykmUnmSfF4ht0rLS2lpKSE7du3s3r1av7+979z880388Mf/pAf/ehHVeuVlJRQXFzMJ598\ngrvz0Ucf8c0339CrV68d8hBJlhkzZjB06NBa34Nnnnkm1113HfPmzeOSSy6hX79+3HrrrZxxxhls\n2rSJ3r17c9xxx3Hdddc1KIY777yT7OxszjnnHMrKyjjxxBN54IEHqoabfP7551m2bBnLli2jd+/e\nwHfDUX7++ef06tWrQduX1KNGJhEREZFoDHL31WaWS9DYtMTd50cdlIiIiDR7+wDvx8yPBJ5y99kA\nZnY9cH+cdQ0CxpjZSKA1sIeZzXL382uuOGXmoqrH+XldyB/QZTfDl1STzJPi8dxTaebMmcycOZPs\n7Gw6depEXl4e9957L2PHVm83veeeeygoKMDMMDNOOeWUqueff/4Ob2GRpHjyySfrLOvbty/l5eVV\n88cffzzHH398veofN24c48aN2+k62dnZ3Hnnndx55521lsdThyRXYWEhhYWFjVa/GplEREREIuDu\nq8O/a83sCeBIYIdGpvHjx9OnTx8AOnToQF5eHvn5+QBVB4lRzQMULt9Ofp+sqsfAbs9XbiNV8qt5\nEN7Q/CrnY3NNhfw0r/nGnK+0eeVCAFr1OqRB85VS7fOYqPwq51Pl9aucr4wxUflBYn8/dvf1rHy8\nfPlypJqWwKaY+aOB2B7XnwFxtQC5+/XA9QBmNgS4srYGJoApEw7erWAl9TX2SfFKQ4YMqVZXbV5+\n+eW465s8eTKTJ0/erVhEpP7Ky8sZM2YMHTt2ZODAgfTv358BAwbQtWvXqENLe/n5+dWO6woKChJa\nv8XTjTSVmJmnW8yJYmZUTG4XybYzCtbH1eVY0oOZ0XvS3Ei2vWLqaL2XkkDfF5IoZoa7a1iTBDOz\nNkCGu280sxzgeaDA3Z+vsV5KH/ck+rsmEd8f3Xv2orhoVYIiCqRajlJ/fXp2Z0VRcULr7L1nN5Z/\nsTqhdTZFiT7uXDF1dMp9Jhsjx1T7nkj11xES91rquAfMbDEw2d3/amZdgNXAke7+blh+JPCku3ev\nZ72VjUxjainzisKzEhB93TqOnsvylV/SoUOHRt2OiIjsnuXLl3PggQeyefNmWrRoQatWrdiyZQsd\nOnTg2GOPZdiwYQwaNIgDDzxQ93VqoEQf86gnk4iIiEjydQWeMDMnOB6bXbOBqTE0RgNMqikuWpXw\nE6GS/lYUFTfCCe3ENlqJiKSQ+4H/NrNDgHxgSWUDU+gYYGFtT9wZd38FeCUhEUrKWbhwIZdddhlT\npkypdrV8Io0cOZJXX311h+VmxvXXX8+1117bKNsVkeRYtmwZLVu2ZPPmzWzdupWtW7cCwf3RHnvs\nMZ599tnKxhEOP/xwzj33XE4//XS6dNHwqlFTI5OIiIhIkrn750BesrerBhgRERGJw81ADjAaKAYu\nrlE+CHg42UFJatq0aRM33HAD06ZNY/PmzUyYMIGlS5eSmZmZ8G3dd999lJWV1VrWqVOnhG9PRJJr\n2bJlbNu2rc7y0tLSqsevvvoq7777LldccQWHHnooF154Iaeddhq5ubnJCFVqUCOTiIiIiIiIiIgA\n4O4VwG/CqbbyM5MbkaSqF198kXPOOYeNGzeyefNmANauXcv06dO5+OKabZMN1717vUZoFJE0s3Ll\nSsrLy8nKymL79u27XL+y0entt9/mo48+4vLLLycvL4+JEydyzjnnkJ2d3dghS0iDF4rE6NOzO2YW\nydSnpw6WRERERERERCS1uTs333wzo0eP5quvvqpqYIKgZ9OkSZPYsGFDhBGKSDq68soreeSRR/jd\n737H6aefTt++fcnOzqZdu3a77B1ZWlrKli1bePPNN/n5z39Ot27dmDp1KuvXr09S9M2bejKJxGiM\n8frjpXH9RURERERERCSVlZWV8eMf/5h//OMftQ5d16ZNG/bYYw9KS0vZY489IohQRNJVx44dGTNm\nDGPGjKlatn79et544w3++c9/8txzz7Fw4UJatGhBaWkp5eXltdazceNGAAoKCvjtb3/LhRdeyDXX\nXEPPnj2TkkdzpJ5MIiIiIiLNWKJ7cqt3toiISNO0cuVKBgwYwDPPPFPt3igAGRkZtG7dmquuuoql\nS5fStWvXiKIUkaakXbt2DBs2jN///ve8/fbbbNy4kXnz5nHZZZeRm5tLTk5Onb2cysrKKC0tZdq0\naey3336MHz+er7/+OskZNA/qySQiIiIi0owluie3emeLiIg0Pa+88gqnnnoqGzZsoKKiolpZmzZt\nOOigg3jwwQfZZ599IopQRJqDrKwsjjzySI488khuu+023nnnHWbPns3s2bPZvHlzVS+mWFu3bgXg\nwQcf5LHHHuPGG2/kkksu2eUQfBI/9WQSERERERERERHMrNzMuoSPZ5iZxjsT7r77bkaMGMG6detq\nbWCaMGEC//rXv9TAJCJJZWYcccQR3HHHHaxZs4a///3vnHDCCbRq1arWBqStW7eyceNGrr32Wvr1\n68drr70WQdRNkxqZREREREREREQEoAxoGz4eB7SKMBZJAbNnz+aaa66p8/5L99xzD3fffTdZWRos\nSUSiY2YMHjyYefPmsWDBAsaOHUurVq1o2bLlDutu2rSJpUuXctJJJ3HGGWfw7bffRhBx06JGJhER\nERERERERAfgX8DczmwkY8MewR9MOU7wVmllLM3vTzN43s4VmNrnRopeEmj9/Pj/96U93aGDKzs4m\nNzeXV199lfPOOy+i6EREanfAAQcwa9YsVqxYwRVXXEGbNm3Izs7eYb2ysjLmzp1Lv379ePvttyOI\ntOlQI5OIiIiIiIiIiACcBzwHdAAc6Azk1jHFxd23AMe7+wAgDxhhZkcmOG5JsE8//ZRRo0bt0MDU\npk0bBgwYwEcffcTAgQMjik5EZNe6dOnCjTfeyKeffsopp5xC69atd1hn69atlJSUMGTIEG699Vbc\nPYJI05/6soqIiIiIiIiICO5eAlwNYGafA+e6+9cJqLc0fNiS4FyUzuKlsK+++or8/Hw2bNhQbXnr\n1q3Jz8/nb3/7W629AkREUlGPHj14/PHHmT9/PhMmTGD16tVs2rSp2jplZWVMnjyZ559/noceeoiO\nHTtGFG16Uk8mERERERERERGpxt37JqKBCcDMMszsfaAYeMHdNS5Ritq8eTMnnXQSa9eurXZFf3Z2\nNvvttx+PPvqoGphEJC0de+yxfPzxx9x0003k5ORgZtXKN23aRGFhIf369WPJkiURRZme1MgkIiIi\nIiIiIiI7MLNRZvZPM/vKzNaa2StmNrK+9bh7RThcXk/gB2bWP/HRSkO5O+eeey6ffPIJ27Ztq1pu\nZnzve99j3rx5tQ43JSKSLjIzM7n00ktZsGAB++233w7faVu3bmXNmjUcddRRvPPOOxFFmX40XJ6I\niIiIiIiIiFRjZj8B/gTMBu4PFx8HPGFml7j7jPrW6e7rzexl4GRgcWzZlJmLqh7n53Uhf0CX3Q29\nTocdciArvyhOeL2xeu/Vg+Urixp1G43lySef5IUXXtjhPkxt27bl5ZdfJjc37ltxiYiktH333ZcP\nPviASy+9lDlz5uzwvbd+/Xry8/N5+umnyc/PjybIBCosLKSwsLDR6lcjk6Sc7j17UVy0KuowRERE\nRERERJqzScAv3f3umGX3mdm7wLVAXI1MZvY9YJu7rzOz1sBJwE0115sy4eAEhLxzK78opqLwrEbd\nRkb+I41af2PZsmULl1xyyQ73KWndujVPP/00BxxwQESRiYg0jlatWjF9+nSGDRvGBRdcQGlpabVh\nQjdt2sSoUaN48MEHGTNmTISRNlx+fn61xrKCgoKE1q9GJkk5xUWr6D1pbiTbXjF1dCTbFRERERER\nAWiZyQ73CGiI3nt2Y/kXqxNWnzQrvYB/1LL8WeCWetTTHbjfzDIIbtvwsLs/k4D4JIFuu+021q9f\nX21ZTk4Od999N4MHD44oKhGRxnfWWWcxcOBAjjvuONauXUt5eXlVWWlpKeeccw7Tpk3jvPPOizDK\n1KZGJhERkd3Qp2d3VhQ17lAbddHJIhERkaZrSzlUTG6XsPoyCqI5XpEmYSVBr6OlNZYPA1bEW4m7\nLwQGJjAuSbCSkhJ+//vfU1paWrXMzDjssMMYN25chJGJiCTHvvvuy7vvvsugQYMoKiqqdl+6srIy\nLr74Yjp16sSoUaMijDJ1pUwjk5n1BGYBXYEK4C/u/sdooxIREandiqLihJ4Aqg+dLBIRERGRJLgF\nuMvMBgL/CpcNAs4DLossKkm4K6+8ku3bt1db1qpVK+69996E9qwUEUllPXr04J133mHIkCEsXbqU\nLVu2VJWVlZVx1llnMX/+fAYMGBBhlKkpI+oAYmwnGOv3IOBoYKKZ9Ys4JhERERERERGRZsfdpwFn\nAwcSNDjdAvQDznL3e6OMTRLn/fff5/HHH2fr1q1Vy1q2bMnYsWM56KCDIoxMRCT5OnfuzBtvvMGA\nAQNo3bp1tbLS0lJOOOEEVq5cGVF0qStlGpncvdjdF4SPNwJLgD2jjUpEREREREREpHly9yfc/Vh3\n7xxOx7r7k1HHJYlz0UUXUVZWVm1ZdnY2U6dOjSgiEZFotW3blpdffpkBAwbQsmXLamXr168nPz+f\ndevWRRRdakqZRqZYZtYHyAPejDYSERERERERERGRpmfBggUsXry42rKcnBx+97vf0blz54iiEhGJ\nXqtWrXjmmWfYc889yczMrFpeXl5OUVERI0aMqHbfpuYu5RqZzKwt8ChwedijSURERERERERERHaT\nu7Nhw4Zqy+64445q9xwB6NixIxMnTkxmaCIiKal9+/YUFhbSvn37asu3bt3KggULuPrqq6uWrVy5\nkmOPPZY5c+YkO8yUkBV1ALHMLIuggemBnXW/Hj9+PH369AGgQ4cO5OXlkZ+fD0BhYSFA051fHtyI\nMb9PVlLnKyUr30qbVy4EoFWvQ5IyX5lzsvdv1XwT37+V85Ui/zw14nz3nr0oLlpFlJrL90VU85U5\nN7Xvi8rHy5cvR0REREREmoZly5ax//77k5eXx0UXXcSIESN45JFHKC8vr1onJyeHX/3qV2RnZ0cY\nqYhI6thrr7148cUXGTRoEKWlpVXLy8rKuPfeexk2bBjffPMNP/vZz9i0aRP9+vVj7NixEUYcDXP3\nqGOoYmazgK/c/Zc7WcdTKeZkMjMqJreLZNsZBetJ1n43M3pPmpuUbdW0Yupo7eNGtmLq6KTlGaUo\n9zE0n/dylJrTd7K7W1I2JjtI9HFPor+bEv1dk4j3dnPIMdES/X3WHHKE1MwzFTWHz6RyrL/GOFZN\nVJ467omGmXlF4VmNuo2Oo+eybkMZjb2djPxHUvL3Yd26dXTu3Jny8nJycnLYunUrLVq0YNOmTVXr\ntG7dmjVr1tC2bdsIIxURST3PPPMMZ5xxRq33sMvOzq5qgDriiCN46623ogixXhJ9zJMyw+WZ2SDg\nR8BQM3vfzN4zs5OjjktERERERCQd9OnZHTNL2NSnZ/eoUxIRkQRp3749rVq1AmDTpk1s27atWgNT\nZmYmRx99NC1atIgqRBGRlDVy5Eiuu+462rRpU235tm3bqvVw+uyzz5IdWkpImXVj2W0AACAASURB\nVOHy3P01IHOXK4qIiIiIiMgOVhQVJ7gHTHHC6hKR9GRmdwO/cff/NKCOnsAsoCtQAfzF3f+YoBCl\nHvbcc0/+/e9/11pWXl7O22+/TceOHfnhD3/IRRddxODBg5McoYhI6rrhhht4/vnnefPNN9m2bVut\n63zzzTdUVFSQkZEyfXuSonllKyIiIiKS5rr37JXQ3ioSHb2WIpKKwkahSmOBtuHyhWa2125UuR34\npbsfBBwNTDSzfg2PVOrr+9///k7LN2zYQGlpKXPmzGHUqFFs3rw5SZGJiKS+Tz/9lK+++oqsrLr7\n7WRnZ/PNN98kMarUkDI9mUREROqre89eFBetijoMEZGkKi5alfB7o0g09FqKSIr62My+Bl4DWgF7\nASuBPkB2fStz92KgOHy80cyWAHsCHycqYInPSSedxFNPPcXGjRvrXCcjI4MOHTpQWFhYNbyeiEhz\n5u78+c9/5qqrrmLz5s07ve9eixYtKCkpoXPnzkmMMHpqZBIRkbSV6JNz9aETeSIiIiLSRHUABgLH\nAacBz5hZCdASGG5mj7t7ye5UbGZ9gDzgzcSEKvUxaNAgKioq6iw3M7p06cIbb7xB7969kxiZiEjq\neuONN5g4cWJc62ZkZLBmzRr69+/fyFGlFg2XJyIiIiIiIiIilbLd/S13vxUoAwYAE4By4ALgczP7\npL6Vmllb4FHgcnevuyuNNJp9992XzMy6b4eek5NDYWGhGphERGIcffTRvPbaaxx11FG0adNmp+tW\nVFSwZs2aJEWWOtSTSUREREREREREKn1rZgsIhstrAbR299fMbDtwNlAEHFGfCs0si6CB6QF3f7K2\ndabMXFT1OD+vC/kDuuxm+FIXM+OII47gpZde2qGsdevWPP744xxwwAERRCYiktqOOeYYXn/9dV55\n5RXGjx/PmjVrKC0t3WG9rVu3pmQjU2FhIYWFhY1WvxqZRERERERERESk0p7A0cAxBOeN3jWztwka\nnAYCX7j7/HrWOQNY7O531rXClAkH72a4Uh8nn3wy8+fPZ+vWrVXL2rRpw9SpUznppJMijExEJPUN\nGTKETz75hKlTp3LjjTeyZcuWasOQbtmyhdWrV0cYYe3y8/PJz8+vmi8oKEho/RouT0RERERERERE\nAHD3r9x9rrtfB5QS9Fq6C3DgFmCdmb0Sb31mNgj4ETDUzN43s/fM7OTGiF127bjjjqNly5ZV823a\ntGHs2LFceumlEUYlIpI+WrRowa9//WsWLVrE0UcfTU5OTrXydevWRRRZdNTIJCIiIiIiIiIidVnn\n7o8A24ChQF/gT/E+2d1fc/dMd89z9wHuPtDd/9FYwcrOff/73696nJGRQf/+/fnzn/8cYUQiIulp\n77335tVXX+Uvf/kL7dq1IysrGDSurKws4siST41MIiIiIiIiIiJSm0OBL8LHK4Bt7l7s7g9HGJM0\nQFZWFqeddhpmRqtWrXj44YerToyKiEj9mBnnnnsuS5YsIS8vj4yMDLZs2RJ1WEmnRiYRERERERER\nEdmBu69y94rw8cHuvirqmKThzj//fNydgoIC9t5776jDERFJez169OD111/n2muvJS8vL+pwkk6X\nKtRD9569KC7S8ZSIiIiIiIiIiKSnIUOGcNddd3HJJZdEHYqISJORlZXFH/7wh6jDiIQameqhuGgV\nvSfNjWz7K6aOjmzbIiIiIiLJoou7REREGk9mZiaXXnpp1GGIiEgToUYmERERERFJKYm+uEsXa4mI\niIiIiDQO3ZNJRERERERERERERERE6k2NTCIiIiIiIiIiIiIiIlJvamQSERERERERERERERGRelMj\nk4iIiIiIiIiINAozu8/MSszsw6hjERERkcRTI5OIiIiIiIiIiDSWmcDwqIMQERGRxqFGJhERERER\nERERaRTuPh/4Juo4REREpHGokUlERERERERERERERETqLSvqAESkeWmZCWYWybZ779mN5V+sjmTb\nIiIiIiIiIiIiIk2NGplEJKm2lEPF5HaRbDujoDiS7YqIiIiIiMjOTZm5qOpxfl4X8gd0iTCa1Nan\n156sWPVlo26j9149WL6yqFG3ISKSjpLxHQyJ/R4uLCyksLAwIXXVRo1MIiIiIiIiIiLSmCyc6jRl\nwsFJCiX9rVj1JRWFZzXqNjLyH2nU+kVE0lUyvoMhsd/D+fn55OfnV80XFBQkrG7QPZlERERERERE\nRKSRmNkc4F/A/ma20swmRB2TiIiIJI56MomIiIiIiIiISKNw97FRxyAiIiKNRz2ZRERERERERERE\nREREpN7UyCQiIiIiIiIiIiIiIiL1pkYmERERERERERERERERqTc1MomIiIiIiIiIiIiIiEi9qZFJ\nRERERERERERERERE6k2NTCIiIiIiIiIiIiIiIlJvamQSERERERERERERERGRelMjk4iIiIiIiIiI\niIiIiNRbSjUymdnJZvaxmf3bzCZFHY+IiIhIY9Fxj4iIiDQXOu4RERFpulKmkcnMMoC7geHAQcC5\nZtYv2qhEREREEk/HPSIiItJcNNfjnsLCwqhDaBRNMS/llB6UU3pQTs1TyjQyAUcCn7r7CnffBjwE\nnBpxTCIiIiKNQcc9IiIi0lw0y+OepnpSsinmpZzSg3JKD8qpeUqlRqY9gVUx81+Ey0RERESaGh33\niIiISHOh4x4REZEmLCvqAEREREREREREpHkb8+t3G7X+si3bGrV+ERGR5srcPeoYADCzo4Ap7n5y\nOH8t4O4+tcZ6qRGwiIhIM+HuFnUMTY2Oe0RERFKTjnsSL57jHh3ziIiIJFcij3lSqZEpE/gEOAFY\nDbwFnOvuSyINTERERCTBdNwjIiIizYWOe0RERJq2lBkuz93LzexS4HmCe0XdpwMOERERaYp03CMi\nIiLNhY57REREmraU6ckkIiIiIiIiIiIiIiIi6SMj6gBERERERERERKR5MrOTzexjM/u3mU2KOp54\nmVlPM3vJzD4ys4Vm9otweUcze97MPjGz58ysfcxzrjOzT81siZkNiy76uplZhpm9Z2ZPhfNpnQ+A\nmbU3s7+GcX5kZj9I97zM7AozW2RmH5rZbDNrkW45mdl9ZlZiZh/GLKt3DmY2MNwP/zazO5KdR6w6\ncvp/YcwLzOwxM2sXU5byOYXx7JBXTNmVZlZhZp1ilqV8XnXlZGaXhXEvNLObYpanZU5mdpiZvW5m\n75vZW2b2/ZiyxOXk7poSOAEDgO3Aq3WU5wNzgbVAKbAY+CPQeyd1jgMqgPLw75fAw0CfmHVaAHeF\n9W4EngT2rKO+lsAHYV0Do95nTW0/AwMJhgH4JlxvGpAT9T5Lo338U+ClcP9VAL1qqacD8ADwbTjN\nAtpHvc+ayj4GegPTgWXhdpcB/xdoFfU+a0r7OVxneVhWEVPf/416n2lqWlOiPwPAFcC62r4TCC5g\nKgJ+H7NsSrisFHgZ6F/jOV3D7/TVwCZgATA2XXIEfgj8A1gTfo4H1/KcuI/T0jzPXX7vpXOOQMdw\nW0vCba8E/gR0aio51vL8Z8P1TmtKOQKF7Pj7O6cpvV9jnnckwf8mG4D1wPz6vGdTPUcS8BuiKdop\nfN2XEvwPkh2+hv2ijivO2LsBeeHjtgT3neoHTAWuCZdPAm4KH/cH3ie4dUWfMG+LOo9a8roC+F/g\nqXA+rfMJY/0fYEL4OAton855AT2Az4AW4fzDBP+jplVOwLFAHvBhzLJ65wC8CRwRPn4GGJ5iOZ0I\nZISPbwJuTKec6sorXN6T4Lf6c8LjC+DAdMirjtcqn+C4KSuc/14TyOk5YFj4eATwcmO8/9STKfF+\nAvw3cLCZHRBbYGYXA/MIDsDPIDj4uBAw4IZd1LuJ4ACmO3AuwRvmSTOzsPxOgoPwswneUO2Ap2PK\nY91C8E+x1ze5FJKS+9nMugMvEHwwjwROBg4iOKBJN1Ht4zYEX4CTqfs9+mD4vGHAcIKGvVn1yC1V\npOo+7kfwz97FBD86lwLnA5FfPbObUnU/Ey6fQnCCpLKu38efmkhcEv0ZeICg0eTMWspGEryfp4f1\nTyI4UTER+D7BicIXzCynRn0HAKMJfjNnAQ+Y2bHpkCOQA7wW5lnXZ70+x2k7k+p5xvO9tyupnGOP\ncLoKOBj4ETAYmBNfalVSOcfYWK4iaFzYndcy1XN0YAbVf38vjiOvmlI6TzP7AcFn8iWC/00GEvwv\nuC2u7AIpnSOJ+Q2RaB0JfOruK9x9G/AQcGrEMcXF3YvdfUH4eCPBRQg9CeK/P1ztfuD/hI/HAA+5\n+3Z3Xw58SpB/yjCzngSfxekxi9M2H4Cw18hx7j4TIIx3HWmeF5AJ5JhZFtCaoJE+rXJy9/kEFyfF\nqlcOZtYN2MPd3w7XmxXznKSrLSd3n+fuFeHsGwTfE5AmOUGdrxXA7cDVNZadShrkVUdOlxA0bG4P\n1/kqXJ7OOVUQNKxDcNF+Ufg4se+/qFrWmuIEtApfyIMIfpBvjinbE9gM3F7Hc9vtpN5xwPoay84l\nuOJuX4ITFVuAc2LKe4blJ9V43qnAQoID8bTsyZTK+5ngCt41Neo4ONzXe0e971J9H9dYfni4vGYv\nm37h/jwqZtmgcNl+Ue+7prCP66j3EmBt1Putqe1ngqt9fhn1ftLUdKdG/Aw8QngFVI3ljwPzYua/\nBK6tEc964KcxyzYA42rUszzez0bUOcYs70ztvV/iPk5L5zxrrBP370u65hiz7giChpi2TSlH4Ahg\nBfA96tmTKR1yJOhV+cf6vD/TNM/XgN828Rwb9BuiKfoJOB24N2b+xw39fEaUR5/wvdcW+KZG2X/C\nv3cR09Mu/FzVq6doEvL4K8HFc0P4ridT2uYTxnUYwVX5M4H3gHsJLoxJ97x+EX4HlgAPpOtrRdCL\nMbbXxX/qkwPBcefzMcuPrXzvpkpONcqeAs5Nt5zqeK3GALeFjz/nu55MaZNXLTm9T3Ah8BsEx4uH\nN4Gc+hEc168EVgF7NUZO6smUWGcCy939I4Irqs43s8yw7CyCrt9Ta3uiu6+v57a2hH9bErz4WQQ9\naCrr+4LgKppjKpeFV6T8ieBE6OZ6bi+VpPJ+bsmOVwZW7ut0upouqn0cj6OBDe7+Rsw2XyPoVXJM\nnc9KPam8j2vTntqvWkl16bCfrzKzr8Lxca83s+x6Pl9kZxrrM3AfcJyZ7V25wMy6AKfwXS+mvgQ9\nBGJ/NzcD/6T69/WrwFlm1skCpxKc2J6X6jnGKa7jtDikep6JkI45tif4/i+Nc/2Uz9HM9gBmEzQG\nf7Wr9WuR8jmGzjGztRbcz+JmM2tbz+endJ5mlktw3FxiZq+G4/P/08yGxlsHKZ5jqKG/ISINFn5/\nPApc7kGPJq+xSs35lGRmo4ASD3pn7ay3dVrkEyOLoCfnf7v7QIJzB9eSpq8TgJl1ILiIvDdBL+sc\nM/sRaZzTTjSFHAAwsxuAbe7+YNSxNJSZtQauJxjBoCnJAjq6+1HANQQN7+nuEoLfp14EvcNnNMZG\n1MiUWBcQDtnl7q8Q/HBVdvPel+Cq9+KGbiRsLLqaoPXxE4ITOOXu/nWNVUvCMswsg2BM3ZvdfVFD\nY4hYyu5ngqEovmdmk8ws28w6AjcS/Ch2b2hMSRTVPo5HN4IhOWpaw3evQzpI5X1cs47ewJUEQ6Wk\nm1Tfz3cSNPznE1xFcgXpuZ8ldTXWZ+B54Iuw/krjCK6mfDyc70bw+1dS47mxv5sQDCEH8BXByfoH\nCK7u2+GmsnWIMsd4xHP8EI9UzzMR0irH8ATPbwmuwK/Y1fqhdMjxz8Az7v78bsQB6ZHjbILhDvMJ\nXsPTCU4Q10eq51nZgDOFoOFmGEGDzHNmdkicdaR6jtDw3xCJXhHQK2a+J98N5ZPywqHKHiXoSfJk\nuLjEzLqG5d0I/leFIK+9Yp6earkOAsaY2WcEQ9QPNbMHgOI0zafSF8Aqd38nnH+MoNEpXV8nCO7x\n85m7/8fdy4EnCC5eSuecKtU3h7TIzczGEwxFOTZmcTrntA9BD84PzOxzghjfCy8aqet7PR3yWkV4\nLOLBcHHlZtaZ9M5pnLv/DcDdHyUYsQAS/P5TI1OCmNm+BD1VYluj5xCMSQ07vwoktp4N4bTezP4U\nU9Q2XLaRoItbJkHX1u1xhngDsMXdK++pUt97AKSEVN/P7r6Y4B+ky4EygmGCPuO7m9WmvFTfx01B\nOu3j8ODuWeA5d7+zvs+PUjrsZ3e/w91fcfdF7j6D4AqTC8MGapEGaczPgAf95mcSXNVeWc8E4H/d\nfWs9Q/0DwVBIQwl6/dxMcD+NXZ4ETaMcG6Q55JluOVpwX7G5BP+ITorzOSmfo5mdRzCs0DXxPqfG\n81M+x7Cu6e7+grt/5O6PEPTKGWZmeXHGlw55Vv6vf4+73+/uH7j7DcDbwM/iiC0dcoQG/IZIyngb\n2NfMeptZC+AcguGk0sUMYHGN/5WeAsaHj8cBT8YsP8fMWljQ43tf4K1kBbor7n69u/dy970JXoeX\n3P08gt+78eFqaZNPJXcvAVaZ2f7hohOAj0jT1ym0EjjKzFqF36MnAItJz5yM6r8p9cohvNhhnZkd\nGe6L82OeE5VqOZnZyQQXpY5x9y0x66VTThCTV3gOo5u77+3ufQkacwe4+xqCvM5Ok7xqvv/+RnBM\nQfid0SK8WDCdcyoysyEAZnYCwb2XIMHvv6xEZtDM/YTgQH6V1biHs5ntCfwbaGdm3XZxtddhMY9j\nhxjYFJY5QfflspiyYiDTzDrXuEq2K8GQNBB8QI41s5pDub1hZg+HBw7pINX3M+7+EPCQBUNUbAoX\nX0nQ2JQOotzH8SgGcmtZ3iUsSwepvo8rY+kGvAh8SPCjkm7SYj/X8BbBwcC+BP9wizREY38GZgK/\nBoab2XqC+z2eHVNeTPB+7krwT0elrmEZFgyVdClwaExP64VmNhi4DLgoxXOMR1zHD7uQDnk2VNrk\nGDYwPUtw36nR9Tgpng45DgUOBDbViPERM/uXuw/exfPTIcfavEvweu4HLIhj/XTIc3X4d0mN5Yup\nfiVuXVI+xwT8hkgKcPdyM7uUoIdbBnCfu9d836YkMxtE0CtyoZm9T/B/wfUEw0g+YmYXEFyMdhYE\nF6Wa2SMEn8NtwM/DRtdUdxPpn88vgNkWDE3+GUHDdiZpmpe7v2VmjxLcP2Zb+PdeYA/SKCczm0PQ\nq7izma0kGHrtJuCv9cxhIvA/BPcSfMbd/5HMPGLVkdP1QAvghfA39Q13/3m65AS15+XuM2NWcb5r\ngEqLvOp4rWYAM81sIUEP6fMh7XP6KfBHC4Y83kx4fJTwnDzCG1E1lYngh+lLglbp/jWmd4FfEXQt\n2wzcUUcd7XdS/zhq3GC+RvnObih9Yjjfu0ZcJxL0rDkD6BH1Pmwq+7mO511AMPRDnTe/TZUp6n1c\nY91abxpOcMO6cuComGXHhMv2i3ofNoV9HJZ1Jzgp8QiQEfV+a6r7uZZ1Tw3X7Rn1PtSU3lNjfwZi\n1vkHwRAx0wmueqpZ/iVwbcx8K2Ad8JNw/uDwPd+/lnqnp0OOMevVemN6dvP4Id3yrLFO3N976ZYj\nwQ3dXyVoIGwTT37plCPB73/N+CoITs71aQo51rHuYeG6xzaV1zIs+wIoqLHsn8DdTSFHGvAbokmT\nJk2aNGnSpClxU+QBNIWJ4KTgFoIbg9Usu4ZgnFQIhkHaTtASOITgCrKjgD8C03ZS/y5PZgJ/Iugu\newIwgODeQO8CVsf6vcMD9YFR77+mtp8JWnsHElwJOZGgt8PEqPdfGu3jrgT/6I8N36MjwvmOMes8\nA3wQbvNogp42f4t6/zWVfUxwgunf4fu7Z7h+5ZQWDU5psp+PAv4rXNaH4AqtL4DHo95/mtJ/auzP\nQExdZ4bbWQ9cVMe2vgF+SHAy8KHwfZ4TlmeF3zeFBGND703Q+3c7MDJNcuwYfo7zw8/6heF815h1\n6nWclsZ57vI3PJ1zJGhgeh1YSDAOfezvY3ZTyLGOOisIhoNtEu9Vgu+ZXxM0hvYmuD/CYoIexE3t\nM3k5wXfwGeF79vqwzoObQo404DdEkyZNmjRp0qRJU+KmyANoChPBuITP1lHWl+o9io4HngbWAqUE\nPQXuBfrupP54TmZmE9xAfi2wkWAMyT13sn7vMK50amRKi/0M3B+WlxF0WR4b9b5Ls308OfwnsrzG\ndH7MOu0JbkL8bTjdTxr0FEuXfRzWUbOscv24rkqPekqT/TyA4GTlfwgaoxcTnPRqFfX+05T+U2N/\nBmLqyia47+AGYI861vkNwY1CS4GX2fGK832AvxIM7bSB4Lfzx+mSY/h9UNtn/Tc16oj7OC2N89zl\nb3g650hwEr2u38ed9pZJlxzrqLOc+BqZ0iJHggtoCvnueP3fwG1Ah6b2mQzXuxpYHtbzBnB8U8qR\n3fwN0aRJkyZNmjRp0pS4ydwdERERERERERERERERkfrIiDoAERERERERERERERERST9qZBIRERER\nEREREREREZF6UyOTiIiIiIiIiIiIiIiI1JsamURERERERERERESaIDOrMLPToo5DRJouNTKJiIiI\niIiIiIiIhMzsf8LGmRtqLB8SLu8UVWx1MbOZZvZULUXdgLnJjmd3mdmzZvaLOsomh/u/PPxbOX2Z\n7DhF5DtqZBIRERERERERERH5jgNlwNVm1rmWsrTh7mvcfVvUccTDzNoCxwNP7GS1jwkazmKnQxo/\nOhGpixqZRERERERERERERKp7GVgO/GZnK5lZfzN72szWm1mJmc0xs64x5WZmvzazlWa22cw+NLMx\nNeq40cw+NrNSM/vczKaaWYuY8slmttDMzjazpeG2nqjsUWVmk4FxwKiYnj6Dw7Jqw+WZ2Q/M7F0z\nKzOzt81seLjO4HrkNNPM5prZL8zsCzP7j5nNMLNWMesMNrPXzWyDmX1rZm+YWf9d7PMRwBJ3X7WT\ndba7+9qw8axy+jrc5v5mttHMzouJ42Qz22JmP4jn9TCz3uH+ONvMCsPX5D0zO8TMDjKz18JtvGpm\nvWu8jqPN7J1w3y4zs9+bWfYuchZJe2pkEhEREREREREREamuArgW+JmZ9a1tBTPrBrzy/9u7/1A/\nyzKO4+/PnLk4TRFzq4xYVJumG1m5Sou50hSUnUwzmzCXUGH/VCxxCrbmWsMWGPZDVsL6MRKWjhxD\n6ZSKKwRLRHKNlNJSMZdlPzyFaXr1x/OcfPbt/PiejXE67v2CwznPfd/P/Vz38/x3Lq77Bn4JvB14\nHzAA3NwZ9mlgFXApcAJNlc62JIs6Y4aBlcCxwCXAh4G9tuoD5gHnA4PA6cCJwPq278vAVuAnwFzg\n1cBdo8Q7QLN13u72/tXtvdUZ08+aAN4DHN/2nw+cA3yqneMQ4IfATpoqo8XAV4Dne2Pq8YFRntO3\nqnoQ+AzwtSTzkhwNbAbWVdXd7bB+vgfA54ENwFuAvwI3ANcClwMnAbPaawCSnAFsaduOAy4GzuXF\nbyS9ZKVqWlV4SpIkSZIkSdIBk2QzcFRVLUtyO/BEVS1PsgS4HTi6qp5KshY4uapO79x7JPBnYHFV\n3ZPkMeC6qlrfGXMH8GhVrRjj+Z8AVlXV/PZ6DXAZMKeqhtu2K4CVnTH/jblnrheA86pqWzvveuCY\nqvpX2/8RmuTI0qra2eeaNgPvBeZV+8/lJN9sr9/fjv8TcGpV/bTPdz4TeLKN474xxqwBrqTZyrBr\ne1Vd2Bl3E/CaNubDgSWdOMf9Hm110sPAx6vq+rb/LJrk3DlVdXPbdhHw1ao6vL2+ExjqmXcQ2FJV\ns/t5B9J0NXOqA5AkSZIkSZKk/1OXAXcl2ThK39uAJUme7mkv4A1JHqBJdvRWFf2MZms4AJKcR1MF\n9EbgFcAh/O8OVL8fSTC1HgfmTHItC4BdIwmm1t1AOtfjrgm4p73ePZK46cSzGKCq/pLkO8BQktuA\n24AbJ9gG71Tgb2MlmDp+Q/PuujEP94z5GPAA8GZgYSfBNJs+vkfr/s7fe2jWv6unbSDJrKp6hua9\nnZRkdWfMDOCwJHOras8E65KmLZNMkiRJkiRJkjSKqvpFkm3ARmBdT/cMYAfN9mvp6dvD+EeVjCQ+\n3kmzFdsa4Ec0W7MNts/rem6U+w/EUSgTramveKrq4iTXAGcCy4D1SQar6sdjPHcQ2N5HfM9W1cMT\njFkIHNHG9FrgkT7m7d3u67lR+kZrm9H5vRb4wShzP9nH86VpyySTJEmSJEmSJI3tCppzjM7sab8X\n+BDwSFWNet5QkseBU4A7Os3vbucDOBl4rKq+2Lln3j7E+CxNBdR4fg2sSHJYp5rpHeydYJlwTf2q\nqvtpKoI2JrkFuAgYK8m0DPjo/jwPIMkRwHdpknQvB7YkWVRVw1X1dB/fY1/dCxxbVQ/t5zzStHMg\nst2SJEmSJEmS9JJQVb8FNtFsadf1dZqKma1JFid5fZLTkmxKMtCO2Qh8NskFSd6U5CqapMZIpdKD\nwDFJlrf3XwJcsA9h/g44Icn8JEclGS3h9H3gBeD6JMclOQ24fGSZk1jTuJLMS7IhybuSvC7JUmAR\n8Ksxxr8VmA3c2cf0M5PM7f3p9G8C/gh8DlgN/B34Rqd/ou8x5rIm6L8KWJ5kbZLjkyxIcm6Sq/tY\nkzStWckkSZIkSZIkSeNbB6wEXjbSUFV/SHIKsAG4FZhFszXbEDBSKXQtzTlLVwNzac4K+mBV7Wrn\n2NGe93QNTeXNEHAleydG+vEtYAnNmUkDwFJgJ50qpaoaTnI2cB1N5c1umm36bgKemcSaJvJPYD6w\nFXglzTZ73wO+NMb4QeCWPiunFtCc/zQiQCU5FFgOnA2c2M71fJILgZ8n2VFVW5nge7R6t84bq+3F\nzqqhJGfRfLtVwL9pEojf7mNN0rSWvc9nkyRJkiRJkiQdDJIM0iSZ5lTVU1MUw33AF6rqxql4vqT9\nYyWTJEmSJEmSJB0EkqwAHgIeBRbSVFBtn8IE06E0Sa5bp+L5kvaflUySWwiwnQAAAIBJREFUJEmS\nJEmSdBBIcinwSeBVwBPADmB1Vf1jSgOTNG2ZZJIkSZIkSZIkSdKkzZjqACRJkiRJkiRJkjT9mGSS\nJEmSJEmSJEnSpJlkkiRJkiRJkiRJ0qSZZJIkSZIkSZIkSdKkmWSSJEmSJEmSJEnSpJlkkiRJkiRJ\nkiRJ0qT9B6uPhwsl8T7UAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x117783410>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## ========================================\n",
    "# Create figures\n",
    "## ========================================\n",
    "print 'Creating figure 3 ...'\n",
    "\n",
    "fig = plt.figure(1, figsize = (29.13, 6))\n",
    "gs = gridspec.GridSpec(2,3)\n",
    "\n",
    "N = 5\n",
    "ind = np.arange(N)\n",
    "width = 0.35\n",
    "labels = ['AC-P04', 'AC-P09', 'AC-P10', 'AC-P12', 'AC-P15']\n",
    "\n",
    "# Generate mutational burden plot, pilot cohort\n",
    "print '...Generating mutational burden plot for pilot cohort...'\n",
    "ax00 = plt.subplot(gs[:,0])\n",
    "\n",
    "pre = ax00.bar(ind, df_burden.ix[:, 'Primary load'].tolist(), width, color = colors_fig3[0])\n",
    "post = ax00.bar(ind + width, df_burden.ix[:, 'Recurrent load'].tolist(), width, color = colors_fig3[1])\n",
    "\n",
    "plt.xlim(-width, N)\n",
    "\n",
    "ax00.set_xticks(ind + width)\n",
    "ax00.set_xticklabels(labels, fontsize = 14)\n",
    "\n",
    "ax00.legend((pre, post), ('primary tumor', 'recurrent tumor'), fontsize = 14)\n",
    "\n",
    "ax00.yaxis.grid(True)\n",
    "\n",
    "# Create annotation for delta purity\n",
    "deltap_04 = (0.60-0.31)/(0.31) # AC-04, primary = 0.31, recurrent = 0.60\n",
    "deltap_09 = (0.67-0.63)/(0.63) # AC-09, primary = 0.63, recurrent = 0.67\n",
    "deltap_10 = (0.20-0.34)/(0.34) # AC-P10, primary = 0.34, recurrent = 0.20\n",
    "deltap_12 = (0.41-0.34)/(0.34) # AC-P12, primary = 0.34, recurrent = 0.41\n",
    "deltap_15 = (0.45-0.45)/(0.45) # AC-P15, primary  = 0.45, recurrent = 0.45\n",
    "\n",
    "delta_p = [deltap_04, deltap_09, deltap_10, deltap_12, deltap_15]\n",
    "#ax00.text(0.36, 2, r'$\\Delta$p=93.5%', ha = 'center', va = 'bottom', fontsize = 13)\n",
    "#ax00.text(1.36, 3.1, r'$\\Delta$p=6.3%', ha = 'center', va = 'bottom', fontsize = 13)\n",
    "#ax00.text(2.36, 6, r'$\\Delta$p=-41.2%', ha = 'center', va = 'bottom', fontsize = 13)\n",
    "#ax00.text(3.36, 3.7, r'$\\Delta$p=20.6%', ha = 'center', va = 'bottom', fontsize = 13)\n",
    "#ax00.text(4.36, 2.5, r'$\\Delta$p=0%', ha = 'center', va = 'bottom', fontsize = 13)\n",
    "\n",
    "plt.ylabel('Total Mutations / Mb', fontsize = 15)\n",
    "plt.title('Mutational Burden, Pre vs Post. Pilot Cohort', fontsize = 17)\n",
    "\n",
    "# Generate extension cohort mutational burden plot\n",
    "print '...Generating mutational burden plot for extension cohort...'\n",
    "ax01 = plt.subplot(gs[:,1])\n",
    "\n",
    "N = 8\n",
    "ind = np.arange(N)\n",
    "width = 0.35\n",
    "labels = ['AC-V08', 'AC-V10', 'AC-V11', 'AC-V12', 'AC-V14', 'AC-V15', 'AC-V16', 'AC-V18']\n",
    "\n",
    "pre = ax01.bar(ind, pre_mut_extension, width, color = colors_fig3[0])\n",
    "post = ax01.bar(ind + width, post_mut_extension, width, color = colors_fig3[1])\n",
    "\n",
    "plt.xlim(-width, N)\n",
    "\n",
    "ax01.set_xticks(ind + width)\n",
    "ax01.set_xticklabels(labels, fontsize = 14)\n",
    "\n",
    "ax01.yaxis.grid(True)\n",
    "\n",
    "ax01.legend((pre, post), ('primary tumor', 'recurrent tumor'), fontsize = 14)\n",
    "\n",
    "plt.ylabel('Nonsyn. Mutations / Mb', fontsize = 15)\n",
    "plt.title('Mutational Burden, Pre vs Post. Extension Cohort', fontsize = 17)\n",
    "\n",
    "### Generate neoantigen and mutational burden plot\n",
    "print '...Generating mutational burden and neoantigen plot for pilot cohort primary tumors...'\n",
    "ax02 = plt.subplot(gs[0,2])\n",
    "\n",
    "# Plot mutational burden of primary tumors, pilot cohort\n",
    "hist_pre_mut = ax02.hist(df_pilot_primary_plt.ix['Total', :].tolist(), range(0, 38, 2), facecolor = colors_fig3[3])\n",
    "\n",
    "ax02.annotate('AC-P02', xy = (35, 1.25), xytext = (27.5, 3), \n",
    "              arrowprops=dict(facecolor = 'black', shrink = 0.05, connectionstyle = \"arc3, rad = -0.3\"), fontsize = 14)\n",
    "ax02.annotate('AC_PD1', xy = (11, 1.25), xytext = (15, 3), \n",
    "              arrowprops=dict(facecolor = 'black', shrink = 0.05, connectionstyle = \"arc3, rad = 0.3\"), fontsize = 14)\n",
    "\n",
    "plt.ylim(0,6)\n",
    "\n",
    "plt.ylabel('# of Samples', fontsize = 14)\n",
    "plt.xlabel('Total Mutations / Mb', fontsize = 14)\n",
    "ax02.xaxis.labelpad = -1\n",
    "\n",
    "plt.title('Primary Tumors of Pilot Cohort', fontsize = 17)\n",
    "\n",
    "ax12 = plt.subplot(gs[1,2])\n",
    "\n",
    "# Plot mutational burden of primary tumors, pilot cohort\n",
    "hist_pre_neo = ax12.hist(pre_neo.tolist(), range(0, 1800, 95), facecolor = colors_fig3[2])\n",
    "\n",
    "ax12.annotate('AC-P02', xy = (1650, 1.5), xytext = (1250, 4), \n",
    "              arrowprops=dict(facecolor = 'black', shrink = 0.05, connectionstyle = \"arc3, rad = -0.3\"), fontsize = 14)\n",
    "ax12.annotate('AC_PD1', xy = (350, 1.5), xytext = (450, 4), \n",
    "              arrowprops=dict(facecolor = 'black', shrink = 0.05, connectionstyle = \"arc3, rad = 0.3\"), fontsize = 14)\n",
    "\n",
    "plt.ylim(0,9)\n",
    "\n",
    "plt.ylabel('# of Samples', fontsize = 14)\n",
    "plt.xlabel('Neoantigens / Exome', fontsize = 14)\n",
    "\n",
    "#plt.show()\n",
    "plt.savefig('figures/figure3.png', bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<a id='supplementary.figure4'></a>\n",
    "## Supplementary Figure 4: CoMut - No Recurrence vs Recurrence of Primary Tumors\n",
    "<a href='#table.of.contents'>Return to top</a><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this section, we will produce Supplementary Figure 4 which features a co-mutation plot of primary tumors in the pilot and extension cohorts. Samples that did not recur are shown on the left and those that did recur are shown on the right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ========================================================\n",
    "# Reorder samples and genes by frequency\n",
    "## ========================================================\n",
    "# We can begin with the data frames that reordered samples and genes by frequency for each cohort\n",
    "list_ = list_remove_nan(df_pilot_primary_R_sorted.columns.tolist())\n",
    "df_pilot_primary_R_sorted_sf4 = df_pilot_primary_R_sorted.ix[:, list_]\n",
    "\n",
    "list_ = list_remove_nan(df_pilot_primary_NR_sorted.columns.tolist())\n",
    "df_pilot_primary_NR_sorted_sf4 = df_pilot_primary_NR_sorted.ix[:, list_]\n",
    "\n",
    "list_ = list_remove_nan(df_ext_primary_R_sorted.columns.tolist())\n",
    "df_ext_primary_R_sorted_sf4 = df_ext_primary_R_sorted.ix[:, list_]\n",
    "\n",
    "list_ = list_remove_nan(df_ext_primary_NR_sorted.columns.tolist())\n",
    "df_ext_primary_NR_sorted_sf4 = df_ext_primary_NR_sorted.ix[:, list_]\n",
    "\n",
    "# We can concatenate cohorts together by recurrence\n",
    "# R = Response (No Recurrence), NR = No Response (Recurrence)\n",
    "df_R = pd.concat([df_pilot_primary_R_sorted_sf4.ix[:,1:], df_ext_primary_R_sorted_sf4.ix[:,1:]], axis = 1)\n",
    "df_NR = pd.concat([df_pilot_primary_NR_sorted_sf4.ix[:,1:], df_ext_primary_NR_sorted_sf4.ix[:,1:]], axis = 1)\n",
    "\n",
    "# Like before, I make an aethetic choice of only displaying some genes. The plot is less crowded now so I display more\n",
    "display_S4 = 15\n",
    "\n",
    "data_00 = df_R.iloc[:display_S4, :]\n",
    "data_01 = df_NR.iloc[:display_S4, :]\n",
    "\n",
    "list_genes = df_pilot_primary_R_sorted_sf4.ix[:,'gene'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ====================================\n",
    "# Clinical annotation 1: Cohort status\n",
    "## ====================================\n",
    "# This annotation displays whether a sample is in the pilot or extension cohort\n",
    "\n",
    "resp_cohort = format_cohort(df_pilot_primary_R_sorted_sf4.ix[:,1:], df_ext_primary_R_sorted_sf4.ix[:,1:])\n",
    "noresp_cohort = format_cohort(df_pilot_primary_NR_sorted_sf4.ix[:,1:], df_ext_primary_NR_sorted_sf4.ix[:,1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===============================================\n",
    "# Clinical annotation 2: All Samples - HPV Status\n",
    "## ===============================================\n",
    "resp_hpv_pilot = format_hpv(info_pilot, df_pilot_primary_R_sorted_sf4.ix[:,1:])\n",
    "resp_hpv_ext = format_hpv(info_extension, df_ext_primary_R_sorted_sf4.ix[:,1:])\n",
    "noresp_hpv_pilot = format_hpv(info_pilot, df_pilot_primary_NR_sorted_sf4.ix[:,1:])\n",
    "noresp_hpv_ext = format_hpv(info_extension, df_ext_primary_NR_sorted_sf4.ix[:,1:])\n",
    "\n",
    "resp_hpv = pd.concat([resp_hpv_pilot, resp_hpv_ext], axis = 1)\n",
    "noresp_hpv = pd.concat([noresp_hpv_pilot, noresp_hpv_ext], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===============================================\n",
    "# Genomic annotation 1: All Samples - 3q CNV\n",
    "## ===============================================\n",
    "resp_3q_pilot = format_gistic_3q(gistic_pilot_3q, df_pilot_primary_R_sorted_sf4.ix[:,1:])\n",
    "resp_3q_ext = format_gistic_3q(gistic_ext_3q, df_ext_primary_R_sorted_sf4.ix[:,1:])\n",
    "noresp_3q_pilot = format_gistic_3q(gistic_pilot_3q, df_pilot_primary_NR_sorted_sf4.ix[:,1:])\n",
    "noresp_3q_ext = format_gistic_3q(gistic_ext_3q, df_ext_primary_NR_sorted_sf4.ix[:,1:])\n",
    "\n",
    "resp_3q = pd.concat([resp_3q_pilot, resp_3q_ext], axis = 1)\n",
    "noresp_3q = pd.concat([noresp_3q_pilot, noresp_3q_ext], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ==================================================\n",
    "# Genomic annotation 2: All Samples - Mutations / MB\n",
    "## ==================================================\n",
    "\n",
    "# We calculate the mutations / Mb for all samples illustrated\n",
    "# For the pilot cohort, MutSig provides mutations / Mb of both total and nonsyn mutations\n",
    "# For the extension cohort, we calculate an estimate of nonsyn mutational burden\n",
    "\n",
    "# Pilot Cohort\n",
    "resp_burden_pilot = format_pilot_load(mutsig_rates, df_pilot_primary_R_sorted_sf4.ix[:,1:])\n",
    "noresp_burden_pilot = format_pilot_load(mutsig_rates, df_pilot_primary_NR_sorted_sf4.ix[:,1:])\n",
    "\n",
    "# Extension Cohort\n",
    "extension_all_calls = pd.read_csv('filter_unmatched/all_genes/df_filter_pass.txt', sep = '\\t', low_memory = False)\n",
    "\n",
    "resp_burden_ext = format_extension_load(info_extension, extension_all_calls, df_ext_primary_R_sorted_sf4.ix[:,1:])\n",
    "noresp_burden_ext = format_extension_load(info_extension, extension_all_calls, df_ext_primary_NR_sorted_sf4.ix[:,1:])\n",
    "\n",
    "resp_burden = pd.concat([resp_burden_pilot, resp_burden_ext], axis = 1)\n",
    "noresp_burden = pd.concat([noresp_burden_pilot, noresp_burden_ext], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===========================================================\n",
    "# Supplementary Figure 4: CoMut of Primary / Recurrent Tumors\n",
    "## ===========================================================\n",
    "\n",
    "fig = plt.figure(figsize = (22,16))\n",
    "gs = gridspec.GridSpec(5, 3, height_ratios = [1, 1, 2*display_S4, 1, display_S4/2], \n",
    "                      width_ratios = [len(data_00), len(data_01), len(data_00)/4])\n",
    "gs.update(wspace = 0.08, hspace = 0.04)\n",
    "\n",
    "##============================================\n",
    "# ax20 - CoMut - Response Primary Samples\n",
    "##============================================\n",
    "ax20, cax20 = plot_comut(2, 0, data_00, list_genes, 1, 'italic', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "##============================================\n",
    "# ax21 - CoMut - No Response Primary Samples\n",
    "##============================================\n",
    "ax21, cax21 = plot_comut(2, 1, data_01, list_genes, 0, 'normal', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "##============================================\n",
    "# ax00 - Cohort Status - Response Primary Samples\n",
    "##============================================\n",
    "ax00, cax00 = plot_comut(0, 0, resp_cohort, resp_cohort.index.tolist(), 1, 'normal', \n",
    "                         cmap_cohort, norm_cohort, 'black', sharex = ax20)\n",
    "\n",
    "plt.title('No Recurrence', y = 1.15, fontsize = 24)\n",
    "\n",
    "##============================================\n",
    "# ax01 - Cohort Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax01, cax01 = plot_comut(0, 1, noresp_cohort, noresp_cohort.index.tolist(), 0, 'normal', \n",
    "                         cmap_cohort, norm_cohort, 'black', sharex = ax21, sharey = ax00)\n",
    "\n",
    "plt.title('Recurrence', y = 1.15, fontsize = 24)\n",
    "\n",
    "##============================================\n",
    "# ax10 - Hpv Status - Response Primary Samples\n",
    "##============================================\n",
    "ax10, cax10 = plot_comut(1, 0, resp_hpv, resp_hpv.index.tolist(), 1, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax20)\n",
    "\n",
    "##============================================\n",
    "# ax11 - Hpv Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax11, cax11 = plot_comut(1, 1, noresp_hpv, resp_hpv.index.tolist(), 0, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax21, sharey = ax10)\n",
    "\n",
    "##============================================\n",
    "# ax30 - 3q Status - Response Primary Samples\n",
    "##============================================\n",
    "ax30, cax30 = plot_comut(3, 0, resp_3q, resp_3q.index.tolist(), 1, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax20)\n",
    "\n",
    "##============================================\n",
    "# ax31 - 3q Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax31, cax31 = plot_comut(3, 1, noresp_3q, noresp_3q.index.tolist(), 0, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax20, sharey = ax30)\n",
    "\n",
    "##==============================================\n",
    "# ax40 - Mutations / Mb - Response Primary Samples\n",
    "##==============================================\n",
    "ax40 = plot_burden(4, 0, resp_burden, 1, colors_bar, 10, 40, \n",
    "                                      display_label = 'Mutations / Mb', sharex = ax20)\n",
    "\n",
    "##==============================================\n",
    "# ax41 - Mutations / Mb - No Response Primary Samples\n",
    "##==============================================\n",
    "ax41 = plot_burden(4, 1, noresp_burden, 0, colors_bar, 10, 40,\n",
    "                                     sharex = ax21, sharey = ax40)\n",
    "\n",
    "##============================================\n",
    "# ax02 - Legend - Cohort Status\n",
    "##============================================\n",
    "labels = ['Pilot', 'Extension']\n",
    "ax02 = create_legend(0, 2, labels, colors_cohort, 1.65, 1.4, 2)\n",
    "\n",
    "##============================================\n",
    "# ax12 - Legend - HPV Status\n",
    "##============================================\n",
    "labels = [' Neg ', 'Positive']\n",
    "ax12 = create_legend(1, 2, labels, colors_tableauHPV, 1.6, 1.4, 2)\n",
    "\n",
    "##============================================\n",
    "# ax22 - Legend - Mutation Type\n",
    "##============================================\n",
    "ax22 = plt.subplot(gs[2,2])\n",
    "\n",
    "# Remove plot frame lines\n",
    "ax22.spines['top'].set_visible(False)\n",
    "ax22.spines['bottom'].set_visible(False)\n",
    "ax22.spines['right'].set_visible(False)\n",
    "ax22.spines['left'].set_visible(False)\n",
    "\n",
    "plt.tick_params(axis = 'both', which = 'both', \n",
    "    bottom = 'off', top = 'off', left = 'off', right = 'off',\n",
    "    labelbottom = 'off', labelleft = 'off', labelright = 'off')\n",
    "\n",
    "labels = [\"Syn.\", \"Missense\", \"Nonsense\", \"Inframe indel\"]\n",
    "\n",
    "ax22.bar(0, 0, label = labels[0], color = colors_fig1_comut[1]) # Syn\n",
    "ax22.bar(1, 0, label = labels[1], color = colors_fig1_comut[2]) # Missense\n",
    "ax22.bar(2, 0, label = labels[2], color = colors_fig1_comut[4]) # Nonsense\n",
    "ax22.bar(4, 0, label = labels[3], color = colors_fig1_comut[6]) # Inframe Indel\n",
    "\n",
    "plt.ylim(0)\n",
    "\n",
    "legend = ax22.legend(bbox_to_anchor = (1.15, 1), ncol = 1, handlelength = 3, fontsize = 14, columnspacing = 1, \n",
    "               frameon = False)\n",
    "legend.get_title().set_fontsize('14')\n",
    "\n",
    "##============================================\n",
    "# ax32 - Legend - 3q CNV status\n",
    "##============================================\n",
    "#labels = [\"Loss\", \"Gain\"]\n",
    "#ax32 = create_legend(3, 2, labels, colors_tableau3q, 1.0, 1.0, 2)\n",
    "ax32 = plt.subplot(gs[3,2])\n",
    "\n",
    "# Remove plot frame lines\n",
    "ax32.spines['top'].set_visible(False)\n",
    "ax32.spines['bottom'].set_visible(False)\n",
    "ax32.spines['right'].set_visible(False)\n",
    "ax32.spines['left'].set_visible(False)\n",
    "\n",
    "plt.tick_params(axis = 'both', which = 'both', \n",
    "    bottom = 'off', top = 'off', left = 'off', right = 'off',\n",
    "    labelbottom = 'off', labelleft = 'off', labelright = 'off')\n",
    "\n",
    "labels = [\"Loss\", \"Gain\"]\n",
    "\n",
    "ax32.bar(0, 0, label = labels[0], color = colors_tableau3q[1]) # Loss\n",
    "ax32.bar(1, 0, label = labels[1], color = colors_tableau3q[3]) # Gain\n",
    "\n",
    "plt.ylim(0)\n",
    "\n",
    "legend = ax32.legend(bbox_to_anchor = (1.4, 1.25), ncol = 2, handlelength = 3, fontsize = 14, columnspacing = 1, \n",
    "               frameon = False)\n",
    "legend.get_title().set_fontsize('14')\n",
    "\n",
    "##============================================\n",
    "# ax42 - Legend - Mutational Burden\n",
    "##============================================\n",
    "labels = [\"Non Syn\", \"Syn\"]\n",
    "ax42 = create_legend(4, 2, labels, colors_bar, 1.52, 1.0, 2)\n",
    "\n",
    "plt.savefig('figures/figureS4.png', bbox_inches = 'tight')\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "<a id='supplementary.figure6a'></a>\n",
    "## Supplementary Figure 6a: CoMut - Primary and Recurrent Pairs\n",
    "<a href='#table.of.contents'>Return to top</a><br>\n",
    "\n",
    "In this section, we will produce Supplementary Figure 6a which features a co-mutation plot of primary and recurrent tumor pairs in the pilot and extension cohorts. Primary and recurrent pairs are shown for the pilot cohort on the left and the extension cohort on the right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ========================================================\n",
    "# Generate dataframes\n",
    "## ========================================================\n",
    "\n",
    "# We create our lists of samples for each cohort\n",
    "pilot_list = ['AC-P04-Tumor-SM-63B4O','AC-P04-Tumor-SM-63B4Q','AC-P09-Tumor-SM-63B4Z',\n",
    "              'AC-P09-Tumor-SM-63B52','AC-P10-Tumor-SM-63B53','AC-P10-Tumor-SM-63B55',\n",
    "              'AC-P12-Tumor-SM-63B58','AC-P12-Tumor-SM-63B59','AC-P15-Tumor-SM-63B5D',\n",
    "              'AC-P15-Tumor-SM-63B5F']\n",
    "\n",
    "extension_list = ['AC-V08-Tumor-SM-9J45Z','AC-V08-Tumor-SM-9J461','AC-V08-Tumor-SM-9J462',\n",
    "                  'AC-V09-Tumor-SM-9J463','AC-V09-Tumor-SM-9J464','AC-V10-Tumor-SM-9J465',\n",
    "                  'AC-V10-Tumor-SM-9J466','AC-V11-Tumor-SM-9J467','AC-V11-Tumor-SM-9J468',\n",
    "                  'AC-V12-Tumor-SM-9N4BH','AC-V12-Tumor-SM-9J46A','AC-V14-Tumor-SM-9J46C',\n",
    "                  'AC-V14-Tumor-SM-9J46D','AC-V15-Tumor-SM-9J46E','AC-V15-Tumor-SM-9J46F',\n",
    "                  'AC-V16-Tumor-SM-9J46G','AC-V16-Tumor-SM-9J46H','AC-V18-Tumor-SM-9J46J',\n",
    "                  'AC-V18-Tumor-SM-9J46K']\n",
    "\n",
    "df_pilot_pairs = populate(pilot_list, df_genes, variants_pilot)\n",
    "df_extension_pairs = populate(extension_list, df_genes, variants_extension)\n",
    "\n",
    "# Sort vertically by frequency\n",
    "df_freq_pilot_pairs = df_pilot_pairs.replace(0, np.nan)\n",
    "df_freq_extension_pairs = df_extension_pairs.replace(0, np.nan)\n",
    "\n",
    "freq_pairs = (df_freq_pilot_pairs.count(axis = 1, numeric_only = True) + \\\n",
    "             df_freq_extension_pairs.count(axis = 1, numeric_only = True))\n",
    "\n",
    "freq_pairs.sort_values(ascending = False, inplace = True)\n",
    "\n",
    "# We sort our dataframes across genes depending on frequency across both cohorts\n",
    "df_pilot_pairs_sorted = pd.DataFrame(df_pilot_pairs, index = freq_pairs.index)\n",
    "df_extension_pairs_sorted = pd.DataFrame(df_extension_pairs, index = freq_pairs.index)\n",
    "\n",
    "df_pilot_pairs_sorted.index = range(0, len(df_pilot_pairs_sorted))\n",
    "df_extension_pairs_sorted.index = range(0, len(df_extension_pairs_sorted))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Like before, I make an aethetic choice of only displaying some genes. The plot is less crowded now so I display more\n",
    "display_S6A = 18\n",
    "\n",
    "data_00 = df_pilot_pairs_sorted.iloc[:display_S6A, 1:]\n",
    "data_01 = df_extension_pairs_sorted.iloc[:display_S6A, 1:]\n",
    "\n",
    "list_genes = df_pilot_pairs_sorted.ix[:,'gene'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===============================================\n",
    "# Clinical annotation 1: All Samples - HPV Status\n",
    "## ===============================================\n",
    "hpv_pilot = format_hpv(info_pilot, df_pilot_pairs_sorted.ix[:,1:])\n",
    "hpv_extension = format_hpv(info_extension, df_extension_pairs_sorted.ix[:,1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===============================================\n",
    "# Genomic annotation 1: All Samples - 3q CNV\n",
    "## ===============================================\n",
    "gistic_pilot_pairs = format_gistic_3q(gistic_pilot_3q, df_pilot_pairs_sorted.ix[:,1:])\n",
    "gistic_extension_pairs = format_gistic_3q(gistic_ext_3q, df_extension_pairs_sorted.ix[:,1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## ==================================================\n",
    "# Genomic annotation 2: All Samples - Mutations / MB\n",
    "## ==================================================\n",
    "burden_pilot_pairs = format_pilot_load(mutsig_rates, df_pilot_pairs_sorted.ix[:,1:])\n",
    "burden_extension_pairs = format_extension_load(info_extension, extension_all_calls, df_extension_pairs_sorted.ix[:,1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## ===========================================================\n",
    "# Supplementary Figure 4: CoMut of Primary / Recurrent Tumors\n",
    "## ===========================================================\n",
    "fig = plt.figure(figsize = (22,16))\n",
    "gs = gridspec.GridSpec(4, 3, height_ratios = [1, 2*display_S6A, 1, display_S6A/2], \n",
    "                      width_ratios = [len(data_00), len(data_01), len(data_00)/4])\n",
    "gs.update(wspace = 0.08, hspace = 0.04)\n",
    "\n",
    "##============================================\n",
    "# ax10 - CoMut - Response Primary Samples\n",
    "##============================================\n",
    "ax10, cax10 = plot_comut(1, 0, data_00, list_genes, 1, 'italic', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "divider = [2,4,6,8,10]\n",
    "for i in range(0,len(divider)-1,1):\n",
    "    ax10.plot((divider[i] - 0.5, divider[i] - 0.5),(-0.5, display_S6A - 0.5),'k-', linewidth=2.0)    \n",
    "\n",
    "##============================================\n",
    "# ax11 - CoMut - No Response Primary Samples\n",
    "##============================================\n",
    "ax11, cax11 = plot_comut(1, 1, data_01, list_genes, 0, 'normal', cmap_fig1_comut, norm_fig1_comut, 'white')\n",
    "\n",
    "divider = [3,5,7,9,11,13,15,17]\n",
    "for i in range(0,len(divider)-1,1):\n",
    "    ax11.plot((divider[i] - 0.5, divider[i] - 0.5),(-0.50, display_S6A - 0.51),'k-', linewidth=2.0)\n",
    "\n",
    "##============================================\n",
    "# ax00 - Hpv Status - Response Primary Samples\n",
    "##============================================\n",
    "ax00, cax00 = plot_comut(0, 0, hpv_pilot, hpv_pilot.index.tolist(), 1, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax10)\n",
    "\n",
    "plt.title('Pilot Cohort', y = 1.05, fontsize = 20)\n",
    "\n",
    "##============================================\n",
    "# ax01 - Hpv Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax01, cax01 = plot_comut(0, 1, hpv_extension, hpv_extension.index.tolist(), 0, 'normal', \n",
    "                         cmapHPV, normHPV, 'black', sharex = ax11, sharey = ax00)\n",
    "\n",
    "plt.title('Extension Cohort', y = 1.05, fontsize = 20)\n",
    "\n",
    "##============================================\n",
    "# ax20 - 3q Status - Response Primary Samples\n",
    "##============================================\n",
    "ax20, cax20 = plot_comut(2, 0, gistic_pilot_pairs, gistic_pilot_pairs.index.tolist(), 1, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax10)\n",
    "\n",
    "##============================================\n",
    "# ax21 - 3q Status - No Response Primary Samples\n",
    "##============================================\n",
    "ax21, cax21 = plot_comut(2, 1, gistic_extension_pairs, gistic_extension_pairs.index.tolist(), 0, 'normal', \n",
    "                         cmap3q, norm3q, 'black', sharex = ax11, sharey = ax20)\n",
    "\n",
    "##==============================================\n",
    "# ax30 - Mutations / Mb - Response Primary Samples\n",
    "##==============================================\n",
    "ax30 = plot_burden(3, 0, burden_pilot_pairs, 1, colors_bar, 10, 25, \n",
    "                                      display_label = 'Mutations / Mb', sharex = ax10)\n",
    "\n",
    "##==============================================\n",
    "# ax31 - Mutations / Mb - No Response Primary Samples\n",
    "##==============================================\n",
    "ax31 = plot_burden(3, 1, burden_extension_pairs, 0, colors_bar, 10, 25,\n",
    "                                     sharex = ax11, sharey = ax30)\n",
    "\n",
    "##============================================\n",
    "# ax12 - Legend - Mutation Type\n",
    "##============================================\n",
    "ax12 = plt.subplot(gs[1,2])\n",
    "\n",
    "# Remove plot frame lines\n",
    "ax12.spines['top'].set_visible(False)\n",
    "ax12.spines['bottom'].set_visible(False)\n",
    "ax12.spines['right'].set_visible(False)\n",
    "ax12.spines['left'].set_visible(False)\n",
    "\n",
    "plt.tick_params(axis = 'both', which = 'both', \n",
    "    bottom = 'off', top = 'off', left = 'off', right = 'off',\n",
    "    labelbottom = 'off', labelleft = 'off', labelright = 'off')\n",
    "\n",
    "labels = [\"Synonymous\", \"Missense\", \"Splice Site\", \"Nonsense\", \"Frameshift\", \"InDel\"]\n",
    "\n",
    "ax12.bar(1, 0, label = labels[1], color = colors_fig1_comut[2]) # Missense\n",
    "ax12.bar(4, 0, label = labels[3], color = colors_fig1_comut[4]) # Nonsense\n",
    "ax12.bar(4, 0, label = labels[5], color = colors_fig1_comut[6]) # Indel\n",
    "\n",
    "plt.ylim(0)\n",
    "\n",
    "legend = ax12.legend(bbox_to_anchor = (0.9, 0.90), ncol = 1, handlelength = 3, fontsize = 14, columnspacing = 1, \n",
    "               frameon = False)\n",
    "\n",
    "##============================================\n",
    "# ax02 - Legend - HPV Status\n",
    "##============================================\n",
    "labels = [' Neg ', 'Positive', 'Unknown']\n",
    "ax02 = create_legend(0, 2, labels, colors_tableauHPV, 1.75, 1.75, 2)\n",
    "\n",
    "##============================================\n",
    "# ax22 - Legend - 3q CNV status\n",
    "##============================================\n",
    "labels = [\"Loss\", \"Gain\"]\n",
    "ax22 = plt.subplot(gs[2,2])\n",
    "\n",
    "# Remove plot frame lines\n",
    "ax22.spines['top'].set_visible(False)\n",
    "ax22.spines['bottom'].set_visible(False)\n",
    "ax22.spines['right'].set_visible(False)\n",
    "ax22.spines['left'].set_visible(False)\n",
    "\n",
    "plt.tick_params(axis = 'both', which = 'both', \n",
    "    bottom = 'off', top = 'off', left = 'off', right = 'off',\n",
    "    labelbottom = 'off', labelleft = 'off', labelright = 'off')\n",
    "\n",
    "ax22.bar(0, 0, label = labels[0], color = colors_tableau3q[1]) # Loss\n",
    "ax22.bar(1, 0, label = labels[1], color = colors_tableau3q[2]) # Gain\n",
    "\n",
    "plt.ylim(0)\n",
    "\n",
    "legend = ax22.legend(bbox_to_anchor = (1.42, 1.70), ncol = 2, handlelength = 3, fontsize = 14, columnspacing = 1, \n",
    "               frameon = False)\n",
    "\n",
    "##============================================\n",
    "# ax32 - Legend - Mutational Burden\n",
    "##============================================\n",
    "labels = [\"Nonsyn burden\", \"Syn burden\"]\n",
    "ax32 = create_legend(3, 2, labels, colors_bar, 1.25, 1.0, 1)\n",
    "\n",
    "\n",
    "plt.savefig('figures/figureS6A.png', bbox_inches = 'tight')\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Calculations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='methods.tumoronly.filter'></a>\n",
    "## Methods: Applying tumor only filter to the pilot cohort\n",
    "<a href='#table.of.contents'>Return to top</a><br>\n",
    "\n",
    "For validation and metric purposes, we applied the tumor only filter to the pilot cohort's unmatched tumor pairs to the matched tumor pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# We import our golden standard, the matched tumor samples\n",
    "allfiles = glob.glob('storage/mafs/pilot_snv/*')\n",
    "df_master_ = pd.DataFrame()\n",
    "list_ = []\n",
    "\n",
    "for file_ in allfiles:\n",
    "    df_ = pd.read_csv(file_, index_col = None, sep = '\\t', \n",
    "                     comment = '#', low_memory = False)\n",
    "    list_.append(df_)\n",
    "df_master_ = pd.concat(list_, ignore_index = True)  \n",
    "\n",
    "# To compare apples to apples, we shrink the reference to only Missense, Nonsense, and Splice site mutations\n",
    "df_master = df_master_[df_master_['Variant_Classification'].isin(['Missense_Mutation', 'Nonsense_Mutation', 'Splice_Site'])]\n",
    "df_master.index = range(0, len(df_master))\n",
    "\n",
    "# We import the comparison, the unmatched tumor samples post-filtering\n",
    "df_passed = pd.read_csv('filter_unmatched/pilot_all_genes/df_filter_pass.txt', sep = '\\t', \n",
    "                       comment = '#', low_memory = False)\n",
    "df_failed = pd.read_csv('filter_unmatched/pilot_all_genes/df_filter_failed.txt', sep = '\\t', \n",
    "                       comment = '#', low_memory = False)\n",
    "\n",
    "print 'We see a', str(format(round((len(df_passed) + len(df_failed))/len(df_master)*100,2))), '% increase in mutation calls in the pilot cohort when considering unmatched vs matched tumor samples, after upstream filtering was performed (such as PoN and FFPE filters)'\n",
    "print 'Specifically, we observe', str(len(df_passed) + len(df_failed)), 'variants being called in the tumor only case and', str(len(df_master)), 'for matched pairs.'\n",
    "print ''\n",
    "print 'After applying our novel tumor only filtering on the unmatched samples, this number shrinks.'\n",
    "print str(len(df_passed)), 'variants passed our filter and', str(len(df_failed)), 'failed.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# For some reason that I could not discover, the start_position column of the passed dataframe was being processed as a float\n",
    "# so we convert it to an integer\n",
    "df_passed.ix[:, 'Start_position'] = df_passed.ix[:,'Start_position'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## For all genes\n",
    "# For quick processing of comparison, we combine identifying columns into one series\n",
    "\n",
    "id_master = df_master['Chromosome'].map(str) + df_master['Start_position'].map(str) + df_master['Tumor_Sample_Barcode'] + \\\n",
    "    df_master['Reference_Allele'] + df_master['Tumor_Seq_Allele1'] + df_master['Tumor_Seq_Allele2']\n",
    "id_passed = df_passed['Chromosome'].map(str) + df_passed['Start_position'].fillna(0.0).map(str) + df_passed['Tumor_Sample_Barcode'] + \\\n",
    "    df_passed['Reference_Allele'] + df_passed['Tumor_Seq_Allele1'] + df_passed['Tumor_Seq_Allele2']\n",
    "id_failed = df_failed['Chromosome'].map(str) + df_failed['Start_position'].fillna(0.0).map(str) + df_failed['Tumor_Sample_Barcode'] + \\\n",
    "    df_failed['Reference_Allele'] + df_failed['Tumor_Seq_Allele1'] + df_failed['Tumor_Seq_Allele2'] \n",
    "    \n",
    "# Calculate True Positives, False Positives, False Negatives, TrueNegatives\n",
    "idx_tp = (id_master).isin(id_passed)\n",
    "TruePositive = len(df_master[idx_tp])         # Correctly called, post-filtering\n",
    "idx_tn = (id_master).isin(id_failed)\n",
    "TrueNegative = len(df_master[~idx_tn])        # Correctly filtered, post-filtering\n",
    "\n",
    "idx_fp = (id_passed).isin(id_master)\n",
    "FalsePositive = len(df_passed[~idx_fp])       # Incorrectly called, post-filtering\n",
    "idx_fn = (id_failed).isin(id_master)\n",
    "FalseNegative = len(df_failed[idx_fn])        # Incorrectly filtered, post-filtering\n",
    "\n",
    "# Calculate PPV and Sensitivity\n",
    "PPV = TruePositive / (TruePositive + FalsePositive)\n",
    "Sensitivity = TruePositive / (TruePositive + FalseNegative)\n",
    "\n",
    "print 'Our filtering approach yields the following results when applied to the pilot cohort, when considering all genes:'\n",
    "print 'PPV:', str(PPV)\n",
    "print 'Sensitivity:', str(Sensitivity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## For 268 genes\n",
    "genes_tcga = pd.read_csv('storage/reference/tcga_genes.txt', sep = '\\t')\n",
    "\n",
    "df_master = df_master[df_master['Hugo_Symbol'].isin(genes_tcga['gene'])]\n",
    "df_passed = df_passed[df_passed['Hugo_Symbol'].isin(genes_tcga['gene'])]\n",
    "df_failed = df_failed[df_failed['Hugo_Symbol'].isin(genes_tcga['gene'])]\n",
    "\n",
    "# For quick processing of comparison, we combine identifying columns into one series\n",
    "\n",
    "id_master = df_master['Chromosome'].map(str) + df_master['Start_position'].map(str) + df_master['Tumor_Sample_Barcode'] + \\\n",
    "    df_master['Reference_Allele'] + df_master['Tumor_Seq_Allele1'] + df_master['Tumor_Seq_Allele2']\n",
    "id_passed = df_passed['Chromosome'].map(str) + df_passed['Start_position'].fillna(0.0).map(str) + df_passed['Tumor_Sample_Barcode'] + \\\n",
    "    df_passed['Reference_Allele'] + df_passed['Tumor_Seq_Allele1'] + df_passed['Tumor_Seq_Allele2']\n",
    "id_failed = df_failed['Chromosome'].map(str) + df_failed['Start_position'].fillna(0.0).map(str) + df_failed['Tumor_Sample_Barcode'] + \\\n",
    "    df_failed['Reference_Allele'] + df_failed['Tumor_Seq_Allele1'] + df_failed['Tumor_Seq_Allele2'] \n",
    "    \n",
    "# Calculate True Positives, False Positives, False Negatives, TrueNegatives\n",
    "idx_tp = (id_master).isin(id_passed)\n",
    "TruePositive = len(df_master[idx_tp])         # Correctly called, post-filtering\n",
    "idx_tn = (id_master).isin(id_failed)\n",
    "TrueNegative = len(df_master[~idx_tn])        # Correctly filtered, post-filtering\n",
    "\n",
    "idx_fp = (id_passed).isin(id_master)\n",
    "FalsePositive = len(df_passed[~idx_fp])       # Incorrectly called, post-filtering\n",
    "idx_fn = (id_failed).isin(id_master)\n",
    "FalseNegative = len(df_failed[idx_fn])        # Incorrectly filtered, post-filtering\n",
    "\n",
    "# Calculate PPV and Sensitivity\n",
    "PPV = TruePositive / (TruePositive + FalsePositive)\n",
    "Sensitivity = TruePositive / (TruePositive + FalseNegative)\n",
    "\n",
    "print 'Our filtering approach yields the following results when applied to the pilot cohort, when considering a genes reported in TCGA(n=268):'\n",
    "print 'PPV:', str(PPV)\n",
    "print 'Sensitivity:', str(Sensitivity)\n",
    "print '* TCGA genes obtained from tumorportal.org'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<a id='results.primary.ascc'></a>\n",
    "## Results: Genomic Landscape of Primary ASCC\n",
    "<a href='#table.of.contents'>Return to top</a><br>\n",
    "\n",
    "We will replicate any calculations cited in this section of the paper, this primarily contains reports from MutSig for the pilot cohort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "input_info_pilot = r'storage/reference/ASCC_samples-pilot.txt'\n",
    "info_pilot = pd.read_csv(input_info_pilot, sep = '\\t')\n",
    "\n",
    "list_pilot_primary = info_pilot[info_pilot['Tumor Type'] == 'Primary'].ix[:,'Tumor Sample ID'].tolist()\n",
    "list_pilot_recurrent = info_pilot[info_pilot['Tumor Type'] != 'Primary'].ix[:,'Tumor Sample ID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## MutSig breakdown for the pilot cohort (both primary and recurrent tumors)\n",
    "# Just in case this section is run independently, we re-import MutSig results\n",
    "df_pilot = pd.read_csv('storage/mutsig-pilot/final_analysis_set.maf', sep = '\\t',\n",
    "                      comment = '#', low_memory = False)\n",
    "\n",
    "pilot_missense = df_pilot[df_pilot['Variant_Classification'] == 'Missense_Mutation']\n",
    "pilot_nonsense = df_pilot[df_pilot['Variant_Classification'] == 'Nonsense_Mutation']\n",
    "pilot_frameshift = df_pilot[df_pilot['Variant_Classification'].isin(['Frame_Shift_Del', 'Frame_Shift_Ins'])]\n",
    "pilot_silent = df_pilot[df_pilot['Variant_Classification'] == 'Silent']\n",
    "\n",
    "print 'A total of', str(len(df_pilot)), 'somatic mutations were found after running MutSig on the pilot cohort, this include:'\n",
    "print str(len(pilot_missense)), 'missense mutations'\n",
    "print str(len(pilot_nonsense)), 'nonsense mutations'\n",
    "print str(len(pilot_frameshift)), 'frameshift mutations'\n",
    "print str(len(pilot_silent)), 'silent mutations \\n'\n",
    "\n",
    "print 'These values reflect mutations reported for both primary and recurrent samples'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Mutational burden for the pilot cohort (Repeat calculations from Figure 3)\n",
    "mutsig_burden = pd.read_csv('storage/mutsig-pilot/patient_counts_and_rates.txt', sep = '\\t')\n",
    "\n",
    "pilot_burden_mean = mutsig_burden.ix[:,'rate_tot'].mean()\n",
    "pilot_burden_mean = pilot_burden_mean*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "pilot_burden_min = mutsig_burden.ix[:,'rate_tot'].min()\n",
    "pilot_burden_min = pilot_burden_min*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "pilot_burden_max = mutsig_burden.ix[:,'rate_tot'].max()\n",
    "pilot_burden_max = pilot_burden_max*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "print 'The mean somatic rate for the pilot cohort was', str(pilot_burden_mean)\n",
    "print 'with a range of', str(pilot_burden_min), 'to', str(pilot_burden_max), 'mutations per Mb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# We repeat the calculation for only primary samples\n",
    "## MutSig breakdown for the pilot cohort (both primary and recurrent tumors)\n",
    "# Just in case this section is run independently, we re-import MutSig results\n",
    "df_pilot = pd.read_csv('storage/mutsig-pilot/final_analysis_set.maf', sep = '\\t',\n",
    "                      comment = '#', low_memory = False)\n",
    "\n",
    "df_pilot_primary = df_pilot[df_pilot['Tumor_Sample_Barcode'].isin(list_pilot_primary)]\n",
    "\n",
    "pilot_missense_primary = df_pilot_primary[df_pilot_primary['Variant_Classification'] == 'Missense_Mutation']\n",
    "pilot_nonsense_primary = df_pilot_primary[df_pilot_primary['Variant_Classification'] == 'Nonsense_Mutation']\n",
    "pilot_frameshift_primary = df_pilot_primary[df_pilot_primary['Variant_Classification'].isin(['Frame_Shift_Del', 'Frame_Shift_Ins'])]\n",
    "pilot_silent_primary = df_pilot_primary[df_pilot_primary['Variant_Classification'] == 'Silent']\n",
    "\n",
    "print 'A total of', str(len(df_pilot_primary)), 'somatic mutations were found after running MutSig on the pilot cohort primary samples, this include:'\n",
    "print str(len(pilot_missense_primary)), 'missense mutations'\n",
    "print str(len(pilot_nonsense_primary)), 'nonsense mutations'\n",
    "print str(len(pilot_frameshift_primary)), 'frameshift mutations'\n",
    "print str(len(pilot_silent_primary)), 'silent mutations \\n'\n",
    "\n",
    "## Mutational burden for the pilot cohort (Repeat calculations from Figure 3)\n",
    "mutsig_burden_primary = mutsig_burden[mutsig_burden['name'].isin(list_pilot_primary)]\n",
    "\n",
    "pilot_burden_mean_primary = mutsig_burden_primary.ix[:,'rate_tot'].mean()\n",
    "pilot_burden_mean_primary = pilot_burden_mean_primary*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "pilot_burden_min_primary = mutsig_burden_primary.ix[:,'rate_tot'].min()\n",
    "pilot_burden_min_primary = pilot_burden_min_primary*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "pilot_burden_max_primary = mutsig_burden_primary.ix[:,'rate_tot'].max()\n",
    "pilot_burden_max_primary = pilot_burden_max_primary*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "print 'The mean somatic rate for the pilot cohort was', str(pilot_burden_mean_primary)\n",
    "print 'with a range of', str(pilot_burden_min_primary), 'to', str(pilot_burden_max_primary), 'mutations per Mb'\n",
    "\n",
    "print 'These values reflect mutations reported for primary samples'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# We repeat the calculation for only primary samples\n",
    "## MutSig breakdown for the pilot cohort (both primary and recurrent tumors)\n",
    "# Just in case this section is run independently, we re-import MutSig results\n",
    "df_pilot = pd.read_csv('storage/mutsig-pilot/final_analysis_set.maf', sep = '\\t',\n",
    "                      comment = '#', low_memory = False)\n",
    "\n",
    "df_pilot_recurrent = df_pilot[df_pilot['Tumor_Sample_Barcode'].isin(list_pilot_recurrent)]\n",
    "\n",
    "pilot_missense_recurrent = df_pilot_recurrent[df_pilot_recurrent['Variant_Classification'] == 'Missense_Mutation']\n",
    "pilot_nonsense_recurrent = df_pilot_recurrent[df_pilot_recurrent['Variant_Classification'] == 'Nonsense_Mutation']\n",
    "pilot_frameshift_recurrent = df_pilot_recurrent[df_pilot_recurrent['Variant_Classification'].isin(['Frame_Shift_Del', 'Frame_Shift_Ins'])]\n",
    "pilot_silent_recurrent = df_pilot_recurrent[df_pilot_recurrent['Variant_Classification'] == 'Silent']\n",
    "\n",
    "print 'A total of', str(len(df_pilot_recurrent)), 'somatic mutations were found after running MutSig on the pilot cohort recurrent samples, this include:'\n",
    "print str(len(pilot_missense_recurrent)), 'missense mutations'\n",
    "print str(len(pilot_nonsense_recurrent)), 'nonsense mutations'\n",
    "print str(len(pilot_frameshift_recurrent)), 'frameshift mutations'\n",
    "print str(len(pilot_silent_recurrent)), 'silent mutations \\n'\n",
    "\n",
    "## Mutational burden for the pilot cohort (Repeat calculations from Figure 3)\n",
    "mutsig_burden_recurrent = mutsig_burden[mutsig_burden['name'].isin(list_pilot_recurrent)]\n",
    "\n",
    "pilot_burden_mean_recurrent = mutsig_burden_recurrent.ix[:,'rate_tot'].mean()\n",
    "pilot_burden_mean_recurrent = pilot_burden_mean_recurrent*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "pilot_burden_min_recurrent = mutsig_burden_recurrent.ix[:,'rate_tot'].min()\n",
    "pilot_burden_min_recurrent = pilot_burden_min_recurrent*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "pilot_burden_max_recurrent = mutsig_burden_recurrent.ix[:,'rate_tot'].max()\n",
    "pilot_burden_max_recurrent = pilot_burden_max_recurrent*10*10*10*10*10*10 # Convert from per base to per Mb\n",
    "\n",
    "print 'The mean somatic rate for the pilot cohort was', str(pilot_burden_mean_recurrent)\n",
    "print 'with a range of', str(pilot_burden_min_recurrent), 'to', str(pilot_burden_max_recurrent), 'mutations per Mb'\n",
    "\n",
    "print 'These values reflect mutations reported for recurrent samples'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='results.recurrent.ascc'></a>\n",
    "## Results: Genomic Landscape of Recurrent ASCC\n",
    "<a href='#table.of.contents'>Return to top</a><br>\n",
    "\n",
    "We will replicate any calculations cited in this section of the paper, this primarily contains calculations that were performed for figure 1. We calculate the mutational burden for all primary and recurrent samples and then compare them with paired t-tests using the Python package scipy, we do not assume equal variance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# First and foremost, we import some files that contain annotations\n",
    "input_info_pilot = r'storage/reference/ASCC_samples-pilot.txt'\n",
    "info_pilot = pd.read_csv(input_info_pilot, sep = '\\t')\n",
    "\n",
    "input_info_extension = r'storage/reference/ASCC_samples-extension.txt'\n",
    "info_extension = pd.read_csv(input_info_extension, sep = '\\t')\n",
    "\n",
    "# Create lists of primary and recurrent tumors from each cohort\n",
    "list_pilot_primary = info_pilot[info_pilot['Tumor Type'] == 'Primary'].ix[:,'Tumor Sample ID'].tolist()\n",
    "list_pilot_recurrent = info_pilot[info_pilot['Tumor Type'] != 'Primary'].ix[:,'Tumor Sample ID'].tolist()\n",
    "\n",
    "list_extension_primary = info_extension[info_extension['Tumor Type'] == 'Primary'].ix[:,'Tumor Sample ID'].tolist()\n",
    "list_extension_recurrent = info_extension[info_extension['Tumor Type'] != 'Primary'].ix[:,'Tumor Sample ID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Because of how I wrote the functions \"format_pilot_load\" and \"format_extension_load\", \n",
    "# I will move the lists created above into dataframes\n",
    "df_list_pilot_primary = pd.DataFrame([], columns = list_pilot_primary)\n",
    "df_list_pilot_recurrent = pd.DataFrame([], columns = list_pilot_recurrent)\n",
    "\n",
    "df_list_extension_primary = pd.DataFrame([], columns = list_extension_primary)\n",
    "df_list_extension_recurrent = pd.DataFrame([], columns = list_extension_recurrent)\n",
    "\n",
    "# Import files necessary to compute mutational burden\n",
    "pilot_mutsig_rates = pd.read_csv('storage/mutsig-pilot/patient_counts_and_rates.txt', sep = '\\t')\n",
    "extension_all_calls = pd.read_csv('filter_unmatched/all_genes/df_filter_pass.txt', sep = '\\t', low_memory = False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Compute mutational burden\n",
    "# This is very similar to the calculatiosn done in Figure 1 and Figure 3 but we are now performing \n",
    "# the calculation for all samples, rather than just the ones required for those figures\n",
    "\n",
    "df_pilot_primary = format_pilot_load(pilot_mutsig_rates, df_list_pilot_primary)\n",
    "df_pilot_recurrent = format_pilot_load(pilot_mutsig_rates, df_list_pilot_recurrent)\n",
    "\n",
    "df_extension_primary = format_extension_load(info_extension, extension_all_calls, df_list_extension_primary)\n",
    "df_extension_recurrent = format_extension_load(info_extension, extension_all_calls, df_list_extension_recurrent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# We now perform some comparison calculations using the package scipy for paired t-tests\n",
    "# We do not assume equal variance\n",
    "\n",
    "print 'We compare the mutational burden in the pilot cohort \\n'\n",
    "\n",
    "statistic, pvalue = stats.ttest_ind(df_pilot_primary.ix['Total', :], df_pilot_recurrent.ix['Total', :], equal_var = False )\n",
    "print 'Total mutational burden:', str(statistic), str(pvalue)\n",
    "statistic, pvalue = stats.ttest_ind(df_pilot_primary.ix['Nonsyn', :], df_pilot_recurrent.ix['Nonsyn', :], equal_var = False )\n",
    "print 'Nonsyn mutational burden:', str(statistic), str(pvalue)\n",
    "statistic, pvalue = stats.ttest_ind(df_pilot_primary.ix['Syn', :], df_pilot_recurrent.ix['Syn', :], equal_var = False )\n",
    "print 'Syn mutational burden:', str(statistic), str(pvalue)\n",
    "\n",
    "print '\\n'\n",
    "\n",
    "print 'We compare the mutational burden in the extension cohort \\n'\n",
    "\n",
    "statistic, pvalue = stats.ttest_ind(df_extension_primary.ix['Nonsyn', :], df_extension_recurrent.ix['Nonsyn', :], equal_var = False )\n",
    "print 'Nonsyn mutational burden:', str(statistic), str(pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## We have estimated the mutational burden for the unmatched samples by estimating the \n",
    "# nonsyn mutational burden. We compare the extension cohort's estimated burden\n",
    "# to both the pilot cohort's total mutational burden and nonsyn burden\n",
    "\n",
    "print 'We compare the mutational burden across the extension cohort to the pilot cohort \\n'\n",
    "print 'For primary tumors...'\n",
    "\n",
    "statistic, pvalue = stats.ttest_ind(df_pilot_primary.ix['Total', :], df_extension_primary.ix['Nonsyn', :], equal_var = False)\n",
    "print 'Total burden of the pilot cohort:', str(statistic), str(pvalue)\n",
    "statistic, pvalue = stats.ttest_ind(df_pilot_primary.ix['Nonsyn', :], df_extension_primary.ix['Nonsyn', :], equal_var = False)\n",
    "print 'Nonsyn burden of the pilot cohort:', str(statistic), str(pvalue)\n",
    "\n",
    "print ''\n",
    "\n",
    "print 'For recurrent tumors...'\n",
    "statistic, pvalue = stats.ttest_ind(df_pilot_recurrent.ix['Total', :], df_extension_recurrent.ix['Nonsyn', :], equal_var = False)\n",
    "print 'Total burden of the pilot cohort:', str(statistic), str(pvalue)\n",
    "statistic, pvalue = stats.ttest_ind(df_pilot_recurrent.ix['Nonsyn', :], df_extension_recurrent.ix['Nonsyn', :], equal_var = False)\n",
    "print 'Nonsyn burden of the pilot cohort:', str(statistic), str(pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## We now combine populations of primary and recurrent tumors together between the pilot\n",
    "# and extension cohort\n",
    "\n",
    "# We combine the total burden of the primary cohort with nonsyn burden of the extension cohort\n",
    "total_primary = df_pilot_primary.ix['Total', :].tolist() + df_extension_primary.ix['Nonsyn', :].tolist()\n",
    "total_recurrent = df_pilot_recurrent.ix['Total', :].tolist() + df_extension_recurrent.ix['Nonsyn', :].tolist()\n",
    "\n",
    "# We combine the total burden of the primary cohort with nonsyn burden of the extension cohort\n",
    "nonsyn_primary = df_pilot_primary.ix['Nonsyn', :].tolist() + df_extension_primary.ix['Nonsyn', :].tolist()\n",
    "nonsyn_recurrent = df_pilot_recurrent.ix['Nonsyn', :].tolist() + df_extension_recurrent.ix['Nonsyn', :].tolist()\n",
    "\n",
    "print 'We compare the mutational burden of primary and recurrent tumors across both cohorts \\n'\n",
    "\n",
    "statistic, pvalue = stats.ttest_ind(total_primary, total_recurrent, equal_var = False)\n",
    "print 'Total burden of the pilot cohort:', str(statistic), str(pvalue)\n",
    "statistic, pvalue = stats.ttest_ind(nonsyn_primary, nonsyn_recurrent, equal_var = False)\n",
    "print 'Nonsyn burden of the pilot cohort:', str(statistic), str(pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print 'Total burden of the pilot cohort with the estimated burden of the extension cohort,'\n",
    "print 'Mean burden of primary tumors:', str(np.mean(total_primary))\n",
    "print 'Mean burden of recurrent tumors:', str(np.mean(total_recurrent))\n",
    "print 'Median burden of primary tumors:', str(np.median(total_primary))\n",
    "print 'Median burden of recurrent tumors:', str(np.median(total_recurrent))\n",
    "\n",
    "print '\\n'\n",
    "\n",
    "print 'Nonsyn burden of the pilot cohort with the estimated burden of the extension cohort,'\n",
    "print 'Mean burden of primary tumors:', str(np.mean(nonsyn_primary))\n",
    "print 'Mean burden of recurrent tumors:', str(np.mean(nonsyn_recurrent))\n",
    "print 'Median burden of primary tumors:', str(np.median(nonsyn_primary))\n",
    "print 'Median burden of recurrent tumors:', str(np.median(nonsyn_recurrent))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='results.paired.ascc'></a>\n",
    "## Results: Genomic evolution in matched primary and recurrent tumors\n",
    "<a href='#table.of.contents'>Return to top</a><br>\n",
    "\n",
    "We will replicate any calculations cited in this section of the paper, this primarily contains calculations that were performed for figure 3. We calculate the mutational burden for paired primary and recurrent samples and then compare them with paired t-tests using the Python package scipy, we do not assume equal variance. This section is very similar to the one above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# First and foremost, we import some files that contain annotations\n",
    "input_info_pilot = r'storage/reference/ASCC_samples-pilot.txt'\n",
    "info_pilot = pd.read_csv(input_info_pilot, sep = '\\t')\n",
    "\n",
    "input_info_extension = r'storage/reference/ASCC_samples-extension.txt'\n",
    "info_extension = pd.read_csv(input_info_extension, sep = '\\t')\n",
    "\n",
    "# Create lists of primary and recurrent tumors from each cohort\n",
    "# For convenience, I just manually create our lists\n",
    "list_pilot_primary = ['AC-P04-Tumor-SM-63B4O','AC-P09-Tumor-SM-63B4Z','AC-P10-Tumor-SM-63B53',\n",
    "                     'AC-P12-Tumor-SM-63B58','AC-P15-Tumor-SM-63B5D']\n",
    "list_pilot_recurrent = ['AC-P04-Tumor-SM-63B4Q', 'AC-P09-Tumor-SM-63B52', 'AC-P10-Tumor-SM-63B55',\n",
    "                        'AC-P12-Tumor-SM-63B59', 'AC-P15-Tumor-SM-63B5F']\n",
    "\n",
    "list_extension_primary = ['AC-V08-Tumor-SM-9J45Z', 'AC-V10-Tumor-SM-9J465',\n",
    "                         'AC-V11-Tumor-SM-9J467', 'AC-V12-Tumor-SM-9N4BH', 'AC-V14-Tumor-SM-9J46D', \n",
    "                          'AC-V15-Tumor-SM-9J46F', 'AC-V16-Tumor-SM-9J46G', 'AC-V18-Tumor-SM-9J46J']\n",
    "list_extension_recurrent =['AC-V08-Tumor-SM-9J461', 'AC-V08-Tumor-SM-9J462', 'AC-V10-Tumor-SM-9J466',\n",
    "                          'AC-V11-Tumor-SM-9J468', 'AC-V12-Tumor-SM-9J46A', 'AC-V14-Tumor-SM-9J46C', \n",
    "                           'AC-V15-Tumor-SM-9J46E','AC-V16-Tumor-SM-9J46H', 'AC-V18-Tumor-SM-9J46K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Because of how I wrote the functions \"format_pilot_load\" and \"format_extension_load\", \n",
    "# I will move the lists created above into dataframes\n",
    "df_list_pilot_primary = pd.DataFrame([], columns = list_pilot_primary)\n",
    "df_list_pilot_recurrent = pd.DataFrame([], columns = list_pilot_recurrent)\n",
    "\n",
    "df_list_extension_primary = pd.DataFrame([], columns = list_extension_primary)\n",
    "df_list_extension_recurrent = pd.DataFrame([], columns = list_extension_recurrent)\n",
    "\n",
    "df_list_pilot_primary_plt = pd.DataFrame([], columns = list_pilot_primary_plt)\n",
    "df_list_extension_recurrent_plt = pd.DataFrame([], columns = list_extension_recurrent_plt)\n",
    "\n",
    "# Import files necessary to compute mutational burden\n",
    "pilot_mutsig_rates = pd.read_csv('storage/mutsig-pilot/patient_counts_and_rates.txt', sep = '\\t')\n",
    "extension_all_calls = pd.read_csv('filter_unmatched/all_genes/df_filter_pass.txt', sep = '\\t', low_memory = False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Compute mutational burden\n",
    "# This is very similar to the calculatiosn done in Figure 1 and Figure 3 but we are now performing \n",
    "# the calculation for all samples, rather than just the ones required for those figures\n",
    "\n",
    "df_pilot_primary = format_pilot_load(pilot_mutsig_rates, df_list_pilot_primary)\n",
    "df_pilot_recurrent = format_pilot_load(pilot_mutsig_rates, df_list_pilot_recurrent)\n",
    "\n",
    "df_extension_primary = format_extension_load(info_extension, extension_all_calls, df_list_extension_primary)\n",
    "df_extension_recurrent = format_extension_load(info_extension, extension_all_calls, df_list_extension_recurrent)\n",
    "\n",
    "df_pilot_primary_plt = format_pilot_load(pilot_mutsig_rates, df_list_pilot_primary_plt)\n",
    "df_extension_recurrent_plt = format_extension_load(info_extension, extension_all_calls, df_list_extension_recurrent_plt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## We now combine populations of primary and recurrent tumors together between the pilot\n",
    "# and extension cohort\n",
    "\n",
    "# We combine the total burden of the primary cohort with nonsyn burden of the extension cohort\n",
    "total_primary = df_pilot_primary.ix['Total', :].tolist() + df_extension_primary.ix['Nonsyn', :].tolist()\n",
    "total_recurrent = df_pilot_recurrent.ix['Total', :].tolist() + df_extension_recurrent.ix['Nonsyn', :].tolist()\n",
    "\n",
    "# We combine the total burden of the primary cohort with nonsyn burden of the extension cohort\n",
    "nonsyn_primary = df_pilot_primary.ix['Nonsyn', :].tolist() + df_extension_primary.ix['Nonsyn', :].tolist()\n",
    "nonsyn_recurrent = df_pilot_recurrent.ix['Nonsyn', :].tolist() + df_extension_recurrent.ix['Nonsyn', :].tolist()\n",
    "\n",
    "print 'We compare the mutational burden of primary and recurrent tumors across both cohorts \\n'\n",
    "\n",
    "statistic, pvalue = stats.ttest_ind(total_primary, total_recurrent, equal_var = False)\n",
    "print 'Total burden of the pilot cohort:', str(statistic), str(pvalue)\n",
    "statistic, pvalue = stats.ttest_ind(nonsyn_primary, nonsyn_recurrent, equal_var = False)\n",
    "print 'Nonsyn burden of the pilot cohort:', str(statistic), str(pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print 'Total burden of the pilot cohort with the estimated burden of the extension cohort,'\n",
    "print 'Mean burden of primary tumors:', str(np.mean(total_primary))\n",
    "print 'Mean burden of recurrent tumors:', str(np.mean(total_recurrent))\n",
    "print 'Median burden of primary tumors:', str(np.median(total_primary))\n",
    "print 'Median burden of recurrent tumors:', str(np.median(total_recurrent))\n",
    "\n",
    "print '\\n'\n",
    "\n",
    "print 'Nonsyn burden of the pilot cohort with the estimated burden of the extension cohort,'\n",
    "print 'Mean burden of primary tumors:', str(np.mean(nonsyn_primary))\n",
    "print 'Mean burden of recurrent tumors:', str(np.mean(nonsyn_recurrent))\n",
    "print 'Median burden of primary tumors:', str(np.median(nonsyn_primary))\n",
    "print 'Median burden of recurrent tumors:', str(np.median(nonsyn_recurrent))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
